<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-06-27" />

<title>Learning about antibodies</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Learning about antibodies</h1>
<h4 class="date">2023-06-27</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-06-27
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity itâ€™s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreec90d5b7e3e30abadd7cb2d5aa30f7936c6bf7229targetblankc90d5b7a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/c90d5b7e3e30abadd7cb2d5aa30f7936c6bf7229" target="_blank">c90d5b7</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreec90d5b7e3e30abadd7cb2d5aa30f7936c6bf7229targetblankc90d5b7a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/c90d5b7e3e30abadd7cb2d5aa30f7936c6bf7229" target="_blank">c90d5b7</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.1.2/
    Ignored:    r_packages_4.2.0/
    Ignored:    r_packages_4.2.2/
    Ignored:    r_packages_4.3.0/

Untracked files:
    Untracked:  analysis/cell_ranger.Rmd
    Untracked:  analysis/tss_xgboost.Rmd
    Untracked:  code/multiz100way/
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz.tbi
    Untracked:  data/ncrna_NONCODE[v3.0].fasta.tar.gz
    Untracked:  data/ncrna_noncode_v3.fa
    Untracked:  data/netmhciipan.out.gz
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/graph.Rmd

</code></pre>
<p>
Note that any generated files, e.g.Â HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/antibody.Rmd</code>) and HTML
(<code>docs/antibody.html</code>) files. If youâ€™ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/c90d5b7e3e30abadd7cb2d5aa30f7936c6bf7229/analysis/antibody.Rmd" target="_blank">c90d5b7</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-27
</td>
<td>
CDR sequences
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/b707032ebe7f14fad454e75f0836e30b8d0263e8/docs/antibody.html" target="_blank">b707032</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/9161c8658327e8aa6d805f44f8c27749675a38da/analysis/antibody.Rmd" target="_blank">9161c86</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-26
</td>
<td>
Learning about antibodies
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Notes based on the study: <a
href="https://pubmed.ncbi.nlm.nih.gov/9359721/">Development of a
humanized monoclonal antibody (MEDI-493) with potent in vitro and in
vivo activity against respiratory syncytial virus</a>.</p>
<p>This study describes the generation of a humanised monoclonal
antibody, MEDI-493, that recognises a conserved neutralising epitope on
the F glycoprotein of RSV. Broad neutralisation of a panel of 57
clinical isolates of the RSV A and B subtypes was demonstrated.</p>
<div id="antibody-information" class="section level2">
<h2>Antibody information</h2>
<p><a
href="https://absoluteantibody.com/product/anti-rsv-f-protein-medi-493-palivizumab-hu129/">Additional
info</a>.</p>
<ul>
<li>UniProt Accession Number of Target Protein: P03420</li>
<li>Alternative Name(s) of Target: F2; Fusion glycoprotein F0; Protein
F; F gycoprotein; Human respiratory syncytial virus; humanized mAb
129</li>
<li>Immunogen: The original antibody was generated by immunizing female
BALB/c mice by primary intranasal infection with A2 strain of RSV.</li>
<li>Specificity: This antibody recognizes the site A of the RSV F
glycoprotein. This antibody binds the NSELLSLINDMPITNDQKKLMSNN epitope
(PMID: <a
href="https://pubmed.ncbi.nlm.nih.gov/20098425/">20098425</a>).</li>
</ul>
</div>
<div id="fusion-glycoprotein-sequence" class="section level2">
<h2>Fusion glycoprotein sequence</h2>
<p><a href="https://www.uniprot.org/uniprotkb/P03420/entry">Fusion
glycoprotein F0</a> of Human respiratory syncytial virus A (strain
A2).</p>
<pre><code>&gt;sp|P03420|FUS_HRSVA Fusion glycoprotein F0 OS=Human respiratory syncytial virus A (strain A2) OX=11259 GN=F PE=1 SV=1
MELLILKANAITTILTAVTFCFASGQNITEEFYQSTCSAVSKGYLSALRTGWYTSVITIE
LSNIKENKCNGTDAKVKLIKQELDKYKNAVTELQLLMQSTPPTNNRARRELPRFMNYTLN
NAKKTNVTLSKKRKRRFLGFLLGVGSAIASGVAVSKVLHLEGEVNKIKSALLSTNKAVVS
LSNGVSVLTSKVLDLKNYIDKQLLPIVNKQSCSISNIETVIEFQQKNNRLLEITREFSVN
AGVTTPVSTYMLT NSELLSLINDMPITNDQKKLMSNN VQIVRQQSYSIMSIIKEEVLAYV
VQLPLYGVIDTPCWKLHTSPLCTTNTKEGSNICLTRTDRGWYCDNAGSVSFFPQAETCKV
QSNRVFCDTMNSLTLPSEINLCNVDIFNPKYDCKIMTSKTDVSSSVITSLGAIVSCYGKT
KCTASNKNRGIIKTFSNGCDYVSNKGMDTVSVGNTLYYVNKQEGKSLYVKGEPIINFYDP
LVFPSDEFDASISQVNEKINQSLAFIRKSDELLHNVNAGKSTTNIMITTIIIVIIVILLS
LIAVGLLLYCKARSTPVTLSKDQLSGINNIAFSN</code></pre>
</div>
<div id="antibody-sequence" class="section level2">
<h2>Antibody sequence</h2>
<p>Design of humanised VL and VH segments based on murine monoclonal
antibody 1129. Human framework regions were derived from K102 for VL. VH
framework (FR1) region was derived from Cor; remaining framework regions
were derived from CE-1 sequence.</p>
<p>VL.</p>
<pre><code>FR1                     CDR1       FR2             CDR2
DIQMTQSPSTLSASVGDRVTITC KCQLSVGYMH WYQQKPGKAPKLLIY DTSKLAS

FR3                              CDR3      FR4
GVPSRFSGSGSGTEFTLTISSLQPDDFATYYC FQGSGYPFT FGGGTKLEIK</code></pre>
<p>VH.</p>
<pre><code>FR1                            CDR1    FR2            CDR2
QVTLRESGPALVKPTQTLTLTCTFSGFSLS TSGMSVG WIRQPPGKALEWLA DIWWDDKKDYNPSLKS

FR3                              CDR3       FR4
RLTISKDTSKNQVVLKVTNMDPADTATYYCAR SMITNWYFDV WGAGTTVTVSS</code></pre>
</div>
<div id="basics" class="section level2">
<h2>Basics</h2>
<p>Notes from <a
href="https://www.genscript.com/gsfiles/techfiles/Analyzing_antibody_sequence_for_recombinant_antibody_expression.pdf">Analysing
antibody sequence for recombinant antibody expression</a>.</p>
<ul>
<li>An antibody (a.k.a. immunoglobulin) is a large Y-shaped protein
produced by B cells. They are used by the immune system to neutralise
foreign objects.</li>
<li>Fab region determines its antigen specificity (Fv region and
CDR).</li>
<li>Fc region determines the effects.</li>
</ul>
<p><img src="assets/igg.png" /></p>
<ul>
<li>Each antibody (Ab) only binds to a specific antigen (Ag).</li>
<li>Ab and Ag interact by spatial complementarity.</li>
<li>Ab-Ag interaction are based on weak and non-covalent binding between
Ab and Ag:
<ul>
<li>Electrostatic interactions</li>
<li>Hydrogen bonds</li>
<li>Van der Waals forces</li>
<li>Hydrophobic interactions</li>
</ul></li>
<li>The hinge region allows better binding.</li>
<li>Ab-Ag binding is reversible.</li>
</ul>
<p><img src="assets/ab_ag.png" /></p>
<ul>
<li>Ab binding residues are mainly located in Complementarity
Determining Regions (CDRs)</li>
<li>6 CDRs form the combining site: L1, L2, L3, H1, H2, and H3</li>
<li>CDR-H3 and L3 have a distinctive role in antigen recognition.</li>
</ul>
<p><img src="assets/ab_cdr.png" /></p>
<p>The <a
href="https://en.wikipedia.org/wiki/Framework_region">framework
region</a> is a subdivision of the variable region (Fab) of the
antibody. The variable region is composed of seven amino acid regions,
four of which are framework regions (FR1-4) and three of which are
hypervariable regions. The framework region makes up about 85% of the
variable region. Located on the tips of the Y-shaped molecule, the
framework regions are responsible for acting as a scaffold for the
complementarity determining regions (CDR), also referred to as
hypervariable regions, of the Fab. These CDRs are in direct contact with
the antigen and are involved in binding antigen, while the framework
regions support the binding of the CDR to the antigen and aid in
maintaining the overall structure of the four variable domains on the
antibody.</p>
<p><img
src="assets/Antibody_Structure_and_Antigen_Binding_Regions.jpg" /></p>
<div id="identifying-cdrs-by-sequence" class="section level3">
<h3>Identifying CDRs by sequence</h3>
<p>CDR-L1</p>
<ul>
<li>Start - approximately residue 24</li>
<li>Residue before - <strong>always</strong> a Cys</li>
<li>Residue after - <strong>always</strong> a Trp. Typically
Trp-Tyr-Gln, but also Trp-Leu-Gln, Trp-Phe-Gln, Trp-Tyr-Leu</li>
<li>Length - 10 to 17 residues</li>
</ul>
<p>CDR-L2</p>
<ul>
<li>Start - <strong>always</strong> 16 residues after the end of L1</li>
<li>Residue before - generally Ile-Tyr, but also, Val-Tyr, Ile-Lys,
Ile-Phe</li>
<li>Length - <strong>always</strong> 7 residues (except NEW [7FAB],
which has a deletion in this region.)</li>
</ul>
<p>CDR-L3</p>
<ul>
<li>Start - <strong>always</strong> 33 residues after end of L2 (except
NEW [7FAB], which has the deletion at the end of CDR-L2)</li>
<li>Residue before - <strong>always</strong> Cys</li>
<li>Residue after - <strong>always</strong> Phe-Gly-XXX-Gly</li>
<li>Length - 7 to 11 residues</li>
</ul>
<p>CDR-H1</p>
<ul>
<li>Start - approximately residue 26 (<strong>always</strong> 4 after a
Cys)</li>
<li>Residue before - <strong>always</strong> Cys-XXX-XXX-XXX</li>
<li>Residue after - <strong>always</strong> a Trp. Typically Trp-Val,
but also, Trp-Ile, Trp-Ala</li>
<li>Length - 10 to 12 residues</li>
</ul>
<p>CDR-H2</p>
<ul>
<li>Start - <strong>always</strong> 15 residues after the end of Kabat /
AbM definition of CDR-H1</li>
<li>Residue before - typically Leu-Glu-Trp-Ile-Gly, but a number of
variations</li>
<li>Residue after - Lys/Arg-Leu/Ile/Val/Phe/Thr/Ala-Thr/Ser/Ile/Ala</li>
<li>Length - Kabat definition 16 to 19 residues</li>
</ul>
<p>CDR-H3</p>
<ul>
<li>Start - <strong>always</strong> 33 residues after end of CHR-H2
(<strong>always</strong> 2 after a Cys)</li>
<li>Residue before - <strong>always</strong> Cys-XXX-XXX (typically
Cys-Ala-Arg)</li>
<li>Residue after - <strong>always</strong> Trp-Gly-XXX-Gly</li>
<li>Length - 3 to 25 residues</li>
</ul>
<p>Leader sequence-FR1-CDR1-FR2-CDR2-FR3-CDR3-FR4-Constant region-Stop
codon</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.2 forcats_1.0.0   stringr_1.5.0   dplyr_1.1.2    
 [5] purrr_1.0.1     readr_2.1.4     tidyr_1.3.0     tibble_3.2.1   
 [9] ggplot2_3.4.2   tidyverse_2.0.0 workflowr_1.7.0

loaded via a namespace (and not attached):
 [1] sass_0.4.5       utf8_1.2.3       generics_0.1.3   stringi_1.7.12  
 [5] hms_1.1.3        digest_0.6.31    magrittr_2.0.3   timechange_0.2.0
 [9] evaluate_0.20    grid_4.3.0       fastmap_1.1.1    rprojroot_2.0.3 
[13] jsonlite_1.8.5   processx_3.8.1   whisker_0.4.1    ps_1.7.5        
[17] promises_1.2.0.1 httr_1.4.5       fansi_1.0.4      scales_1.2.1    
[21] jquerylib_0.1.4  cli_3.6.1        rlang_1.1.0      munsell_0.5.0   
[25] withr_2.5.0      cachem_1.0.7     yaml_2.3.7       tools_4.3.0     
[29] tzdb_0.3.0       colorspace_2.1-0 httpuv_1.6.9     vctrs_0.6.2     
[33] R6_2.5.1         lifecycle_1.0.3  git2r_0.32.0     fs_1.6.2        
[37] pkgconfig_2.0.3  callr_3.7.3      pillar_1.9.0     bslib_0.4.2     
[41] later_1.3.0      gtable_0.3.3     glue_1.6.2       Rcpp_1.0.10     
[45] xfun_0.39        tidyselect_1.2.0 rstudioapi_0.14  knitr_1.42      
[49] htmltools_0.5.5  rmarkdown_2.21   compiler_4.3.0   getPass_0.2-2   </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
