<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-02-13" />

<title>Using the Bioconductor ensembldb package</title>

<script src="site_libs/header-attrs-2.20/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Using the Bioconductor ensembldb
package</h1>
<h4 class="date">2023-02-13</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-02-13
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree28276f9e28ef38c5c899d0c12642d27625a27635targetblank28276f9a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/28276f9e28ef38c5c899d0c12642d27625a27635" target="_blank">28276f9</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree28276f9e28ef38c5c899d0c12642d27625a27635targetblank28276f9a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/28276f9e28ef38c5c899d0c12642d27625a27635" target="_blank">28276f9</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.1.2/
    Ignored:    r_packages_4.2.0/

Untracked files:
    Untracked:  analysis/cell_ranger.Rmd
    Untracked:  analysis/tss_xgboost.Rmd
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz.tbi
    Untracked:  data/ncrna_NONCODE[v3.0].fasta.tar.gz
    Untracked:  data/ncrna_noncode_v3.fa

Unstaged changes:
    Modified:   analysis/graph.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/ensembldb.Rmd</code>) and HTML
(<code>docs/ensembldb.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/28276f9e28ef38c5c899d0c12642d27625a27635/analysis/ensembldb.Rmd" target="_blank">28276f9</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-13
</td>
<td>
Using ensembldb
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The <code>ensembldb</code> package can be used to retrieve genomic
and protein annotations and to map between protein, transcript, and
genome coordinates. This mapping relies on annotations of proteins
(their sequences) to their encoding transcripts which are stored in
<code>EnsDb</code> databases.</p>
<blockquote>
<p>All functions, except <code>proteinToGenome</code> and
<code>transcriptToGenome</code> return <code>IRanges</code> with
negative coordinates if the mapping failed (e.g. because the identifier
is unknown to the database, or if, for mappings to and from protein
coordinates, the input coordinates are not within the coding region of a
transcript). <code>proteinToGenome</code> and
<code>transcriptToGenome</code> return empty <code>GRanges</code> if
mappings fail.</p>
</blockquote>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>To begin, install the <a
href="https://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a>
and <a
href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a>
packages.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

if (!require(&quot;ensembldb&quot;, quietly = TRUE))
  BiocManager::install(&quot;ensembldb&quot;)

if (!require(&quot;AnnotationHub&quot;, quietly = TRUE))
  BiocManager::install(&quot;AnnotationHub&quot;)

library(ensembldb)
library(AnnotationHub)</code></pre>
</div>
<div id="annotationhub" class="section level2">
<h2>AnnotationHub</h2>
<blockquote>
<p>The AnnotationHub server provides easy R / Bioconductor access to
large collections of publicly available whole genome resources, e.g,.
ENSEMBL genome fasta or gtf files, UCSC chain resources, ENCODE data
tracks at UCSC, etc.</p>
</blockquote>
<p>Create an <code>AnnotationHub</code> object.</p>
<pre class="r"><code>ah &lt;- AnnotationHub(ask = FALSE)</code></pre>
<pre><code>snapshotDate(): 2022-10-31</code></pre>
<p>Query.</p>
<pre class="r"><code>ensdb_homo &lt;- query(ah, c(&quot;EnsDb&quot;, &quot;Homo sapiens&quot;))
ensdb_homo</code></pre>
<pre><code>AnnotationHub with 23 records
# snapshotDate(): 2022-10-31
# $dataprovider: Ensembl
# $species: Homo sapiens
# $rdataclass: EnsDb
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., &#39;object[[&quot;AH53211&quot;]]&#39; 

             title                             
  AH53211  | Ensembl 87 EnsDb for Homo Sapiens 
  AH53715  | Ensembl 88 EnsDb for Homo Sapiens 
  AH56681  | Ensembl 89 EnsDb for Homo Sapiens 
  AH57757  | Ensembl 90 EnsDb for Homo Sapiens 
  AH60773  | Ensembl 91 EnsDb for Homo Sapiens 
  ...        ...                               
  AH95744  | Ensembl 104 EnsDb for Homo sapiens
  AH98047  | Ensembl 105 EnsDb for Homo sapiens
  AH100643 | Ensembl 106 EnsDb for Homo sapiens
  AH104864 | Ensembl 107 EnsDb for Homo sapiens
  AH109336 | Ensembl 108 EnsDb for Homo sapiens</code></pre>
<p>Latest available Ensembl version.</p>
<pre class="r"><code>latest &lt;- nrow(mcols(ensdb_homo))
edb &lt;- ensdb_homo[[latest]]</code></pre>
<pre><code>loading from cache</code></pre>
<pre class="r"><code>edb</code></pre>
<pre><code>EnsDb for Ensembl:
|Backend: SQLite
|Db type: EnsDb
|Type of Gene ID: Ensembl Gene ID
|Supporting package: ensembldb
|Db created by: ensembldb package from Bioconductor
|script_version: 0.3.7
|Creation time: Fri Oct 28 05:24:43 2022
|ensembl_version: 108
|ensembl_host: localhost
|Organism: Homo sapiens
|taxonomy_id: 9606
|genome_build: GRCh38
|DBSCHEMAVERSION: 2.2
| No. of genes: 70616.
| No. of transcripts: 275721.
|Protein data available.</code></pre>
</div>
<div id="mapping-protein-coordinates-to-the-genome"
class="section level2">
<h2>Mapping protein coordinates to the genome</h2>
<blockquote>
<p>The <code>proteinToGenome</code> function allows to map coordinates
within the amino acid sequence of a protein to the corresponding DNA
sequence on the genome. A protein identifier and the coordinates of the
sequence within its amino acid sequence are required and have to be
passed as an <code>IRanges</code> object to the function. The protein
identifier can either be passed as names of this object, or provided in
a metadata column (mcols).</p>
</blockquote>
<p>The example below (from the vignette) maps positions 5 to 9 within
the amino acid sequence of the protein ENSP00000385415.</p>
<pre class="r"><code>GAGE10_prt &lt;- IRanges(start = 5, end = 9, names = &quot;ENSP00000385415&quot;)
GAGE10_gnm &lt;- proteinToGenome(GAGE10_prt, edb)</code></pre>
<pre><code>Fetching CDS for 1 proteins ... 1 found
Checking CDS and protein sequence lengths ... 1/1 OK</code></pre>
<pre class="r"><code>GAGE10_gnm</code></pre>
<pre><code>$ENSP00000385415
GRanges object with 1 range and 7 metadata columns:
      seqnames            ranges strand |      protein_id           tx_id
         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt;
  [1]        X 49304872-49304886      + | ENSP00000385415 ENST00000407599
              exon_id exon_rank    cds_ok protein_start protein_end
          &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;   &lt;integer&gt;
  [1] ENSE00001692657         2      TRUE             5           9
  -------
  seqinfo: 1 sequence from GRCh38 genome</code></pre>
<p>The result is returned in a list, with one element for each range in
the input <code>IRanges</code>.</p>
<p>Below is an example with two proteins.</p>
<pre class="r"><code>two_prt &lt;- IRanges(
  start = c(6, 15),
  end = c(6, 15),
  names = c(&quot;ENSP00000366863&quot;, &quot;ENSP00000358262&quot;)
)

two_prt_to_gnm &lt;- proteinToGenome(two_prt, edb)</code></pre>
<pre><code>Fetching CDS for 2 proteins ... 2 found
Checking CDS and protein sequence lengths ... 2/2 OK</code></pre>
<pre class="r"><code>two_prt_to_gnm</code></pre>
<pre><code>$ENSP00000366863
GRanges object with 1 range and 7 metadata columns:
      seqnames            ranges strand |      protein_id           tx_id
         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt;
  [1]       13 75481750-75481752      - | ENSP00000366863 ENST00000377636
              exon_id exon_rank    cds_ok protein_start protein_end
          &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;   &lt;integer&gt;
  [1] ENSE00003893703         1      TRUE             6           6
  -------
  seqinfo: 2 sequences from GRCh38 genome

$ENSP00000358262
GRanges object with 1 range and 7 metadata columns:
      seqnames              ranges strand |      protein_id           tx_id
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;     &lt;character&gt;
  [1]        1 147242746-147242748      + | ENSP00000358262 ENST00000369258
              exon_id exon_rank    cds_ok protein_start protein_end
          &lt;character&gt; &lt;integer&gt; &lt;logical&gt;     &lt;integer&gt;   &lt;integer&gt;
  [1] ENSE00003728289         1      TRUE            15          15
  -------
  seqinfo: 2 sequences from GRCh38 genome</code></pre>
<p>We use <code>sapply()</code> to convert the results into a data
frame.</p>
<pre class="r"><code>get_pos &lt;- function(x, add_chr = TRUE){
  chr &lt;- as.character(seqnames(x))
  if(add_chr){
    chr &lt;- paste0(&quot;chr&quot;, chr)
  }
  list(
    chr = chr,
    start = start(x),
    end = end(x)
  )
}

as.data.frame(
  t(sapply(two_prt_to_gnm, get_pos))
)</code></pre>
<pre><code>                  chr     start       end
ENSP00000366863 chr13  75481750  75481752
ENSP00000358262  chr1 147242746 147242748</code></pre>
</div>
<div id="further-reading" class="section level2">
<h2>Further reading</h2>
<ul>
<li><a
href="https://bioconductor.org/packages/release/bioc/vignettes/ensembldb/inst/doc/coordinate-mapping.html">Mapping
between genome, transcript, and protein coordinates</a></li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] AnnotationHub_3.6.0     BiocFileCache_2.6.0     dbplyr_2.3.0           
 [4] ensembldb_2.22.0        AnnotationFilter_1.22.0 GenomicFeatures_1.50.4 
 [7] AnnotationDbi_1.60.0    Biobase_2.58.0          GenomicRanges_1.50.2   
[10] GenomeInfoDb_1.34.9     IRanges_2.32.0          S4Vectors_0.36.1       
[13] BiocGenerics_0.44.0     BiocManager_1.30.19     workflowr_1.7.0        

loaded via a namespace (and not attached):
 [1] ProtGenerics_1.30.0           bitops_1.0-7                 
 [3] matrixStats_0.63.0            fs_1.6.1                     
 [5] bit64_4.0.5                   filelock_1.0.2               
 [7] progress_1.2.2                httr_1.4.4                   
 [9] rprojroot_2.0.3               tools_4.2.0                  
[11] bslib_0.4.2                   utf8_1.2.3                   
[13] R6_2.5.1                      lazyeval_0.2.2               
[15] DBI_1.1.3                     withr_2.5.0                  
[17] tidyselect_1.2.0              prettyunits_1.1.1            
[19] processx_3.8.0                bit_4.0.5                    
[21] curl_5.0.0                    compiler_4.2.0               
[23] git2r_0.31.0                  cli_3.6.0                    
[25] xml2_1.3.3                    DelayedArray_0.24.0          
[27] rtracklayer_1.58.0            sass_0.4.5                   
[29] callr_3.7.3                   rappdirs_0.3.3               
[31] stringr_1.5.0                 digest_0.6.31                
[33] Rsamtools_2.14.0              rmarkdown_2.20               
[35] XVector_0.38.0                pkgconfig_2.0.3              
[37] htmltools_0.5.4               MatrixGenerics_1.10.0        
[39] fastmap_1.1.0                 rlang_1.0.6                  
[41] rstudioapi_0.14               RSQLite_2.2.20               
[43] shiny_1.7.4                   jquerylib_0.1.4              
[45] BiocIO_1.8.0                  generics_0.1.3               
[47] jsonlite_1.8.4                BiocParallel_1.32.5          
[49] dplyr_1.1.0                   RCurl_1.98-1.10              
[51] magrittr_2.0.3                GenomeInfoDbData_1.2.9       
[53] Matrix_1.5-3                  Rcpp_1.0.10                  
[55] fansi_1.0.4                   lifecycle_1.0.3              
[57] stringi_1.7.12                whisker_0.4.1                
[59] yaml_2.3.7                    SummarizedExperiment_1.28.0  
[61] zlibbioc_1.44.0               grid_4.2.0                   
[63] blob_1.2.3                    parallel_4.2.0               
[65] promises_1.2.0.1              crayon_1.5.2                 
[67] lattice_0.20-45               Biostrings_2.66.0            
[69] hms_1.1.2                     KEGGREST_1.38.0              
[71] knitr_1.42                    ps_1.7.2                     
[73] pillar_1.8.1                  rjson_0.2.21                 
[75] codetools_0.2-18              biomaRt_2.54.0               
[77] BiocVersion_3.16.0            XML_3.99-0.13                
[79] glue_1.6.2                    evaluate_0.20                
[81] getPass_0.2-2                 png_0.1-8                    
[83] vctrs_0.5.2                   httpuv_1.6.8                 
[85] purrr_1.0.1                   assertthat_0.2.1             
[87] cachem_1.0.6                  xfun_0.37                    
[89] mime_0.12                     xtable_1.8-4                 
[91] restfulr_0.0.15               later_1.3.0                  
[93] tibble_3.1.8                  GenomicAlignments_1.34.0     
[95] memoise_2.0.1                 interactiveDisplayBase_1.36.0
[97] ellipsis_0.3.2               </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
