<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-10-23" />

<title>GSVA: gene set variation analysis</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GSVA: gene set variation analysis</h1>
<h4 class="date">2023-10-23</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-10-23
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree115752b21e76786cb7afc584780d9dc116e36894targetblank115752ba">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/115752b21e76786cb7afc584780d9dc116e36894" target="_blank">115752b</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree115752b21e76786cb7afc584780d9dc116e36894targetblank115752ba"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/115752b21e76786cb7afc584780d9dc116e36894" target="_blank">115752b</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.3.0/
    Ignored:    r_packages_4.3.1/

Untracked files:
    Untracked:  analysis/cell_ranger.Rmd
    Untracked:  analysis/complex_heatmap.Rmd
    Untracked:  analysis/sleuth.Rmd
    Untracked:  analysis/tss_xgboost.Rmd
    Untracked:  code/multiz100way/
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz.tbi
    Untracked:  data/ncrna_NONCODE[v3.0].fasta.tar.gz
    Untracked:  data/ncrna_noncode_v3.fa
    Untracked:  data/netmhciipan.out.gz
    Untracked:  data/test
    Untracked:  export/davetang039sblog.WordPress.2023-06-30.xml
    Untracked:  export/output/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/graph.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/gsva.Rmd</code>) and HTML
(<code>docs/gsva.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/115752b21e76786cb7afc584780d9dc116e36894/analysis/gsva.Rmd" target="_blank">115752b</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-10-23
</td>
<td>
Technical notes
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/14e3e6b20a8c19ff7f18c7610f41e480a95077af/docs/gsva.html" target="_blank">14e3e6b</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-10-23
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/9b9327a0ae0192e36a438c7b49a61b7a7ad383b9/analysis/gsva.Rmd" target="_blank">9b9327a</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-10-23
</td>
<td>
No progress bar
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/b0af0930f65b6a1f879b667c4b0df492500003a2/docs/gsva.html" target="_blank">b0af093</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-10-23
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/aeb90e1edd28e75a88d9dcaf4e02bda4eed2b967/analysis/gsva.Rmd" target="_blank">aeb90e1</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-10-23
</td>
<td>
GSVA
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Following the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.html">vignette</a>.</p>
<blockquote>
<p>Gene set variation analysis (GSVA) is a particular type of gene set
enrichment method that works on single samples and enables
pathway-centric analyses of molecular data by performing a conceptually
simple but powerful change in the functional unit of analysis, from
genes to gene sets. The GSVA package provides the implementation of four
single-sample gene set enrichment methods, concretely zscore, plage,
ssGSEA and its own called GSVA. While this methodology was initially
developed for gene expression data, it can be applied to other types of
molecular profiling data. In this vignette we illustrate how to use the
GSVA package with bulk microarray and RNA-seq expression data.</p>
</blockquote>
<blockquote>
<p>Gene set variation analysis (GSVA) provides an estimate of pathway
activity by transforming an input gene-by-sample expression data matrix
into a corresponding gene-set-by-sample expression data matrix. This
resulting expression data matrix can be then used with classical
analytical methods such as differential expression, classification,
survival analysis, clustering or correlation analysis in a
pathway-centric manner. One can also perform sample-wise comparisons
between pathways and other molecular data types such as microRNA
expression or binding data, copy-number variation (CNV) data or single
nucleotide polymorphisms (SNPs).</p>
</blockquote>
<div id="package" class="section level2">
<h2>Package</h2>
<p>Install GSVA. (Dependencies are listed in the Imports section in the
<a
href="https://github.com/rcastelo/GSVA/blob/devel/DESCRIPTION">DESCRIPTION</a>
file.)</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

if (!require(&quot;GSVA&quot;, quietly = TRUE))
  BiocManager::install(&quot;GSVA&quot;)</code></pre>
<p>Load package.</p>
<pre class="r"><code>library(GSVA)</code></pre>
</div>
<div id="quick-start" class="section level2">
<h2>Quick start</h2>
<p>Generate example expression matrix.</p>
<pre class="r"><code>p &lt;- 10000
n &lt;- 30
set.seed(1984)
X &lt;- matrix(
  rnorm(p*n),
  nrow=p,
  dimnames=list(paste0(&quot;g&quot;, 1:p), paste0(&quot;s&quot;, 1:n))
)
X[1:5, 1:5]</code></pre>
<pre><code>           s1         s2         s3          s4           s5
g1  0.4092032  1.4676435  0.3515056  1.53512312 -1.279009469
g2 -0.3230250 -1.8501416 -0.9198650  1.40036448  0.086613315
g3  0.6358523  1.6084120  1.6380322  0.23799146  0.216628121
g4 -1.8461288 -0.2928844  0.4651573 -0.09766558 -0.009887299
g5  0.9536474 -0.4816006  0.1807824  1.03141311  0.206414282</code></pre>
<p>Generate 100 gene sets that are contain from 10 to up to 100 genes
sampled from <code>1:p</code>.</p>
<pre class="r"><code>set.seed(1984)
gs &lt;- as.list(sample(10:100, size=100, replace=TRUE))

gs &lt;- lapply(gs, function(n, p){
  paste0(&quot;g&quot;, sample(1:p, size=n, replace=FALSE))
}, p)
names(gs) &lt;- paste0(&quot;gs&quot;, 1:length(gs))

sapply(gs, length)</code></pre>
<pre><code>  gs1   gs2   gs3   gs4   gs5   gs6   gs7   gs8   gs9  gs10  gs11  gs12  gs13 
   49    29    67    90    94    87    41    26    86    77    97    90    45 
 gs14  gs15  gs16  gs17  gs18  gs19  gs20  gs21  gs22  gs23  gs24  gs25  gs26 
   47    54    83    11    75    95    99    94    89    93    50    49    87 
 gs27  gs28  gs29  gs30  gs31  gs32  gs33  gs34  gs35  gs36  gs37  gs38  gs39 
   36    61    84    99    58    30    63    29    35    29    69    41    46 
 gs40  gs41  gs42  gs43  gs44  gs45  gs46  gs47  gs48  gs49  gs50  gs51  gs52 
   38    17    48    72    15    81   100    93    37    99    89    43    36 
 gs53  gs54  gs55  gs56  gs57  gs58  gs59  gs60  gs61  gs62  gs63  gs64  gs65 
   84    83    40    72    90    86    37    23    69    96    20    93    36 
 gs66  gs67  gs68  gs69  gs70  gs71  gs72  gs73  gs74  gs75  gs76  gs77  gs78 
   21    46    76    71    57    48    25    73    26    46    29    53    69 
 gs79  gs80  gs81  gs82  gs83  gs84  gs85  gs86  gs87  gs88  gs89  gs90  gs91 
   69    42    76    30    16    49    35    12    83    99    88    66    10 
 gs92  gs93  gs94  gs95  gs96  gs97  gs98  gs99 gs100 
   51    82    73    97    59    59    42    10    64 </code></pre>
<p>Calculate GSVA enrichment scores using the <code>gsva()</code>
function, which does all the work and requires the following two input
arguments:</p>
<ol style="list-style-type: decimal">
<li>A <strong>normalised</strong> gene expression dataset, which can be
provided in one of the following containers:
<ul>
<li>A matrix of expression values with genes corresponding to rows and
samples corresponding to columns.</li>
<li>An ExpressionSet object; see package Biobase.</li>
<li>A SummarizedExperiment object, see package
SummarizedExperiment.</li>
</ul></li>
<li>A collection of gene sets; which can be provided in one of the
following containers:
<ul>
<li>A list object where each element corresponds to a gene set defined
by a vector of gene identifiers, and the element names correspond to the
names of the gene sets.</li>
<li>A GeneSetCollection object; see package GSEABase.</li>
</ul></li>
</ol>
<blockquote>
<p>The first argument to the <code>gsva()</code> function is the gene
expression data matrix and the second the collection of gene sets. The
<code>gsva()</code> function can take the input expression data and gene
sets using different specialized containers that facilitate the access
and manipulation of molecular and phenotype data, as well as their
associated metadata. Another advanced features include the use of
on-disk and parallel backends to enable, respectively, using GSVA on
large molecular data sets and speed up computing time.</p>
</blockquote>
<p>The <code>gsva()</code> function will apply the following filters
before the actual calculations take place:</p>
<ol style="list-style-type: decimal">
<li>Discard genes in the input expression data matrix with constant
expression.</li>
<li>Discard genes in the input gene sets that do not map to a gene in
the input gene expression data matrix.</li>
<li>Discard gene sets that, after applying the previous filters, do not
meet a minimum and maximum size, which by default is 1 for the minimum
size and <code>Inf</code> for the maximum size.</li>
</ol>
<p>When <code>method="gsva"</code> is used (the default), the following
parameters can be tuned:</p>
<ul>
<li><code>kcdf</code>: The first step of the GSVA algorithm brings gene
expression profiles to a common scale by calculating an expression
statistic through a non-parametric estimation of the CDF across samples.
Such a non-parametric estimation employs a kernel function and the
<code>kcdf</code> parameter allows the user to specify three possible
values for that function:</li>
</ul>
<ol style="list-style-type: decimal">
<li>“Gaussian”, the default value, which is suitable for continuous
expression data, such as microarray fluorescent units in logarithmic
scale and RNA-seq log-CPMs, log-RPKMs or log-TPMs units of
expression;</li>
<li>“Poisson”, which is suitable for integer counts, such as those
derived from RNA-seq alignments;</li>
<li>“none”, which will enforce a direct estimation of the CDF without a
kernel function.</li>
</ol>
<ul>
<li><code>mx.diff</code>: The last step of the GSVA algorithm calculates
the gene set enrichment score from two Kolmogorov-Smirnov random walk
statistics. This parameter is a logical flag that allows the user to
specify two possible ways to do such calculation:</li>
</ul>
<ol style="list-style-type: decimal">
<li><code>TRUE</code>, the default value, where the enrichment score is
calculated as the magnitude difference between the largest positive and
negative random walk deviations;</li>
<li><code>FALSE</code>, where the enrichment score is calculated as the
maximum distance of the random walk from zero.</li>
</ol>
<ul>
<li><p><code>abs.ranking</code>: Logical flag used only when
<code>mx.diff=TRUE</code>. By default, <code>abs.ranking=FALSE</code>
and it implies that a modified Kuiper statistic is used to calculate
enrichment scores, taking the magnitude difference between the largest
positive and negative random walk deviations. When
<code>abs.ranking=TRUE</code> the original Kuiper statistic is used, by
which the largest positive and negative random walk deviations are added
together. In this case, gene sets with genes enriched on either extreme
(high or low) will be regarded as highly activated.</p></li>
<li><p><code>tau</code>: Exponent defining the weight of the tail in the
random walk. By default <code>tau=1</code>. When
<code>method="ssgsea"</code>, this parameter is also used and its
default value becomes then <code>tau=0.25</code> to match the
methodology described in (Barbie et al. 2009).</p></li>
</ul>
<p>In general, the default values for the previous parameters are
suitable for most analysis settings, which usually consist of some kind
of normalized continuous expression values.</p>
<pre class="r"><code>es_gsva &lt;- gsva(X, gs, verbose=FALSE)
dim(es_gsva)</code></pre>
<pre><code>[1] 100  30</code></pre>
<p>Median enrichment scores.</p>
<pre class="r"><code>apply(es_gsva, 2, median)</code></pre>
<pre><code>          s1           s2           s3           s4           s5           s6 
 0.009061514 -0.008458424 -0.005713628 -0.021937531  0.006417182  0.019422486 
          s7           s8           s9          s10          s11          s12 
 0.010140618  0.005812097  0.006495584  0.008887644  0.024577619 -0.031697634 
         s13          s14          s15          s16          s17          s18 
 0.001642502  0.008919786 -0.022622470  0.027695420 -0.015799537 -0.011108686 
         s19          s20          s21          s22          s23          s24 
-0.013956442 -0.015493300 -0.004809844 -0.014081494  0.026845336  0.023895676 
         s25          s26          s27          s28          s29          s30 
 0.006358240  0.010642450 -0.012690144 -0.005999451 -0.005058572 -0.012403422 </code></pre>
<blockquote>
<p>ssgsea (Barbie et al. 2009). Single sample GSEA (ssGSEA) is a
non-parametric method that calculates a gene set enrichment score per
sample as the normalized difference in empirical cumulative distribution
functions (CDFs) of gene expression ranks inside and outside the gene
set. By default, the implementation in the GSVA package follows the last
step described in (Barbie et al. 2009, online methods, pg. 2) by which
pathway scores are normalized, dividing them by the range of calculated
values. This normalization step may be switched off using the argument
ssgsea.norm in the call to the gsva() function; see below.</p>
</blockquote>
<pre class="r"><code>es_ssgsea &lt;- gsva(X, gs, method = &quot;ssgsea&quot;, verbose=FALSE)</code></pre>
<pre><code>[1] &quot;Calculating ranks...&quot;
[1] &quot;Calculating absolute values from ranks...&quot;
[1] &quot;Normalizing...&quot;</code></pre>
<pre class="r"><code>apply(es_ssgsea, 2, median)</code></pre>
<pre><code>       s1        s2        s3        s4        s5        s6        s7        s8 
0.1056332 0.1105148 0.1149790 0.1045303 0.1192256 0.1274702 0.1192217 0.1287768 
       s9       s10       s11       s12       s13       s14       s15       s16 
0.1316429 0.1257247 0.1293241 0.1083643 0.1222038 0.1012195 0.1018693 0.1174057 
      s17       s18       s19       s20       s21       s22       s23       s24 
0.1103184 0.1112410 0.1164373 0.1126160 0.1212393 0.1067666 0.1167117 0.1418775 
      s25       s26       s27       s28       s29       s30 
0.1221909 0.1235449 0.1112199 0.1012381 0.1254514 0.1210392 </code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] GSVA_1.48.3         BiocManager_1.30.22 workflowr_1.7.1    

loaded via a namespace (and not attached):
 [1] blob_1.2.4                  Biostrings_2.68.1          
 [3] bitops_1.0-7                fastmap_1.1.1              
 [5] SingleCellExperiment_1.22.0 RCurl_1.98-1.12            
 [7] promises_1.2.1              rsvd_1.0.5                 
 [9] XML_3.99-0.14               digest_0.6.33              
[11] lifecycle_1.0.3             processx_3.8.2             
[13] KEGGREST_1.40.1             RSQLite_2.3.1              
[15] magrittr_2.0.3              compiler_4.3.1             
[17] rlang_1.1.1                 sass_0.4.7                 
[19] tools_4.3.1                 utf8_1.2.3                 
[21] yaml_2.3.7                  knitr_1.44                 
[23] S4Arrays_1.0.6              bit_4.0.5                  
[25] DelayedArray_0.26.7         abind_1.4-5                
[27] BiocParallel_1.34.2         HDF5Array_1.28.1           
[29] BiocGenerics_0.46.0         grid_4.3.1                 
[31] stats4_4.3.1                fansi_1.0.5                
[33] git2r_0.32.0                beachmat_2.16.0            
[35] xtable_1.8-4                Rhdf5lib_1.22.1            
[37] SummarizedExperiment_1.30.2 cli_3.6.1                  
[39] rmarkdown_2.25              crayon_1.5.2               
[41] rstudioapi_0.15.0           httr_1.4.7                 
[43] DelayedMatrixStats_1.22.6   DBI_1.1.3                  
[45] cachem_1.0.8                rhdf5_2.44.0               
[47] stringr_1.5.0               zlibbioc_1.46.0            
[49] parallel_4.3.1              AnnotationDbi_1.62.2       
[51] XVector_0.40.0              matrixStats_1.0.0          
[53] vctrs_0.6.4                 Matrix_1.5-4.1             
[55] jsonlite_1.8.7              BiocSingular_1.16.0        
[57] callr_3.7.3                 IRanges_2.34.1             
[59] S4Vectors_0.38.2            bit64_4.0.5                
[61] irlba_2.3.5.1               GSEABase_1.62.0            
[63] jquerylib_0.1.4             annotate_1.78.0            
[65] glue_1.6.2                  codetools_0.2-19           
[67] ps_1.7.5                    stringi_1.7.12             
[69] later_1.3.1                 GenomeInfoDb_1.36.4        
[71] GenomicRanges_1.52.1        ScaledMatrix_1.8.1         
[73] tibble_3.2.1                pillar_1.9.0               
[75] htmltools_0.5.6.1           rhdf5filters_1.12.1        
[77] graph_1.78.0                GenomeInfoDbData_1.2.10    
[79] R6_2.5.1                    sparseMatrixStats_1.12.2   
[81] rprojroot_2.0.3             evaluate_0.22              
[83] Biobase_2.60.0              lattice_0.21-8             
[85] png_0.1-8                   memoise_2.0.1              
[87] httpuv_1.6.11               bslib_0.5.1                
[89] Rcpp_1.0.11                 whisker_0.4.1              
[91] xfun_0.40                   fs_1.6.3                   
[93] MatrixGenerics_1.12.3       getPass_0.2-2              
[95] pkgconfig_2.0.3            </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
