<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-12-13" />

<title>Downloading molecular signatures from MSigDB in R</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Downloading molecular signatures from
MSigDB in R</h1>
<h4 class="date">2023-12-13</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-12-13
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree5b0e19403f3c936540ae08673bd2826e92327a08targetblank5b0e194a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/5b0e19403f3c936540ae08673bd2826e92327a08" target="_blank">5b0e194</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree5b0e19403f3c936540ae08673bd2826e92327a08targetblank5b0e194a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/5b0e19403f3c936540ae08673bd2826e92327a08" target="_blank">5b0e194</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.3.2/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/msigdb.Rmd</code>) and HTML
(<code>docs/msigdb.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/5b0e19403f3c936540ae08673bd2826e92327a08/analysis/msigdb.Rmd" target="_blank">5b0e194</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-12-13
</td>
<td>
Downloading molecular signatures in R
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Following the <a
href="https://bioconductor.org/packages/release/data/experiment/vignettes/msigdb/inst/doc/msigdb.html">vignette</a>.</p>
<blockquote>
<p>The molecular signatures database (MSigDB) is one of the largest
collections of molecular signatures or gene expression signatures. A
variety of gene expression signatures are hosted on this database
including experimentally derived signatures and signatures representing
pathways and ontologies from other curated databases. This rich
collection of gene expression signatures (&gt;25,000) can facilitate a
wide variety of signature-based analyses, the most popular being gene
set enrichment analyses. These signatures can be used to perform
enrichment analysis in a DE experiment using tools such as {GSEA}, {fry}
(from {limma}) and {camera} (from {limma}). Alternatively, they can be
used to perform single-sample gene-set analysis of individual
transcriptomic profiles using approaches such as {singscore}, {ssGSEA}
and {GSVA}.</p>
<p>This package provides the gene sets in the MSigDB in the form of
<code>GeneSet</code> objects. This data structure is specifically
designed to store information about gene sets, including their member
genes and metadata. Other packages, such as {msigdbr} and {EGSEAdata}
provide these gene sets too, however, they do so by storing them as
lists or tibbles. These structures are not specific to gene sets
therefore do not allow storage of important metadata associated with
each gene set, for example, their short and long descriptions.
Additionally, the lack of structure allows creation of invalid gene
sets. Accessory functions implemented in the {GSEABase} package provide
a neat interface to interact with <code>GeneSet</code> objects.</p>
</blockquote>
<div id="package" class="section level2">
<h2>Package</h2>
<p>Install GSVA. (Dependencies are listed in the Imports section in the
<a
href="https://github.com/rcastelo/GSVA/blob/devel/DESCRIPTION">DESCRIPTION</a>
file.)</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

if (!require(&quot;msigdb&quot;, quietly = TRUE))
  BiocManager::install(&quot;msigdb&quot;)</code></pre>
<p>Load package.</p>
<pre class="r"><code>library(msigdb)
packageVersion(&quot;msigdb&quot;)</code></pre>
<pre><code>[1] &#39;1.10.0&#39;</code></pre>
</div>
<div id="downloading-the-msigdb-database" class="section level2">
<h2>Downloading the MSigDB database</h2>
<p>In order to download the MSigDB database, we need to load
{ExperimentHub} and {GSEABase}.</p>
<pre class="r"><code>library(ExperimentHub)
library(GSEABase)</code></pre>
<p>Query an <code>ExperimentHub</code> object.</p>
<pre class="r"><code>eh &lt;- ExperimentHub(ask = FALSE)
AnnotationHub::query(x = eh, pattern = &#39;msigdb&#39;)</code></pre>
<pre><code>ExperimentHub with 49 records
# snapshotDate(): 2023-10-24
# $dataprovider: Broad Institute, Emory University, EBI
# $species: Homo sapiens, Mus musculus
# $rdataclass: GSEABase::GeneSetCollection, list, data.frame
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., &#39;object[[&quot;EH5421&quot;]]&#39; 

           title                   
  EH5421 | msigdb.v7.2.hs.SYM      
  EH5422 | msigdb.v7.2.hs.EZID     
  EH5423 | msigdb.v7.2.mm.SYM      
  EH5424 | msigdb.v7.2.mm.EZID     
  EH6727 | MSigDB C8 MANNO MIDBRAIN
  ...      ...                     
  EH8296 | msigdb.v7.5.1.hs.SYM    
  EH8297 | msigdb.v7.5.1.mm.EZID   
  EH8298 | msigdb.v7.5.1.mm.idf    
  EH8299 | msigdb.v7.5.1.mm.SYM    
  EH8300 | imex_hsmm_0722          </code></pre>
<p>Specify a more specific pattern to look for only human
collections.</p>
<pre class="r"><code>AnnotationHub::query(x = eh, pattern = &#39;msigdb.*hs.SYM&#39;)</code></pre>
<pre><code>ExperimentHub with 7 records
# snapshotDate(): 2023-10-24
# $dataprovider: Broad Institute
# $species: Homo sapiens
# $rdataclass: GSEABase::GeneSetCollection
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., &#39;object[[&quot;EH5421&quot;]]&#39; 

           title                
  EH5421 | msigdb.v7.2.hs.SYM   
  EH6772 | msigdb.v7.3.hs.SYM   
  EH6778 | msigdb.v7.4.hs.SYM   
  EH7359 | msigdb.v7.5.hs.SYM   
  EH8284 | msigdb.v2022.1.hs.SYM
  EH8290 | msigdb.v2023.1.hs.SYM
  EH8296 | msigdb.v7.5.1.hs.SYM </code></pre>
<p>The experiment hubs seem to be ordered from earliest to latest.</p>
<pre class="r"><code>AnnotationHub::query(x = eh, pattern = &#39;msigdb.*hs.SYM&#39;) |&gt;
  tail(1) -&gt; msigdb_hs_latest

names(msigdb_hs_latest)</code></pre>
<pre><code>[1] &quot;EH8296&quot;</code></pre>
<pre class="r"><code>msigdb_hs_latest</code></pre>
<pre><code>ExperimentHub with 1 record
# snapshotDate(): 2023-10-24
# names(): EH8296
# package(): msigdb
# $dataprovider: Broad Institute
# $species: Homo sapiens
# $rdataclass: GSEABase::GeneSetCollection
# $rdatadateadded: 2023-07-03
# $title: msigdb.v7.5.1.hs.SYM
# $description: Gene expression signatures (Homo sapiens) from the Molecular...
# $taxonomyid: 9606
# $genome: NA
# $sourcetype: XML
# $sourceurl: https://data.broadinstitute.org/gsea-msigdb/msigdb/release/7.5...
# $sourcesize: NA
# $tags: c(&quot;Homo_sapiens_Data&quot;, &quot;Mus_musculus_Data&quot;) 
# retrieve record with &#39;object[[&quot;EH8296&quot;]]&#39; </code></pre>
<p>Data can be downloaded using the unique ID.</p>
<pre class="r"><code>eh[[names(msigdb_hs_latest)]]</code></pre>
<pre><code>GeneSetCollection
  names: chr1p11, chr1p12, ..., GOMF_STARCH_BINDING (45226 total)
  unique identifiers: RPL22P6, NBPF8, ..., POM121L15P (41072 total)
  types in collection:
    geneIdType: SymbolIdentifier (1 total)
    collectionType: BroadCollection (1 total)</code></pre>
<p>Data can also be downloaded using
<code>msigdb::getMsigdb()</code>.</p>
<pre class="r"><code>msigdb_ver &lt;- sub(pattern = &quot;msigdb.v(.*).hs.SYM&quot;, replacement = &quot;\\1&quot;, msigdb_hs_latest$title)

msigdb.hs &lt;- msigdb::getMsigdb(org = &quot;hs&quot;, id = &quot;SYM&quot;, version = msigdb_ver)</code></pre>
<pre><code>see ?msigdb and browseVignettes(&#39;msigdb&#39;) for documentation</code></pre>
<pre><code>loading from cache</code></pre>
<pre class="r"><code>msigdb.hs</code></pre>
<pre><code>GeneSetCollection
  names: chr1p11, chr1p12, ..., GOMF_STARCH_BINDING (45226 total)
  unique identifiers: RPL22P6, NBPF8, ..., POM121L15P (41072 total)
  types in collection:
    geneIdType: SymbolIdentifier (1 total)
    collectionType: BroadCollection (1 total)</code></pre>
</div>
<div id="accessing-the-geneset-and-genesetcollection-objects"
class="section level2">
<h2>Accessing the GeneSet and GeneSetCollection objects</h2>
<p>A <code>GeneSetCollection</code> object is effectively a list and
therefore all list processing functions work.</p>
<pre class="r"><code>length(msigdb.hs)</code></pre>
<pre><code>[1] 45226</code></pre>
<p>Each signature is stored in a <code>GeneSet</code> object and can be
processed using functions from the {GSEABase} package.</p>
<pre class="r"><code>gs &lt;- msigdb.hs[[1000]]
geneIds(gs)</code></pre>
<pre><code> [1] &quot;HOXC13&quot;      &quot;MYC&quot;         &quot;BTG3&quot;        &quot;PIK3CG&quot;      &quot;MXI1&quot;       
 [6] &quot;NCSTN&quot;       &quot;HOXA11&quot;      &quot;GADD45B&quot;     &quot;RFNG&quot;        &quot;FZD6&quot;       
[11] &quot;CCND2&quot;       &quot;PRKCD&quot;       &quot;SOCS2&quot;       &quot;TNFSF13B&quot;    &quot;FANCA&quot;      
[16] &quot;FOSB&quot;        &quot;VSX2&quot;        &quot;MBD1&quot;        &quot;PSEN2&quot;       &quot;HES3&quot;       
[21] &quot;PRDM1&quot;       &quot;POU2AF1&quot;     &quot;XBP1&quot;        &quot;STAT5A&quot;      &quot;AKT1&quot;       
[26] &quot;COMMD3-BMI1&quot; &quot;DTX2&quot;        &quot;MIB1&quot;        &quot;IRF4&quot;        &quot;TRAF3&quot;      
[31] &quot;IRF6&quot;       </code></pre>
<p>Details of a gene set.</p>
<pre class="r"><code>details(gs)</code></pre>
<pre><code>setName: SHIN_B_CELL_LYMPHOMA_CLUSTER_2 
geneIds: HOXC13, MYC, ..., IRF6 (total: 31)
geneIdType: Symbol
collectionType: Broad
  bcCategory: c2 (Curated)
  bcSubCategory: CGP
setIdentifier: LVY1HGGWMJ7:35020:Fri May 26 12:20:46 2023:93736
description: Cluster 2 of genes distinguishing among different B lymphocyte neoplasms.
  (longDescription available)
organism: Mus musculus
pubMedIds: 19010892
urls: https://data.broadinstitute.org/gsea-msigdb/msigdb/release/7.5.1/msigdb_v7.5.1.xml
contributor: Jessica Robertson
setVersion: 7.5.1
creationDate: </code></pre>
<p><a href="https://www.gsea-msigdb.org/gsea/msigdb">Human collection
types</a>:</p>
<ul>
<li>H - hallmark gene sets are coherently expressed signatures derived
by aggregating many MSigDB gene sets to represent well-defined
biological states or processes.</li>
<li>C1 - positional gene sets corresponding to human chromosome
cytogenetic bands.</li>
<li>C2 - curated gene sets from online pathway databases, publications
in PubMed, and knowledge of domain experts.</li>
<li>C3 - regulatory target gene sets based on gene target predictions
for microRNA seed sequences and predicted transcription factor binding
sites.</li>
<li>C4 - computational gene sets defined by mining large collections of
cancer-oriented expression data.</li>
<li>C5 - ontology gene sets consist of genes annotated by the same
ontology term.</li>
<li>C6 - oncogenic signature gene sets defined directly from microarray
gene expression data from cancer gene perturbations.</li>
<li>C7 - immunologic signature gene sets represent cell states and
perturbations within the immune system.</li>
<li>C8 - cell type signature gene sets curated from cluster markers
identified in single-cell sequencing studies of human tissue.</li>
</ul>
<pre class="r"><code>table(sapply(lapply(msigdb.hs, collectionType), bcCategory))</code></pre>
<pre><code>
   c1    c2    c3    c4    c5    c6    c7    c8     h 
  299  6180  3726   858 28005   189  5219   700    50 </code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] GSEABase_1.64.0      graph_1.80.0         annotate_1.80.0     
 [4] XML_3.99-0.16        AnnotationDbi_1.64.1 IRanges_2.36.0      
 [7] S4Vectors_0.40.2     Biobase_2.62.0       ExperimentHub_2.10.0
[10] AnnotationHub_3.10.0 BiocFileCache_2.10.1 dbplyr_2.4.0        
[13] BiocGenerics_0.48.1  msigdb_1.10.0        BiocManager_1.30.22 
[16] workflowr_1.7.1     

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0              dplyr_1.1.4                  
 [3] blob_1.2.4                    filelock_1.0.3               
 [5] Biostrings_2.70.1             bitops_1.0-7                 
 [7] fastmap_1.1.1                 RCurl_1.98-1.13              
 [9] promises_1.2.1                digest_0.6.33                
[11] mime_0.12                     lifecycle_1.0.4              
[13] ellipsis_0.3.2                processx_3.8.3               
[15] KEGGREST_1.42.0               interactiveDisplayBase_1.40.0
[17] RSQLite_2.3.4                 magrittr_2.0.3               
[19] compiler_4.3.2                rlang_1.1.2                  
[21] sass_0.4.8                    tools_4.3.2                  
[23] utf8_1.2.4                    yaml_2.3.8                   
[25] knitr_1.45                    bit_4.0.5                    
[27] curl_5.2.0                    withr_2.5.2                  
[29] purrr_1.0.2                   fansi_1.0.6                  
[31] git2r_0.33.0                  xtable_1.8-4                 
[33] cli_3.6.2                     rmarkdown_2.25               
[35] crayon_1.5.2                  generics_0.1.3               
[37] rstudioapi_0.15.0             httr_1.4.7                   
[39] DBI_1.1.3                     cachem_1.0.8                 
[41] stringr_1.5.1                 zlibbioc_1.48.0              
[43] XVector_0.42.0                vctrs_0.6.5                  
[45] jsonlite_1.8.8                callr_3.7.3                  
[47] bit64_4.0.5                   jquerylib_0.1.4              
[49] glue_1.6.2                    ps_1.7.5                     
[51] stringi_1.8.3                 BiocVersion_3.18.1           
[53] later_1.3.2                   GenomeInfoDb_1.38.1          
[55] tibble_3.2.1                  pillar_1.9.0                 
[57] rappdirs_0.3.3                htmltools_0.5.7              
[59] GenomeInfoDbData_1.2.11       R6_2.5.1                     
[61] rprojroot_2.0.4               evaluate_0.23                
[63] shiny_1.8.0                   png_0.1-8                    
[65] memoise_2.0.1                 httpuv_1.6.13                
[67] bslib_0.6.1                   Rcpp_1.0.11                  
[69] whisker_0.4.1                 xfun_0.41                    
[71] fs_1.6.3                      getPass_0.2-4                
[73] pkgconfig_2.0.3              </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
