<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2023-06-28" />

<title>Position weight matrix</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Position weight matrix</h1>
<h4 class="date">2023-06-28</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-06-28
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreed944430aec849d37698f6ceda035658ca86d43fctargetblankd944430a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/d944430aec849d37698f6ceda035658ca86d43fc" target="_blank">d944430</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreed944430aec849d37698f6ceda035658ca86d43fctargetblankd944430a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/d944430aec849d37698f6ceda035658ca86d43fc" target="_blank">d944430</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.1.2/
    Ignored:    r_packages_4.2.0/
    Ignored:    r_packages_4.2.2/
    Ignored:    r_packages_4.3.0/

Untracked files:
    Untracked:  analysis/cell_ranger.Rmd
    Untracked:  analysis/tss_xgboost.Rmd
    Untracked:  code/multiz100way/
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz
    Untracked:  data/HG00702_SH089_CHSTrio.chr1.vcf.gz.tbi
    Untracked:  data/ncrna_NONCODE[v3.0].fasta.tar.gz
    Untracked:  data/ncrna_noncode_v3.fa
    Untracked:  data/netmhciipan.out.gz
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/graph.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/pwm.Rmd</code>) and HTML
(<code>docs/pwm.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/d944430aec849d37698f6ceda035658ca86d43fc/analysis/pwm.Rmd" target="_blank">d944430</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-28
</td>
<td>
Position weight matrix
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The process of transcription, is influenced by the interaction of
proteins called <a
href="https://en.wikipedia.org/wiki/Transcription_factor">transcription
factors</a> (TFs) that bind to specific sites called Transcription
Factor Binding Sites (TFBSs), which are proximal or distal to a
transcription starting site. TFs generally have distinct binding
preferences towards specific TFBSs, however TFs can tolerate variations
in the target TFBS. Thus to model a TFBS, the nucleotides are weighted
accordingly, to the tolerance of the TF. One common way to represent
this is by using a <a
href="https://en.wikipedia.org/wiki/Position_weight_matrix">Position
Weight Matrix</a> (PWM), also called position-specific weight matrix
(PSWM) or position-specific scoring matrix (PSSM), which is a commonly
used representation of motifs (in our case TFBS) in biological
sequences.</p>
<p>How do we find TFBSs? DNA sequences that interact with TFs can be
experimentally determined from <a
href="https://en.wikipedia.org/wiki/Systematic_evolution_of_ligands_by_exponential_enrichment">SELEX</a>
experiments. Since this process involves synthesis of a large number of
randomly generated oligonucleotides, DNA sequences that interact with
TFs can be determined, as well as the tolerance at specific sites. From
SELEX experiments, a position frequency matrix (PFM) can be constructed
by recording the position-dependent frequency of each nucleotide in the
DNA sequence that interacted with the TF. Here’s an example of a PFM as
shown in this review “<a
href="https://www.nature.com/articles/nrg1315">Applied bioinformatics
for the identification of regulatory elements</a>” (sorry paywall!):</p>
<table>
<colgroup>
<col width="12%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">nuc</th>
<th align="left">1</th>
<th align="left">2</th>
<th align="left">3</th>
<th align="left">4</th>
<th align="left">5</th>
<th align="left">6</th>
<th align="left">7</th>
<th align="left">8</th>
<th align="left">9</th>
<th align="left">10</th>
<th align="left">11</th>
<th align="left">12</th>
<th align="left">13</th>
<th align="left">14</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">A</td>
<td align="left">0</td>
<td align="left">4</td>
<td align="left">4</td>
<td align="left">0</td>
<td align="left">3</td>
<td align="left">7</td>
<td align="left">4</td>
<td align="left">3</td>
<td align="left">5</td>
<td align="left">4</td>
<td align="left">2</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">4</td>
</tr>
<tr class="even">
<td align="left">C</td>
<td align="left">3</td>
<td align="left">0</td>
<td align="left">4</td>
<td align="left">8</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">3</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">2</td>
<td align="left">2</td>
</tr>
<tr class="odd">
<td align="left">G</td>
<td align="left">2</td>
<td align="left">3</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">6</td>
<td align="left">8</td>
<td align="left">5</td>
<td align="left">0</td>
</tr>
<tr class="even">
<td align="left">T</td>
<td align="left">3</td>
<td align="left">1</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">5</td>
<td align="left">1</td>
<td align="left">4</td>
<td align="left">2</td>
<td align="left">2</td>
<td align="left">4</td>
<td align="left">0</td>
<td align="left">0</td>
<td align="left">1</td>
<td align="left">0</td>
</tr>
</tbody>
</table>
<p><em>The construction of this PFM was done by collecting
experimentally validated binding sites from 8 published studies for
MEF2.</em></p>
<p>To convert a PFM to the corresponding PWM, the frequencies are
converted to normalised frequency values on a log-scale. To perform this
conversion we can use these formulae from the review paper:</p>
<p><span class="math display">\[ W_{b,i} = log_{2}\frac{p(b,i)}{p(b)}
\]</span></p>
<p>where <span class="math inline">\(W_{b,i}\)</span> = PWM value of
base <em>b</em> in position <em>i</em>, <span
class="math inline">\(p(b)\)</span> = background probability of base
<em>b</em> and <span class="math inline">\(p(b,i)\)</span>:</p>
<p><span class="math display">\[ p(b,i) = \frac{f_{b,i} + s(b)}{N + \sum
s(b&#39;)} \]</span></p>
<p>where <span class="math inline">\(b&#39; \in \{A, C, G, T\}\)</span>;
<span class="math inline">\(f_{b,i}\)</span> = counts of base <em>b</em>
in position <em>i</em>; <span class="math inline">\(N\)</span> = number
of sites; <span class="math inline">\(p(b,i)\)</span> = corrected
probability of base <em>b</em> in position <em>i</em> and <span
class="math inline">\(s(b)\)</span> = pseudocount function.</p>
<p>The pseudocount is a sample correction that is added when assessing
the probability to correct for small samples sizes and this calculation
varies widely between applications. One approach is to take the square
root of the number of sites that contribute to the model, which was
apparently the approach used in the paper. However, when I used this as
the pseudocount function, I could not replicate the numbers in the table
of the review (Box 1, d). In fact, if you examine the PWM in the paper
(Box 1, d), there’s several typos. For example, in position 1, the
frequency of C’s and T’s are identical, however the PWM values are
different. Also the PWM for a frequency of 1 is different in position
13, T and all other positions with a frequency of 1. Most other values
seem to be consistent though. So in order to find out the pseudocount
function, I substituted some of the (consistent) PWM values back into
the equation.</p>
<p>Firstly recall the conversion between the logarithmic form to the
exponential form:</p>
<p><span class="math display">\[ y = log_{a}x \leftrightarrow a^{y} = x.
\]</span></p>
<p>Therefore:</p>
<p><span class="math display">\[ W_{b,i} = log_{2}\frac{p(b,i)}{p(b)}
\leftrightarrow 2^{W_{b,i}} = \frac{p(b,i)}{p(b)} \]</span></p>
<p>and substituting <span class="math inline">\(p(b,i)\)</span> in:</p>
<p><span class="math display">\[ 2^{W_{b,i}} \cdot p(b) = \frac{f_{b,i}
+ s(b)}{N + \sum s(b&#39;)} \]</span></p>
<p>The PWM value for a frequency of 0 seems to be consistent, so let’s
take <span class="math inline">\(W_{A,1}\)</span> and substitute it into
the equation:</p>
<p><span class="math display">\[ 2^{-1.93} \cdot p(b) = \frac{0 +
s_{A}}{8 + s_{A} + s_{C} + s_{G} + s_{T}}. \]</span></p>
<p>The PWM value for a frequency of 3 seems to be 0.45, so let’s take
<span class="math inline">\(W_{C,1}\)</span> and substitute it into the
equation:</p>
<p><span class="math display">\[ 2^{0.45} \cdot p(b) = \frac{3 +
s_{C}}{8 + s_{A} + s_{C} + s_{G} + s_{T}}. \]</span></p>
<p>We now have two simultaneous equations that we can divide together,
i.e. dividing <span class="math inline">\(W\_{A,1}\)</span> by <span
class="math inline">\(W_{C,1}\)</span> (the denominators cancel each
other out), to work out the pseudocount:</p>
<p><span class="math display">\[ \frac{2^{-1.93} \cdot p(b)}{2^{0.45}
\cdot p(b)} = \frac{s_{A}}{3 + s_{C}}. \]</span></p>
<p>The background probabilities cancel each other out and since the
pseudocounts should be the same for the different nucleotides, we can
refer to them as an <em>s</em>:</p>
<p><span class="math display">\[ \frac{0.2624292}{1.36604} = \frac{s}{3
+ s}. \]</span></p>
<p><span class="math display">\[ 0.1921094 = \frac{s}{3 + s}.
\]</span></p>
<p>Invert the two sides:</p>
<p><span class="math display">\[ 5.205367 = \frac{3 + s}{s} =
\frac{3}{s} + \frac{s}{s}. \]</span></p>
<p>Solving <em>s</em>, we get:</p>
<p><span class="math display">\[ 5.205367 = \frac{3}{s} + 1
\]</span></p>
<p><span class="math display">\[ 4.205367 = \frac{3}{s} \]</span></p>
<p><span class="math display">\[ s = \frac{3}{4.205367} = 0.7133741
\]</span></p>
<p>Perhaps I missed it, but it wasn’t pointed out exactly how <span
class="math inline">\(p(b)\)</span> or the background probability of
base <em>b</em> is defined. Since we worked out <em>s</em>, we can
substitute it into the equation for <span
class="math inline">\(W_{A,1}\)</span> and work out <span
class="math inline">\(p(b)\)</span>:</p>
<p><span class="math display">\[ 2^{-1.93} \cdot p(b) = \frac{0 +
0.71}{8 + 0.71 + 0.71 + 0.71 + 0.71}. \]</span></p>
<p><span class="math display">\[ p(b) = \frac{0.06572758}{2^{-1.93}} =
0.2504584. \]</span></p>
<p>So the background probability of base <em>b</em> is simply 0.25,
i.e. base b divided by the total number of bases. But where did this
0.71 pseudocount come from? If we take the square root of the number of
sites that contribute to the model, the square root of 8 is 2.828427.
Perhaps the pseudocount needs to be scaled by base <em>b</em>, since the
square root of 8 multiply by 1/4 is 0.7071068.</p>
<p>Since some of the values in the PWM of the review paper is incorrect,
let’s calculate it again using the formulae above (since we now know the
pseudocount function and the background probability) using R:</p>
<pre class="r"><code>calc_pwm &lt;- function(freq, total, bg=0.25){
  p &lt;- (freq + (sqrt(total) * 1/4)) / (total + (4 * (sqrt(total) * 1/4)))
  log2(p/bg)
}

# work out all possible PWM values
map_dbl(0:8, calc_pwm, total = 8)</code></pre>
<pre><code>[1] -1.9367518 -0.6651985  0.0000000  0.4535419  0.7980888  1.0760078  1.3089388
[8]  1.5094376  1.6854416</code></pre>
<p>Now let’s calculate the PWM by first defining the matrix.</p>
<pre class="r"><code># define the frequencies of nucleotides
A &lt;- c(0, 4, 4, 0, 3, 7, 4, 3, 5, 4, 2, 0, 0, 4)
C &lt;- c(3, 0, 4, 8, 0, 0, 0, 3, 0, 0, 0, 0, 2, 4)
G &lt;- c(2, 3, 0, 0, 0, 0, 0, 0, 1, 0, 6, 8, 5, 0)
T &lt;- c(3, 1, 0, 0, 5, 1, 4, 2, 2, 4, 0, 0, 1, 0)

M &lt;- matrix(
  data = c(A,C,G,T),
  nrow = 4,
  byrow = TRUE,
  dimnames = list(c(&#39;A&#39;,&#39;C&#39;,&#39;G&#39;,&#39;T&#39;))
)

M</code></pre>
<pre><code>  [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
A    0    4    4    0    3    7    4    3    5     4     2     0     0     4
C    3    0    4    8    0    0    0    3    0     0     0     0     2     4
G    2    3    0    0    0    0    0    0    1     0     6     8     5     0
T    3    1    0    0    5    1    4    2    2     4     0     0     1     0</code></pre>
<p>Calculate the PWM using our function.</p>
<pre class="r"><code>pwm &lt;- calc_pwm(M, 8)
pwm</code></pre>
<pre><code>        [,1]       [,2]       [,3]      [,4]       [,5]       [,6]       [,7]
A -1.9367518  0.7980888  0.7980888 -1.936752  0.4535419  1.5094376  0.7980888
C  0.4535419 -1.9367518  0.7980888  1.685442 -1.9367518 -1.9367518 -1.9367518
G  0.0000000  0.4535419 -1.9367518 -1.936752 -1.9367518 -1.9367518 -1.9367518
T  0.4535419 -0.6651985 -1.9367518 -1.936752  1.0760078 -0.6651985  0.7980888
        [,8]       [,9]      [,10]     [,11]     [,12]      [,13]      [,14]
A  0.4535419  1.0760078  0.7980888  0.000000 -1.936752 -1.9367518  0.7980888
C  0.4535419 -1.9367518 -1.9367518 -1.936752 -1.936752  0.0000000  0.7980888
G -1.9367518 -0.6651985 -1.9367518  1.308939  1.685442  1.0760078 -1.9367518
T  0.0000000  0.0000000  0.7980888 -1.936752 -1.936752 -0.6651985 -1.9367518</code></pre>
<p>Now that we have the PWM, we can generate a quantitative score for
any DNA sequence (of the same length) by summing the values that
correspond to the observed nucleotides at each position. Here we will
use the same sequence as shown in the review.</p>
<pre class="r"><code>seq &lt;- &#39;TTACATAAGTAGTC&#39;
seq &lt;- unlist(strsplit(x = seq, split=&#39;&#39;))
seq</code></pre>
<pre><code> [1] &quot;T&quot; &quot;T&quot; &quot;A&quot; &quot;C&quot; &quot;A&quot; &quot;T&quot; &quot;A&quot; &quot;A&quot; &quot;G&quot; &quot;T&quot; &quot;A&quot; &quot;G&quot; &quot;T&quot; &quot;C&quot;</code></pre>
<p>A quantitative score for any DNA sequence can be generated by summing
the values that correspond to the observed nucleotide at each
position.</p>
<pre class="r"><code>seq_score &lt;- map_dbl(1:length(seq), function(x) pwm[seq[x], x])
# slightly different to the review due to rounding
sum(seq_score)</code></pre>
<pre><code>[1] 5.26307</code></pre>
<p>The maximum score can be calculated by summing all the maximum values
of each column.</p>
<pre class="r"><code>sum(apply(pwm, 2, max))</code></pre>
<pre><code>[1] 14.31481</code></pre>
<p>Lastly, we will generate a sequence logo from a PFM. We need the
<strong>seqLogo</strong> package, so install it if you have not
already.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

if (!require(&quot;seqLogo&quot;, quietly = TRUE))
  BiocManager::install(&quot;seqLogo&quot;)</code></pre>
<p>Create a data frame.</p>
<pre class="r"><code>library(seqLogo)

A &lt;- c(0, 4, 4, 0, 3, 7, 4, 3, 5, 4, 2, 0, 0, 4)
C &lt;- c(3, 0, 4, 8, 0, 0, 0, 3, 0, 0, 0, 0, 2, 4)
G &lt;- c(2, 3, 0, 0, 0, 0, 0, 0, 1, 0, 6, 8, 5, 0)
T &lt;- c(3, 1, 0, 0, 5, 1, 4, 2, 2, 4, 0, 0, 1, 0)

my_df &lt;- data.frame(A, C, G, T)
my_df</code></pre>
<pre><code>   A C G T
1  0 3 2 3
2  4 0 3 1
3  4 4 0 0
4  0 8 0 0
5  3 0 0 5
6  7 0 0 1
7  4 0 0 4
8  3 3 0 2
9  5 0 1 2
10 4 0 0 4
11 2 0 6 0
12 0 0 8 0
13 0 2 5 1
14 4 4 0 0</code></pre>
<p>First we divide the frequency by the row sum to get the proportions.
We then construct a PWM and then plot a sequence logo. Note that this
PWM is calculated in a different manner as our example above but they
both show the same information.</p>
<pre class="r"><code>mef2 &lt;- apply(my_df, 1, function(x) x / sum(x))
mef2 &lt;- makePWM(mef2)
seqLogo(mef2)</code></pre>
<p><img src="figure/pwm.Rmd/seq_logo-1.png" width="672" style="display: block; margin: auto;" /></p>
<div id="conclusions" class="section level3">
<h3>Conclusions</h3>
<p>With respect to transcription factors (TFs), a position weight matrix
(PWM) can be generated from a position frequency matrix (PFM), which is
a collection of experimentally validated binding sites. Using this PWM,
any given sequence can be quantitatively scored against the motif model.
The PWM models appropriately the tolerance of TFs to binding sites and
one can use sequence logos to visualise PFMs.</p>
<p>I’d like to thank my colleague (if he ever comes across this page)
for his help.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] seqLogo_1.66.0      BiocManager_1.30.20 lubridate_1.9.2    
 [4] forcats_1.0.0       stringr_1.5.0       dplyr_1.1.2        
 [7] purrr_1.0.1         readr_2.1.4         tidyr_1.3.0        
[10] tibble_3.2.1        ggplot2_3.4.2       tidyverse_2.0.0    
[13] workflowr_1.7.0    

loaded via a namespace (and not attached):
 [1] sass_0.4.5       utf8_1.2.3       generics_0.1.3   stringi_1.7.12  
 [5] hms_1.1.3        digest_0.6.31    magrittr_2.0.3   timechange_0.2.0
 [9] evaluate_0.20    fastmap_1.1.1    rprojroot_2.0.3  jsonlite_1.8.5  
[13] processx_3.8.1   whisker_0.4.1    ps_1.7.5         promises_1.2.0.1
[17] httr_1.4.5       fansi_1.0.4      scales_1.2.1     jquerylib_0.1.4 
[21] cli_3.6.1        rlang_1.1.0      munsell_0.5.0    withr_2.5.0     
[25] cachem_1.0.7     yaml_2.3.7       tools_4.3.0      tzdb_0.3.0      
[29] colorspace_2.1-0 httpuv_1.6.9     vctrs_0.6.2      R6_2.5.1        
[33] stats4_4.3.0     lifecycle_1.0.3  git2r_0.32.0     fs_1.6.2        
[37] pkgconfig_2.0.3  callr_3.7.3      pillar_1.9.0     bslib_0.4.2     
[41] later_1.3.0      gtable_0.3.3     glue_1.6.2       Rcpp_1.0.10     
[45] highr_0.10       xfun_0.39        tidyselect_1.2.0 rstudioapi_0.14 
[49] knitr_1.42       htmltools_0.5.5  rmarkdown_2.21   compiler_4.3.0  
[53] getPass_0.2-2   </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
