<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2020-12-01" />

<title>Read GTF file into R</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/anchor-sections-1.0/anchor-sections.css" rel="stylesheet" />
<script src="site_libs/anchor-sections-1.0/anchor-sections.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Read GTF file into R</h1>
<h4 class="date">2020-12-01</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-12-01
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeacb173357057c528ee677e301fd5cc52cf2b8827targetblankacb1733a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/davetang/muse/tree/acb173357057c528ee677e301fd5cc52cf2b8827" target="_blank">acb1733</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeacb173357057c528ee677e301fd5cc52cf2b8827targetblankacb1733a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/davetang/muse/tree/acb173357057c528ee677e301fd5cc52cf2b8827" target="_blank">acb1733</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.Rhistory

Untracked files:
    Untracked:  analysis/linear_regression.Rmd
    Untracked:  data/gencode.v36.annotation.gtf.gz

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/read_gtf.Rmd</code>) and HTML (<code>docs/read_gtf.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/acb173357057c528ee677e301fd5cc52cf2b8827/analysis/read_gtf.Rmd" target="_blank">acb1733</a>
</td>
<td>
davetang
</td>
<td>
2020-12-01
</td>
<td>
Read GTF files into R
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4">Gene Transfer Format</a> (GTF) is a refinement of the <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format3">General Feature Format</a> (GFF). A GFF file has nine columns:</p>
<table>
<colgroup>
<col width="41%" />
<col width="58%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Column</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">seqname</td>
<td align="left">The name of the sequence; must be a chromosome or scaffold</td>
</tr>
<tr class="even">
<td align="left">source</td>
<td align="left">The program that generated this feature</td>
</tr>
<tr class="odd">
<td align="left">feature</td>
<td align="left">The name of this type of feature, e.g. “CDS”, “start_codon”, “stop_codon”, and “exon”</td>
</tr>
<tr class="even">
<td align="left">start</td>
<td align="left">The starting position of the feature in the sequence; the first base is numbered 1</td>
</tr>
<tr class="odd">
<td align="left">end</td>
<td align="left">The ending position of the feature (inclusive)</td>
</tr>
<tr class="even">
<td align="left">score</td>
<td align="left">A score between 0 and 1000</td>
</tr>
<tr class="odd">
<td align="left">strand</td>
<td align="left">Valid entries include “+”, “-”, or “.”</td>
</tr>
<tr class="even">
<td align="left">frame</td>
<td align="left">If the feature is a coding exon, frame should be a number between 0-2 that represents the reading frame of the first base. If the feature is not a coding exon, the value should be “.”</td>
</tr>
<tr class="odd">
<td align="left">group</td>
<td align="left">All lines with the same group are linked together into a single item</td>
</tr>
</tbody>
</table>
<p>The first eight fields in a GTF file are the same as GFF but the group field has been expanded into a list of attributes, where each attribute consists of a type/value pair. Attributes must end in a semi-colon and be separated from any following attribute by exactly one space. The attribute list must begin with the two mandatory attributes:</p>
<ol style="list-style-type: decimal">
<li><strong>gene_id value</strong> – A globally unique identifier for the genomic source of the sequence.</li>
<li><strong>transcript_id value</strong> – A globally unique identifier for the predicted transcript.</li>
</ol>
<p>To get started, I will use the latest GENCODE GTF file.</p>
<pre class="r"><code>gencode_ver &lt;- 36
my_url &lt;- paste0(&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_&quot;, gencode_ver, &quot;/gencode.v&quot;, gencode_ver, &quot;.annotation.gtf.gz&quot;)
my_gtf &lt;- basename(my_url)

if (!file.exists(paste0(&quot;data/&quot;, my_gtf))){
  download.file(url = my_url, destfile = paste0(&quot;data/&quot;, my_gtf))
}</code></pre>
<p>We will use <a href="https://www.bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a> to import the GTF file into R.</p>
<pre class="r"><code>library(rtracklayer)</code></pre>
<pre><code>Loading required package: GenomicRanges</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: BiocGenerics</code></pre>
<pre><code>Loading required package: parallel</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:parallel&#39;:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>Loading required package: S4Vectors</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    expand.grid</code></pre>
<pre><code>Loading required package: IRanges</code></pre>
<pre><code>Loading required package: GenomeInfoDb</code></pre>
<pre class="r"><code>my_obj &lt;- import(paste0(&quot;data/&quot;, my_gtf))
class(my_obj)</code></pre>
<pre><code>[1] &quot;GRanges&quot;
attr(,&quot;package&quot;)
[1] &quot;GenomicRanges&quot;</code></pre>
<p>The <a href="https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a> package serves as the foundation for representing genomic locations within the Bioconductor project. The GRanges class represents a collection of genomic features that each have a single start and end location on the genome. This includes features such as contiguous binding sites, transcripts, and exons.</p>
<pre class="r"><code>my_obj</code></pre>
<pre><code>GRanges object with 3014855 ranges and 21 metadata columns:
            seqnames      ranges strand |   source       type     score
               &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt;
        [1]     chr1 11869-14409      + |   HAVANA gene              NA
        [2]     chr1 11869-14409      + |   HAVANA transcript        NA
        [3]     chr1 11869-12227      + |   HAVANA exon              NA
        [4]     chr1 12613-12721      + |   HAVANA exon              NA
        [5]     chr1 13221-14409      + |   HAVANA exon              NA
        ...      ...         ...    ... .      ...        ...       ...
  [3014851]     chrM 15888-15953      + |  ENSEMBL transcript        NA
  [3014852]     chrM 15888-15953      + |  ENSEMBL exon              NA
  [3014853]     chrM 15956-16023      - |  ENSEMBL gene              NA
  [3014854]     chrM 15956-16023      - |  ENSEMBL transcript        NA
  [3014855]     chrM 15956-16023      - |  ENSEMBL exon              NA
                phase           gene_id                          gene_type
            &lt;integer&gt;       &lt;character&gt;                        &lt;character&gt;
        [1]      &lt;NA&gt; ENSG00000223972.5 transcribed_unprocessed_pseudogene
        [2]      &lt;NA&gt; ENSG00000223972.5 transcribed_unprocessed_pseudogene
        [3]      &lt;NA&gt; ENSG00000223972.5 transcribed_unprocessed_pseudogene
        [4]      &lt;NA&gt; ENSG00000223972.5 transcribed_unprocessed_pseudogene
        [5]      &lt;NA&gt; ENSG00000223972.5 transcribed_unprocessed_pseudogene
        ...       ...               ...                                ...
  [3014851]      &lt;NA&gt; ENSG00000210195.2                            Mt_tRNA
  [3014852]      &lt;NA&gt; ENSG00000210195.2                            Mt_tRNA
  [3014853]      &lt;NA&gt; ENSG00000210196.2                            Mt_tRNA
  [3014854]      &lt;NA&gt; ENSG00000210196.2                            Mt_tRNA
  [3014855]      &lt;NA&gt; ENSG00000210196.2                            Mt_tRNA
              gene_name       level     hgnc_id          havana_gene
            &lt;character&gt; &lt;character&gt; &lt;character&gt;          &lt;character&gt;
        [1]     DDX11L1           2  HGNC:37102 OTTHUMG00000000961.2
        [2]     DDX11L1           2  HGNC:37102 OTTHUMG00000000961.2
        [3]     DDX11L1           2  HGNC:37102 OTTHUMG00000000961.2
        [4]     DDX11L1           2  HGNC:37102 OTTHUMG00000000961.2
        [5]     DDX11L1           2  HGNC:37102 OTTHUMG00000000961.2
        ...         ...         ...         ...                  ...
  [3014851]       MT-TT           3   HGNC:7499                 &lt;NA&gt;
  [3014852]       MT-TT           3   HGNC:7499                 &lt;NA&gt;
  [3014853]       MT-TP           3   HGNC:7494                 &lt;NA&gt;
  [3014854]       MT-TP           3   HGNC:7494                 &lt;NA&gt;
  [3014855]       MT-TP           3   HGNC:7494                 &lt;NA&gt;
                transcript_id      transcript_type transcript_name
                  &lt;character&gt;          &lt;character&gt;     &lt;character&gt;
        [1]              &lt;NA&gt;                 &lt;NA&gt;            &lt;NA&gt;
        [2] ENST00000456328.2 processed_transcript     DDX11L1-202
        [3] ENST00000456328.2 processed_transcript     DDX11L1-202
        [4] ENST00000456328.2 processed_transcript     DDX11L1-202
        [5] ENST00000456328.2 processed_transcript     DDX11L1-202
        ...               ...                  ...             ...
  [3014851] ENST00000387460.2              Mt_tRNA       MT-TT-201
  [3014852] ENST00000387460.2              Mt_tRNA       MT-TT-201
  [3014853]              &lt;NA&gt;                 &lt;NA&gt;            &lt;NA&gt;
  [3014854] ENST00000387461.2              Mt_tRNA       MT-TP-201
  [3014855] ENST00000387461.2              Mt_tRNA       MT-TP-201
            transcript_support_level         tag    havana_transcript
                         &lt;character&gt; &lt;character&gt;          &lt;character&gt;
        [1]                     &lt;NA&gt;        &lt;NA&gt;                 &lt;NA&gt;
        [2]                        1       basic OTTHUMT00000362751.1
        [3]                        1       basic OTTHUMT00000362751.1
        [4]                        1       basic OTTHUMT00000362751.1
        [5]                        1       basic OTTHUMT00000362751.1
        ...                      ...         ...                  ...
  [3014851]                       NA       basic                 &lt;NA&gt;
  [3014852]                       NA       basic                 &lt;NA&gt;
  [3014853]                     &lt;NA&gt;        &lt;NA&gt;                 &lt;NA&gt;
  [3014854]                       NA       basic                 &lt;NA&gt;
  [3014855]                       NA       basic                 &lt;NA&gt;
            exon_number           exon_id         ont  protein_id      ccdsid
            &lt;character&gt;       &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
        [1]        &lt;NA&gt;              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
        [2]        &lt;NA&gt;              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
        [3]           1 ENSE00002234944.1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
        [4]           2 ENSE00003582793.1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
        [5]           3 ENSE00002312635.1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
        ...         ...               ...         ...         ...         ...
  [3014851]        &lt;NA&gt;              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [3014852]           1 ENSE00001544475.2        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [3014853]        &lt;NA&gt;              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [3014854]        &lt;NA&gt;              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [3014855]           1 ENSE00001544473.2        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
<p>We can use the awesome <code>plyranges</code> package by Stuart Lee to find out the number of transcripts on each chromosome.</p>
<pre class="r"><code>library(plyranges)</code></pre>
<pre><code>
Attaching package: &#39;plyranges&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:IRanges&#39;:

    slice</code></pre>
<pre><code>The following object is masked from &#39;package:stats&#39;:

    filter</code></pre>
<pre class="r"><code>my_obj %&gt;%
  group_by(seqnames) %&gt;%
  summarise(total = n()) %&gt;%
  as.data.frame()</code></pre>
<pre><code>   seqnames  total
1      chr1 276557
2      chr2 225529
3      chr3 191057
4      chr4 126025
5      chr5 134857
6      chr6 137896
7      chr7 143222
8      chr8 114067
9      chr9 107691
10    chr10 116513
11    chr11 176975
12    chr12 172270
13    chr13  48692
14    chr14 107092
15    chr15 111227
16    chr16 139150
17    chr17 181102
18    chr18  53969
19    chr19 176133
20    chr20  68882
21    chr21  33899
22    chr22  64343
23     chrX  97807
24     chrY   9757
25     chrM    143</code></pre>
<p>Fetch mitochondrial transcripts.</p>
<pre class="r"><code>my_obj %&gt;%
  filter(seqnames == &quot;chrM&quot;) %&gt;%
  head()</code></pre>
<pre><code>GRanges object with 6 ranges and 21 metadata columns:
      seqnames    ranges strand |   source       type     score     phase
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
  [1]     chrM   577-647      + |  ENSEMBL gene              NA      &lt;NA&gt;
  [2]     chrM   577-647      + |  ENSEMBL transcript        NA      &lt;NA&gt;
  [3]     chrM   577-647      + |  ENSEMBL exon              NA      &lt;NA&gt;
  [4]     chrM  648-1601      + |  ENSEMBL gene              NA      &lt;NA&gt;
  [5]     chrM  648-1601      + |  ENSEMBL transcript        NA      &lt;NA&gt;
  [6]     chrM  648-1601      + |  ENSEMBL exon              NA      &lt;NA&gt;
                gene_id   gene_type   gene_name       level     hgnc_id
            &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
  [1] ENSG00000210049.1     Mt_tRNA       MT-TF           3   HGNC:7481
  [2] ENSG00000210049.1     Mt_tRNA       MT-TF           3   HGNC:7481
  [3] ENSG00000210049.1     Mt_tRNA       MT-TF           3   HGNC:7481
  [4] ENSG00000211459.2     Mt_rRNA     MT-RNR1           3   HGNC:7470
  [5] ENSG00000211459.2     Mt_rRNA     MT-RNR1           3   HGNC:7470
  [6] ENSG00000211459.2     Mt_rRNA     MT-RNR1           3   HGNC:7470
      havana_gene     transcript_id transcript_type transcript_name
      &lt;character&gt;       &lt;character&gt;     &lt;character&gt;     &lt;character&gt;
  [1]        &lt;NA&gt;              &lt;NA&gt;            &lt;NA&gt;            &lt;NA&gt;
  [2]        &lt;NA&gt; ENST00000387314.1         Mt_tRNA       MT-TF-201
  [3]        &lt;NA&gt; ENST00000387314.1         Mt_tRNA       MT-TF-201
  [4]        &lt;NA&gt;              &lt;NA&gt;            &lt;NA&gt;            &lt;NA&gt;
  [5]        &lt;NA&gt; ENST00000389680.2         Mt_rRNA     MT-RNR1-201
  [6]        &lt;NA&gt; ENST00000389680.2         Mt_rRNA     MT-RNR1-201
      transcript_support_level         tag havana_transcript exon_number
                   &lt;character&gt; &lt;character&gt;       &lt;character&gt; &lt;character&gt;
  [1]                     &lt;NA&gt;        &lt;NA&gt;              &lt;NA&gt;        &lt;NA&gt;
  [2]                       NA       basic              &lt;NA&gt;        &lt;NA&gt;
  [3]                       NA       basic              &lt;NA&gt;           1
  [4]                     &lt;NA&gt;        &lt;NA&gt;              &lt;NA&gt;        &lt;NA&gt;
  [5]                       NA       basic              &lt;NA&gt;        &lt;NA&gt;
  [6]                       NA       basic              &lt;NA&gt;           1
                exon_id         ont  protein_id      ccdsid
            &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
  [1]              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [2]              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [3] ENSE00001544501.1        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [4]              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [5]              &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  [6] ENSE00001544499.2        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Summarise biotypes and plot.</p>
<pre class="r"><code>my_obj %&gt;%
  group_by(transcript_type) %&gt;%
  summarise(number = n()) %&gt;%
  as.data.frame() -&gt; my_biotypes

my_biotypes %&gt;%
  dplyr::filter(!is.na(transcript_type)) -&gt; my_biotypes

my_biotypes %&gt;%
  arrange(desc(number)) %&gt;%
  dplyr::pull(transcript_type) -&gt; my_order

my_biotypes$transcript_type &lt;- factor(my_biotypes$transcript_type, levels = my_order)

library(ggplot2)

ggplot(my_biotypes, aes(transcript_type, number)) +
  geom_col() +
  theme_bw() +
  scale_y_log10() +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))</code></pre>
<p><img src="figure/read_gtf.Rmd/transcript_biotypes-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Visualise transcripts near chr17:7661779-7687538.</p>
<pre class="r"><code>library(Gviz)</code></pre>
<pre><code>Loading required package: grid</code></pre>
<pre class="r"><code>library(GenomicFeatures)</code></pre>
<pre><code>Loading required package: AnnotationDbi</code></pre>
<pre><code>Loading required package: Biobase</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>
Attaching package: &#39;AnnotationDbi&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:plyranges&#39;:

    select</code></pre>
<pre class="r"><code>my_txdb &lt;- makeTxDbFromGFF(paste0(&quot;data/&quot;, my_gtf))</code></pre>
<pre><code>Import genomic features from the file as a GRanges object ...</code></pre>
<pre><code>OK</code></pre>
<pre><code>Prepare the &#39;metadata&#39; data frame ... OK
Make the TxDb object ... </code></pre>
<pre><code>Warning in .get_cds_IDX(mcols0$type, mcols0$phase): The &quot;phase&quot; metadata column contains non-NA values for features of type
  stop_codon. This information was ignored.</code></pre>
<pre><code>OK</code></pre>
<pre class="r"><code>my_start &lt;- 7661779 - 5000
my_end &lt;- 7687538 + 5000
geneTrack &lt;- GeneRegionTrack(my_txdb, chromosome=&quot;chr17&quot;, from=my_start, to=my_end)

plotTracks(geneTrack, chromosome=&quot;chr17&quot;, from=my_start, to=my_end, showId=TRUE)</code></pre>
<p><img src="figure/read_gtf.Rmd/gviz-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS  10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8

attached base packages:
 [1] grid      parallel  stats4    stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] GenomicFeatures_1.40.1 AnnotationDbi_1.50.3   Biobase_2.48.0        
 [4] Gviz_1.32.0            ggplot2_3.3.2          plyranges_1.8.0       
 [7] rtracklayer_1.48.0     GenomicRanges_1.40.0   GenomeInfoDb_1.24.2   
[10] IRanges_2.22.2         S4Vectors_0.26.1       BiocGenerics_0.34.0   
[13] workflowr_1.6.2       

loaded via a namespace (and not attached):
  [1] colorspace_2.0-0            ellipsis_0.3.1             
  [3] rprojroot_2.0.2             biovizBase_1.36.0          
  [5] htmlTable_2.1.0             XVector_0.28.0             
  [7] base64enc_0.1-3             fs_1.5.0                   
  [9] dichromat_2.0-0             rstudioapi_0.13            
 [11] farver_2.0.3                bit64_4.0.5                
 [13] xml2_1.3.2                  splines_4.0.2              
 [15] knitr_1.30                  Formula_1.2-4              
 [17] Rsamtools_2.4.0             cluster_2.1.0              
 [19] dbplyr_2.0.0                png_0.1-7                  
 [21] compiler_4.0.2              httr_1.4.2                 
 [23] backports_1.2.0             assertthat_0.2.1           
 [25] Matrix_1.2-18               lazyeval_0.2.2             
 [27] later_1.1.0.1               htmltools_0.5.0            
 [29] prettyunits_1.1.1           tools_4.0.2                
 [31] gtable_0.3.0                glue_1.4.2                 
 [33] GenomeInfoDbData_1.2.3      dplyr_1.0.2                
 [35] rappdirs_0.3.1              Rcpp_1.0.5                 
 [37] vctrs_0.3.5                 Biostrings_2.56.0          
 [39] xfun_0.19                   stringr_1.4.0              
 [41] lifecycle_0.2.0             ensembldb_2.12.1           
 [43] XML_3.99-0.5                zlibbioc_1.34.0            
 [45] scales_1.1.1                BSgenome_1.56.0            
 [47] VariantAnnotation_1.34.0    ProtGenerics_1.20.0        
 [49] hms_0.5.3                   promises_1.1.1             
 [51] SummarizedExperiment_1.18.2 AnnotationFilter_1.12.0    
 [53] RColorBrewer_1.1-2          yaml_2.2.1                 
 [55] curl_4.3                    memoise_1.1.0              
 [57] gridExtra_2.3               biomaRt_2.44.4             
 [59] rpart_4.1-15                latticeExtra_0.6-29        
 [61] stringi_1.5.3               RSQLite_2.2.1              
 [63] checkmate_2.0.0             BiocParallel_1.22.0        
 [65] rlang_0.4.9                 pkgconfig_2.0.3            
 [67] matrixStats_0.57.0          bitops_1.0-6               
 [69] evaluate_0.14               lattice_0.20-41            
 [71] purrr_0.3.4                 GenomicAlignments_1.24.0   
 [73] htmlwidgets_1.5.2           bit_4.0.4                  
 [75] tidyselect_1.1.0            magrittr_2.0.1             
 [77] R6_2.5.0                    generics_0.1.0             
 [79] Hmisc_4.4-2                 DelayedArray_0.14.1        
 [81] DBI_1.1.0                   pillar_1.4.7               
 [83] whisker_0.4                 foreign_0.8-80             
 [85] withr_2.3.0                 survival_3.2-7             
 [87] RCurl_1.98-1.2              nnet_7.3-14                
 [89] tibble_3.0.4                crayon_1.3.4               
 [91] BiocFileCache_1.12.1        rmarkdown_2.5              
 [93] jpeg_0.1-8.1                progress_1.2.2             
 [95] data.table_1.13.2           blob_1.2.1                 
 [97] git2r_0.27.1                digest_0.6.27              
 [99] httpuv_1.5.4                openssl_1.4.3              
[101] munsell_0.5.0               askpass_1.1                </code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
