<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2022-04-28" />

<title>Using the R seqinr package</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Using the R seqinr package</h1>
<h4 class="date">2022-04-28</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-04-28
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2020fbee9c97a4f26011f1606ea46d918daecc6dtargetblank2020fbea"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/davetang/muse/tree/2020fbee9c97a4f26011f1606ea46d918daecc6d" target="_blank">2020fbe</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2020fbee9c97a4f26011f1606ea46d918daecc6dtargetblank2020fbea" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/davetang/muse/tree/2020fbee9c97a4f26011f1606ea46d918daecc6d" target="_blank">2020fbe</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.1.2/

Untracked files:
    Untracked:  data/ncrna_NONCODE[v3.0].fasta.tar.gz
    Untracked:  data/ncrna_noncode_v3.fa

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/seqinr.Rmd</code>) and HTML (<code>docs/seqinr.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/2020fbee9c97a4f26011f1606ea46d918daecc6d/analysis/seqinr.Rmd" target="_blank">2020fbe</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
<td>
Fix link to FASTA
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/seqinr.html" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/9b739a1064d159ea0b0dd6d1a140857d3177dd74/analysis/seqinr.Rmd" target="_blank">9b739a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
<td>
Using the seqinr package
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The <a href="https://cran.r-project.org/web/packages/seqinr/index.html">seqinr package</a> provides many useful functions for working with biological sequences in R. We will use data from <a href="http://noncode.org/download.php">NONCODE</a> to demonstrate some features of <code>seqinr</code>.</p>
<pre><code>wget &#39;http://www.noncode.org/datadownload/ncrna_NONCODE[v3.0].fasta.tar.gz&#39;
tar xzf ncrna_NONCODE\[v3.0\].fasta.tar.gz
mv ncrna_NONCODE\[v3.0\].fasta ncrna_noncode_v3.fa
cat ncrna_noncode_v3.fa | grep &quot;^&gt;&quot; | wc -l
411553</code></pre>
<p>Show last couple of entries.</p>
<pre class="bash"><code>cat data/ncrna_noncode_v3.fa | grep &quot;^&gt;&quot; | tail -3</code></pre>
<pre><code>&gt;n424067 | GQ859162 | mRNAlike lncRNA | Homo sapiens | NEAT1 | lncRNAdb | 424067 | http://lncrnadb.com/Detail.aspx?TKeyID=105 | -1.3349500 | -0.2589092
&gt;n424073 | AK035706 | mRNAlike lncRNA | Mus musculus | HOTAIR | lncRNAdb | 343067 | http://lncrnadb.com/Detail.aspx?TKeyID=97 | -1.4335300 | -0.2747758
Binary file (standard input) matches</code></pre>
<p>Install (if missing) and load <code>seqinr</code>.</p>
<pre class="r"><code>if(!require(&quot;seqinr&quot;, quietly = TRUE)){
  install.packages(&quot;seqinr&quot;)
}
library(&quot;seqinr&quot;)</code></pre>
<p>The <code>read.fasta</code> function is used to load a FASTA file and we will use it to load <code>ncrna_noncode_v3.fa</code>.</p>
<pre class="r"><code>ncrna &lt;- read.fasta(&quot;data/ncrna_noncode_v3.fa&quot;)
length(ncrna)</code></pre>
<pre><code>[1] 411553</code></pre>
<p>The entries are saved in a list.</p>
<pre class="r"><code>class(ncrna)</code></pre>
<pre><code>[1] &quot;list&quot;</code></pre>
<p>Each list item is named after the first annotation in the FASTA file.</p>
<pre class="r"><code>head(names(ncrna))</code></pre>
<pre><code>[1] &quot;nncid&quot; &quot;n1&quot;    &quot;n2&quot;    &quot;n3&quot;    &quot;n4&quot;    &quot;n5&quot;   </code></pre>
<p>Check the first entry, which is stored in index 2, as the first entry is a fake FASTA entry that contains some information on the annotations stored in the FASTA file.</p>
<pre class="r"><code>ncrna[[2]]</code></pre>
<pre><code>  [1] &quot;a&quot; &quot;c&quot; &quot;c&quot; &quot;t&quot; &quot;c&quot; &quot;g&quot; &quot;a&quot; &quot;c&quot; &quot;c&quot; &quot;a&quot; &quot;c&quot; &quot;c&quot; &quot;c&quot; &quot;t&quot; &quot;t&quot; &quot;a&quot; &quot;a&quot; &quot;c&quot;
 [19] &quot;t&quot; &quot;t&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot; &quot;t&quot; &quot;g&quot; &quot;c&quot; &quot;a&quot; &quot;g&quot; &quot;g&quot; &quot;t&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot;
 [37] &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;g&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;g&quot;
 [55] &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;g&quot; &quot;g&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot;
 [73] &quot;t&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;g&quot; &quot;g&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot; &quot;a&quot; &quot;t&quot; &quot;g&quot; &quot;g&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot;
 [91] &quot;t&quot; &quot;t&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;g&quot; &quot;a&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;t&quot; &quot;c&quot; &quot;a&quot;
[109] &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;t&quot; &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot;
[127] &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot;
[145] &quot;t&quot; &quot;c&quot; &quot;g&quot; &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;t&quot; &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot; &quot;c&quot; &quot;c&quot; &quot;c&quot;
[163] &quot;t&quot; &quot;t&quot; &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot; &quot;t&quot; &quot;c&quot; &quot;c&quot; &quot;a&quot; &quot;t&quot; &quot;c&quot; &quot;t&quot; &quot;t&quot; &quot;a&quot;
[181] &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;t&quot; &quot;a&quot; &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;g&quot; &quot;t&quot; &quot;t&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;a&quot; &quot;a&quot; &quot;c&quot; &quot;t&quot;
[199] &quot;t&quot; &quot;a&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;g&quot;
[217] &quot;c&quot; &quot;t&quot; &quot;t&quot; &quot;g&quot; &quot;t&quot; &quot;g&quot; &quot;g&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;g&quot; &quot;t&quot; &quot;c&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot; &quot;t&quot;
[235] &quot;a&quot; &quot;c&quot; &quot;a&quot; &quot;a&quot; &quot;t&quot; &quot;g&quot; &quot;t&quot; &quot;c&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot; &quot;g&quot;
attr(,&quot;name&quot;)
[1] &quot;n1&quot;
attr(,&quot;Annot&quot;)
[1] &quot;&gt;n1 | AB002583 | tmRNA | chloroplast Cyanidioschyzon merolae | ssrA | NONCODE v2.0 | NULL | NULL | -1.0577600 | -0.3460597&quot;
attr(,&quot;class&quot;)
[1] &quot;SeqFastadna&quot;</code></pre>
<p>Count nucleotides.</p>
<pre class="r"><code>count(ncrna[[2]], 1)</code></pre>
<pre><code>
 a  c  g  t 
86 40 44 76 </code></pre>
<p>Count di-nucleotides.</p>
<pre class="r"><code>count(ncrna[[2]], 2)</code></pre>
<pre><code>
aa ac ag at ca cc cg ct ga gc gg gt ta tc tg tt 
38 15  9 24 16  7  5 12 15  4 14 10 16 14 16 30 </code></pre>
<p>GC percent.</p>
<pre class="r"><code>GC(ncrna[[2]])</code></pre>
<pre><code>[1] 0.3414634</code></pre>
<p>Store entire FASTA headers.</p>
<pre class="r"><code>my_header &lt;- getAnnot(ncrna)</code></pre>
<p>Create data frame using FASTA header.</p>
<pre class="r"><code>head(my_header[-1])</code></pre>
<pre><code>[[1]]
[1] &quot;&gt;n1 | AB002583 | tmRNA | chloroplast Cyanidioschyzon merolae | ssrA | NONCODE v2.0 | NULL | NULL | -1.0577600 | -0.3460597&quot;

[[2]]
[1] &quot;&gt;n2 | AB002583 | RNase P RNA | chloroplast Cyanidioschyzon merolae | rnpB | NONCODE v2.0 | NULL | NULL | -1.1143400 | -0.4482952&quot;

[[3]]
[1] &quot;&gt;n3 | AB003477 | tmRNA | Synechococcus sp | 10Sa | NONCODE v2.0 | NULL | NULL | -1.3614100 | -0.3138589&quot;

[[4]]
[1] &quot;&gt;n4 | AB007644 | snoRNA | Arabidopsis thaliana (thale cress) | U3 | NONCODE v2.0 | NULL | NULL | -0.5276610 | -0.3278265&quot;

[[5]]
[1] &quot;&gt;n5 | AB009049 | snoRNA | Arabidopsis thaliana (thale cress) | U24 | NONCODE v2.0 | NULL | NULL | -1.0309300 | -0.5177013&quot;

[[6]]
[1] &quot;&gt;n6 | AB009051 | snRNA | Arabidopsis thaliana (thale cress) | U6 | NONCODE v2.0 | NULL | NULL | -1.3622000 | -0.1069238&quot;</code></pre>
<pre class="r"><code>my_split &lt;- lapply(my_header[-1], function(x){
  str_split(x, &quot; \\| &quot;, simplify = TRUE)
})

my_df &lt;- do.call(rbind.data.frame, my_split)

names(my_df) &lt;- as.vector(str_split(my_header[[1]], &quot; \\| &quot;, simplify = TRUE))

my_df %&gt;%
  rename(nncid = &quot;&gt;nncid&quot;) %&gt;%
  mutate(nncid = sub(&quot;^&gt;&quot;, &quot;&quot;, nncid)) -&gt; my_df

head(my_df)</code></pre>
<pre><code>  nncid     accn       class                            organism name
1    n1 AB002583       tmRNA chloroplast Cyanidioschyzon merolae ssrA
2    n2 AB002583 RNase P RNA chloroplast Cyanidioschyzon merolae rnpB
3    n3 AB003477       tmRNA                    Synechococcus sp 10Sa
4    n4 AB007644      snoRNA  Arabidopsis thaliana (thale cress)   U3
5    n5 AB009049      snoRNA  Arabidopsis thaliana (thale cress)  U24
6    n6 AB009051       snRNA  Arabidopsis thaliana (thale cress)   U6
           ref transcriptID  url   cpcScore       cnci
1 NONCODE v2.0         NULL NULL -1.0577600 -0.3460597
2 NONCODE v2.0         NULL NULL -1.1143400 -0.4482952
3 NONCODE v2.0         NULL NULL -1.3614100 -0.3138589
4 NONCODE v2.0         NULL NULL -0.5276610 -0.3278265
5 NONCODE v2.0         NULL NULL -1.0309300 -0.5177013
6 NONCODE v2.0         NULL NULL -1.3622000 -0.1069238</code></pre>
<p>Organisms with the most ncRNAs.</p>
<pre class="r"><code>my_df %&gt;%
  group_by(organism) %&gt;%
  summarise(tally = n()) %&gt;%
  arrange(desc(tally)) %&gt;%
  head()</code></pre>
<pre><code># A tibble: 6 × 2
  organism                tally
  &lt;chr&gt;                   &lt;int&gt;
1 Mus musculus           119597
2 Drosophila melanogas   102171
3 Homo sapiens            91067
4 Norway rat              66760
5 NULL                    15781
6 Caenorhabditis elegans   4718</code></pre>
<p>Class with the most ncRNAs.</p>
<pre class="r"><code>my_df %&gt;%
  group_by(class) %&gt;%
  summarise(tally = n()) %&gt;%
  arrange(desc(tally)) %&gt;%
  head()</code></pre>
<pre><code># A tibble: 6 × 2
  class              tally
  &lt;chr&gt;              &lt;int&gt;
1 piRNA             174724
2 mature_transcript 102046
3 lncRNA             50615
4 mRNAlike lncRNA    43530
5 miRNA              20550
6 other               4192</code></pre>
<p>Find all human piwi-interacting RNAs (piRNAs) and store their <code>nncid</code>.</p>
<pre class="r"><code>my_df %&gt;%
  filter(class == &quot;piRNA&quot;, organism ==&quot;Homo sapiens&quot;) %&gt;%
  pull(nncid) -&gt; nncid_human_pirna

length(nncid_human_pirna)</code></pre>
<pre><code>[1] 32152</code></pre>
<p>Create <code>pirna_human</code> for storing human piRNAs.</p>
<pre class="r"><code>pirna_human &lt;- ncrna[nncid_human_pirna]
getSequence(pirna_human[[1]], as.string = TRUE)</code></pre>
<pre><code>[[1]]
[1] &quot;tagtgatgtgttcgttggtaagaggga&quot;</code></pre>
<p>Report number of sequences with N’s and remove them.</p>
<pre class="r"><code>my_n &lt;- grepl(&#39;n&#39;, unlist(getSequence(pirna_human, as.string = TRUE)))
pirna_human &lt;- pirna_human[!my_n]
table(my_n)</code></pre>
<pre><code>my_n
FALSE  TRUE 
31580   572 </code></pre>
<p>Save <code>pirna_human</code> as a FASTA file (not run).</p>
<pre class="r"><code>my_anno &lt;- getAnnot(pirna_human)
my_anno &lt;- lapply(my_anno, function(x) sub(&quot;^&gt;&quot;, &quot;&quot;, x))
write.fasta(sequences = pirna_human, names = my_anno, file.out = &quot;human_pirna.fa&quot;)</code></pre>
<p>piRNAs typically start with U/T.</p>
<pre class="r"><code>prop.table(table(sapply(pirna_human, function(x) x[1])))</code></pre>
<pre><code>
         a          c          g          t 
0.05984801 0.06614946 0.08429386 0.78970868 </code></pre>
<p>In addition, piRNAs typically have an A at the tenth base and the proportion below is slightly higher than 0.25 if we expect an equal distribution.</p>
<pre class="r"><code>prop.table(table(sapply(pirna_human, function(x) x[10])))</code></pre>
<pre><code>
        a         c         g         t 
0.3257758 0.1906586 0.2499683 0.2335972 </code></pre>
<p>Length distribution of piRNAs.</p>
<pre class="r"><code>table(getLength(pirna_human))</code></pre>
<pre><code>
  16   18   19   20   21   22   23   24   25   26   27   28   29   30   31   32 
   2    1    3    6    1    6    1    4    7 4186 3792 3901 6580 7662 4238 1188 
  33 
   2 </code></pre>
<p>piRNAs are typically 26-31 nucleotide long, as observed below.</p>
<pre class="r"><code>as.data.frame(table(getLength(pirna_human))) %&gt;%
  ggplot(., aes(Var1, Freq)) +
    geom_col() +
    labs(x = &quot;piRNA length&quot;, y = &quot;Frequency&quot;) +
    theme_bw()</code></pre>
<p><img src="figure/seqinr.Rmd/plot_length_dist-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_length_dist-1">
Past versions of plot_length_dist-1.png
</button>
</p>
<div id="fig-plot_length_dist-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/plot_length_dist-1.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>10th base proportion for different lengths.</p>
<pre class="r"><code>my_len &lt;- 26:32
my_prop &lt;- sapply(my_len, function(x){
  wanted &lt;- getLength(pirna_human) == x
  prop.table(table(sapply(pirna_human[wanted], function(x) x[10])))
})

as.data.frame(t(my_prop), row.names = my_len)</code></pre>
<pre><code>           a         c         g         t
26 0.2730530 0.1889632 0.2859532 0.2520306
27 0.3364979 0.1993671 0.2217827 0.2423523
28 0.3429890 0.1920021 0.2260959 0.2389131
29 0.3354103 0.1925532 0.2428571 0.2291793
30 0.3308536 0.1866353 0.2582877 0.2242234
31 0.3265691 0.1885323 0.2550731 0.2298254
32 0.3324916 0.1860269 0.2584175 0.2230640</code></pre>
<p>Frequencies of nucleotides along every position.</p>
<pre class="r"><code>my_lens &lt;- 26:32
for (my_len in my_lens){
  wanted &lt;- getLength(pirna_human) == my_len
  my_seq &lt;- getSequence(pirna_human[wanted])
  my_freq &lt;- apply(do.call(rbind, my_seq), 2, function(x) prop.table(table(x)))
  
  as.data.frame(my_freq) %&gt;%
    mutate(nuc = c(&#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;)) %&gt;%
    select(nuc, everything()) %&gt;%
    pivot_longer(!nuc, names_to = &quot;Position&quot;, values_to = &quot;Frequency&quot;) %&gt;%
    mutate(Position = as.integer(sub(&quot;^V&quot;, &quot;&quot;, Position))) -&gt; my_freq_df
  
  ggplot(my_freq_df, aes(Position, Frequency, colour = nuc)) +
    geom_line() +
    geom_point() +
    theme_bw() +
    ggtitle(paste0(&quot;Nucleotide frequencies along &quot;, my_len, &quot; nt human piRNAs&quot;)) -&gt; p
  
  print(p)
}</code></pre>
<p><img src="figure/seqinr.Rmd/freq_nuc_along-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-freq_nuc_along-1">
Past versions of freq_nuc_along-1.png
</button>
</p>
<div id="fig-freq_nuc_along-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/freq_nuc_along-1.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/seqinr.Rmd/freq_nuc_along-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-freq_nuc_along-2">
Past versions of freq_nuc_along-2.png
</button>
</p>
<div id="fig-freq_nuc_along-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/freq_nuc_along-2.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/seqinr.Rmd/freq_nuc_along-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-freq_nuc_along-3">
Past versions of freq_nuc_along-3.png
</button>
</p>
<div id="fig-freq_nuc_along-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/freq_nuc_along-3.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/seqinr.Rmd/freq_nuc_along-4.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-freq_nuc_along-4">
Past versions of freq_nuc_along-4.png
</button>
</p>
<div id="fig-freq_nuc_along-4" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/freq_nuc_along-4.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/seqinr.Rmd/freq_nuc_along-5.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-freq_nuc_along-5">
Past versions of freq_nuc_along-5.png
</button>
</p>
<div id="fig-freq_nuc_along-5" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/freq_nuc_along-5.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/seqinr.Rmd/freq_nuc_along-6.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-freq_nuc_along-6">
Past versions of freq_nuc_along-6.png
</button>
</p>
<div id="fig-freq_nuc_along-6" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/freq_nuc_along-6.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/seqinr.Rmd/freq_nuc_along-7.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-freq_nuc_along-7">
Past versions of freq_nuc_along-7.png
</button>
</p>
<div id="fig-freq_nuc_along-7" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b814eb157bc5fda2affabe514c6c5df78bd14a2b/docs/figure/seqinr.Rmd/freq_nuc_along-7.png" target="_blank">b814eb1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Please refer to the <a href="https://cran.r-project.org/web/packages/seqinr/seqinr.pdf">seqinr manual</a> for further information.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] seqinr_4.2-8    forcats_0.5.1   stringr_1.4.0   dplyr_1.0.7    
 [5] purrr_0.3.4     readr_2.1.1     tidyr_1.1.4     tibble_3.1.6   
 [9] ggplot2_3.3.5   tidyverse_1.3.1 workflowr_1.7.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.8       lubridate_1.8.0  getPass_0.2-2    ps_1.6.0        
 [5] assertthat_0.2.1 rprojroot_2.0.2  digest_0.6.29    utf8_1.2.2      
 [9] R6_2.5.1         cellranger_1.1.0 backports_1.4.1  reprex_2.0.1    
[13] evaluate_0.14    highr_0.9        httr_1.4.2       pillar_1.6.5    
[17] rlang_1.0.0      readxl_1.3.1     rstudioapi_0.13  whisker_0.4     
[21] callr_3.7.0      jquerylib_0.1.4  rmarkdown_2.11   labeling_0.4.2  
[25] munsell_0.5.0    broom_0.7.11     compiler_4.1.2   httpuv_1.6.5    
[29] modelr_0.1.8     xfun_0.29        pkgconfig_2.0.3  htmltools_0.5.2 
[33] tidyselect_1.1.1 fansi_1.0.2      crayon_1.4.2     tzdb_0.2.0      
[37] dbplyr_2.1.1     withr_2.4.3      later_1.3.0      MASS_7.3-54     
[41] grid_4.1.2       jsonlite_1.7.3   gtable_0.3.0     lifecycle_1.0.1 
[45] DBI_1.1.2        git2r_0.29.0     magrittr_2.0.2   scales_1.1.1    
[49] cli_3.1.1        stringi_1.7.6    farver_2.1.0     fs_1.5.2        
[53] promises_1.2.0.1 xml2_1.3.3       ellipsis_0.3.2   generics_0.1.1  
[57] vctrs_0.3.8      tools_4.1.2      ade4_1.7-19      glue_1.6.1      
[61] hms_1.1.1        processx_3.5.2   fastmap_1.1.0    yaml_2.2.2      
[65] colorspace_2.0-2 rvest_1.0.2      knitr_1.37       haven_2.4.3     </code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
