<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2022-11-08" />

<title>Sequence composition and Random Forests</title>

<script src="site_libs/header-attrs-2.17/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Sequence composition and Random
Forests</h1>
<h4 class="date">2022-11-08</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-11-08
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreece8e44718528b510196ec41432184aa2313dd702targetblankce8e447a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/ce8e44718528b510196ec41432184aa2313dd702" target="_blank">ce8e447</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreece8e44718528b510196ec41432184aa2313dd702targetblankce8e447a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/ce8e44718528b510196ec41432184aa2313dd702" target="_blank">ce8e447</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/figure/
    Ignored:    r_packages_4.1.2/
    Ignored:    r_packages_4.2.0/

Untracked files:
    Untracked:  analysis/cell_ranger.Rmd
    Untracked:  data/ncrna_NONCODE[v3.0].fasta.tar.gz
    Untracked:  data/ncrna_noncode_v3.fa
    Untracked:  data/tss.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/tss_predict.Rmd</code>) and HTML
(<code>docs/tss_predict.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/ce8e44718528b510196ec41432184aa2313dd702/analysis/tss_predict.Rmd" target="_blank">ce8e447</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-11-08
</td>
<td>
Update notebook
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/9ec0dec626f4d02266a3b53c5ae208519acea033/docs/tss_predict.html" target="_blank">9ec0dec</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/56b1a796a52ee226a819290f1a36cf591ed9fd60/analysis/tss_predict.Rmd" target="_blank">56b1a79</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
<td>
Positional di-nuc freq
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/tss_predict.html" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/06164641bd8f7307763af70a208dad096f524607/analysis/tss_predict.Rmd" target="_blank">0616464</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
<td>
Random forest model
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/fd41c759545ae9429df95eaf79a7ddca9447d5c9/docs/tss_predict.html" target="_blank">fd41c75</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/e372c815ee309a4b014e441268a2ec5416f8ced8/analysis/tss_predict.Rmd" target="_blank">e372c81</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-22
</td>
<td>
TSS
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Transcriptional starting sites (TSSs) demarcate the first position in
the DNA sequence that gets transcribed into RNA. In this work, we will
build a Random Forest classifier that predicts whether a sequence of
nucleotides is a TSSs or not. We will train our classifier using TSS
sequences based on the NCBI Reference Sequence Database (RefSeq
database). Building an accurate classifier will allow us to predict
whether an unknown sequence is likely a TSS or not.</p>
</div>
<div id="packages" class="section level2">
<h2>Packages</h2>
<p>First we’ll define a function for loading required packages (and to
install them first, if missing).</p>
<pre class="r"><code>load_package &lt;- function(x, source = &quot;bioc&quot;){
  if(!require(x, character.only = TRUE, quietly = TRUE)){
    if (source == &quot;bioc&quot;){
      BiocManager::install(x, character.only = TRUE)
    } else if (source == &quot;cran&quot;){
      install.packages(x, character.only = TRUE)
    } else {
      stop(&quot;Unrecognised source&quot;)
    }
  }
  library(x, character.only = TRUE)
}</code></pre>
</div>
<div id="refseq" class="section level2">
<h2>RefSeq</h2>
<p>We will use the Bioconductor package <code>biomaRt</code> to download
the entire collection of RefSeq sequences.</p>
<pre class="r"><code>load_package(&#39;biomaRt&#39;)</code></pre>
<p>Find the human gene set.</p>
<pre class="r"><code>ensembl &lt;- useMart(&#39;ensembl&#39;, host = params$host)
biomaRt::listDatasets(ensembl) %&gt;%
  filter(grepl(&#39;human&#39;, description, TRUE))</code></pre>
<pre><code>                dataset              description    version
1 hsapiens_gene_ensembl Human genes (GRCh38.p13) GRCh38.p13</code></pre>
<p>List RefSeq attributes for use with <code>getBM</code>.</p>
<pre class="r"><code>biomaRt::listDatasets(ensembl) %&gt;%
  filter(grepl(&#39;human&#39;, description, TRUE)) %&gt;%
  pull(dataset) -&gt; my_dataset

ensembl &lt;- useMart(&#39;ensembl&#39;, dataset = my_dataset, host = params$host)
listAttributes(ensembl) %&gt;%
  filter(grepl(&#39;refseq&#39;, description, TRUE))</code></pre>
<pre><code>                           name                                  description
1        transcript_mane_select        RefSeq match transcript (MANE Select)
2 transcript_mane_plus_clinical RefSeq match transcript (MANE Plus Clinical)
3                   refseq_mrna                               RefSeq mRNA ID
4         refseq_mrna_predicted                     RefSeq mRNA predicted ID
5                  refseq_ncrna                              RefSeq ncRNA ID
6        refseq_ncrna_predicted                    RefSeq ncRNA predicted ID
7                refseq_peptide                            RefSeq peptide ID
8      refseq_peptide_predicted                  RefSeq peptide predicted ID
          page
1 feature_page
2 feature_page
3 feature_page
4 feature_page
5 feature_page
6 feature_page
7 feature_page
8 feature_page</code></pre>
<p>Fetch all RefSeq mRNA IDs on assembled chromosomes.</p>
<pre class="r"><code>my_chr &lt;- c(1:22, &#39;X&#39;, &#39;Y&#39;)

my_refseq &lt;- getBM(
  attributes=&#39;refseq_mrna&#39;,
  filters = &#39;chromosome_name&#39;,
  values = my_chr,
  mart = ensembl
)

head(my_refseq)</code></pre>
<pre><code>   refseq_mrna
1    NM_004676
2    NM_004654
3 NM_001002758
4    NM_004681
5 NM_001278612
6    NM_004678</code></pre>
<p>Number of RefSeq IDs.</p>
<pre class="r"><code>length(my_refseq$refseq_mrna)</code></pre>
<pre><code>[1] 62674</code></pre>
<p>We will perform another <code>getBM</code> query to build a table
with the start and end coordinates of each mRNA/transcript. Note that
when working with transcripts, use the attributes
<code>transcript_start</code> and <code>transcript_end</code>, and not
the attributes <code>start_position</code> and
<code>end_position</code>; using those will give you the Ensembl gene
coordinates and not the RefSeq coordinates, which is what we want! (I
have retrieved the Ensembl coordinates for illustrative purposes.)</p>
<pre class="r"><code>my_att &lt;- c(
  &#39;refseq_mrna&#39;,
  &#39;chromosome_name&#39;,
  &#39;transcript_start&#39;,
  &#39;transcript_end&#39;,
  &#39;start_position&#39;,
  &#39;end_position&#39;,
  &#39;strand&#39;
)

my_refseq_loci &lt;- getBM(
  attributes = my_att,
  filters = c(&#39;refseq_mrna&#39;, &#39;chromosome_name&#39;),
  values = list(refseq_mrna = my_refseq$refseq_mrna, chromosome_name = my_chr),
  mart = ensembl
)

dim(my_refseq_loci)</code></pre>
<pre><code>[1] 62704     7</code></pre>
<p>Check out the table.</p>
<pre class="r"><code>head(my_refseq_loci)</code></pre>
<pre><code>  refseq_mrna chromosome_name transcript_start transcript_end start_position
1   NM_000015               8         18391282       18401218       18391282
2   NM_000046               5         78777209       78985310       78777209
3   NM_000071              21         43053191       43075835       43053191
4   NM_000084               X         50067576       50092406       49922596
5   NM_000099              20         23633657       23637955       23626706
6   NM_000100              21         43773950       43776308       43772511
  end_position strand
1     18401218      1
2     78986087     -1
3     43076943     -1
4     50099235      1
5     23638473     -1
6     43776330     -1</code></pre>
<p>Check for duplicated entries.</p>
<pre class="r"><code>table(duplicated(my_refseq_loci$refseq_mrna))</code></pre>
<pre><code>
FALSE  TRUE 
62674    30 </code></pre>
<p>Removed duplicated entries.</p>
<pre class="r"><code>my_refseq_loci_uniq &lt;- my_refseq_loci[!duplicated(my_refseq_loci$refseq_mrna),]
dim(my_refseq_loci_uniq)</code></pre>
<pre><code>[1] 62674     7</code></pre>
<p>We will modify <code>chromosome_name</code> and <code>strand</code>
to make it compatible with <code>BSgenome.Hsapiens.UCSC.hg38</code>.</p>
<pre class="r"><code>my_refseq_loci_uniq %&gt;%
  mutate(strand = ifelse(strand == 1, yes = &#39;+&#39;, no = &#39;-&#39;)) %&gt;%
  mutate(chromosome_name = sub(&quot;^&quot;, &quot;chr&quot;, chromosome_name)) -&gt; my_refseq_loci_uniq

head(my_refseq_loci_uniq)</code></pre>
<pre><code>  refseq_mrna chromosome_name transcript_start transcript_end start_position
1   NM_000015            chr8         18391282       18401218       18391282
2   NM_000046            chr5         78777209       78985310       78777209
3   NM_000071           chr21         43053191       43075835       43053191
4   NM_000084            chrX         50067576       50092406       49922596
5   NM_000099           chr20         23633657       23637955       23626706
6   NM_000100           chr21         43773950       43776308       43772511
  end_position strand
1     18401218      +
2     78986087      -
3     43076943      -
4     50099235      +
5     23638473      -
6     43776330      -</code></pre>
</div>
<div id="transcription-start-site" class="section level2">
<h2>Transcription Start Site</h2>
<p>The TSS is defined by <code>transcript_start</code> when the
transcript is on the <code>+</code> strand and by
<code>transcript_end</code> if the strand is <code>-</code>. We want to
retrieve nucleotides upstream and downstream of the TSS, so we will
subtract and add coordinates accordingly.</p>
<pre class="r"><code>my_tss &lt;- list()

my_tss$loci &lt;- my_refseq_loci_uniq %&gt;%
  dplyr::select(-c(start_position, end_position)) %&gt;%
  mutate(tss_start = if_else(strand == &quot;+&quot;, transcript_start - params$span, transcript_end - params$span)) %&gt;%
  mutate(tss_end = if_else(strand == &quot;+&quot;, transcript_start + params$span, transcript_end + params$span))

head(my_tss$loci)</code></pre>
<pre><code>  refseq_mrna chromosome_name transcript_start transcript_end strand tss_start
1   NM_000015            chr8         18391282       18401218      +  18391280
2   NM_000046            chr5         78777209       78985310      -  78985308
3   NM_000071           chr21         43053191       43075835      -  43075833
4   NM_000084            chrX         50067576       50092406      +  50067574
5   NM_000099           chr20         23633657       23637955      -  23637953
6   NM_000100           chr21         43773950       43776308      -  43776306
   tss_end
1 18391284
2 78985312
3 43075837
4 50067578
5 23637957
6 43776310</code></pre>
</div>
<div id="random-loci" class="section level2">
<h2>Random loci</h2>
<p>Non-TSS sequences will be generated by randomly sampling hg38, which
is available via the <code>BSgenome.Hsapiens.UCSC.hg38</code>
package.</p>
<pre class="r"><code>load_package(&#39;BSgenome.Hsapiens.UCSC.hg38&#39;)</code></pre>
<p>Save chromosome names and their sizes.</p>
<pre class="r"><code>hg38_ref &lt;- seqnames(BSgenome.Hsapiens.UCSC.hg38)

# remove unassembled and mitochondrial sequences
hg38_chr &lt;- hg38_ref[!grepl(&quot;_&quot;, hg38_ref)]
hg38_chr &lt;- hg38_chr[!grepl(&quot;chrM&quot;, hg38_chr)]

hg38_chr_size &lt;- sapply(hg38_chr, function(x){
  length(BSgenome.Hsapiens.UCSC.hg38[[x]])
})

head(hg38_chr_size)</code></pre>
<pre><code>     chr1      chr2      chr3      chr4      chr5      chr6 
248956422 242193529 198295559 190214555 181538259 170805979 </code></pre>
<p>Sample chromosome to match the number of TSSs.</p>
<pre class="r"><code>set.seed(1984)
n &lt;- nrow(my_tss$loci)

sstrand &lt;- sample(c(&quot;+&quot;, &quot;-&quot;), n, TRUE)
sstart &lt;- sapply(my_tss$loci$chromosome_name, function(x){
  sample(hg38_chr_size[x] - (params$span * 2), 1)
})
send &lt;- sstart + (params$span*2)

my_random &lt;- list()

my_random$loci &lt;- data.frame(
  chr = my_tss$loci$chromosome_name,
  start = sstart,
  end = send,
  strand = sstrand
)

head(my_random$loci)</code></pre>
<pre><code>    chr     start       end strand
1  chr8 130846725 130846729      -
2  chr5 136222485 136222489      -
3 chr21  17034887  17034891      -
4  chrX 122209123 122209127      -
5 chr20  27691322  27691326      +
6 chr21  10253649  10253653      +</code></pre>
</div>
<div id="removing-overlaps" class="section level2">
<h2>Removing overlaps</h2>
<p>We will use the <code>GenomicRanges</code> package to remove randomly
sampled regions that may have overlapped with TSS regions.</p>
<pre class="r"><code>load_package(&quot;GenomicRanges&quot;)</code></pre>
<p>First we will create a <code>GRanges</code> object using
<code>my_refseq_loci</code>.</p>
<pre class="r"><code>my_tss$gr &lt;- with(
  my_tss$loci,
  GRanges(chromosome_name,
          IRanges(tss_start, tss_end, names = refseq_mrna),
          strand
  )
)

head(my_tss$gr)</code></pre>
<pre><code>GRanges object with 6 ranges and 0 metadata columns:
            seqnames            ranges strand
               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
  NM_000015     chr8 18391280-18391284      +
  NM_000046     chr5 78985308-78985312      -
  NM_000071    chr21 43075833-43075837      -
  NM_000084     chrX 50067574-50067578      +
  NM_000099    chr20 23637953-23637957      -
  NM_000100    chr21 43776306-43776310      -
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Next we will create a <code>GRanges</code> object for the random
loci.</p>
<pre class="r"><code>my_random$gr &lt;- with(
  my_random$loci,
  GRanges(chr,
          IRanges(start, end, names = paste(chr, start, end, sep = &quot;_&quot;)),
          strand
  )
)

head(my_random$gr)</code></pre>
<pre><code>GRanges object with 6 ranges and 0 metadata columns:
                           seqnames              ranges strand
                              &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
  chr8_130846725_130846729     chr8 130846725-130846729      -
  chr5_136222485_136222489     chr5 136222485-136222489      -
   chr21_17034887_17034891    chr21   17034887-17034891      -
  chrX_122209123_122209127     chrX 122209123-122209127      -
   chr20_27691322_27691326    chr20   27691322-27691326      +
   chr21_10253649_10253653    chr21   10253649-10253653      +
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
<p>The <code>countOverlaps</code> function will let us know the number
of overlaps between random regions and TSSs.</p>
<pre class="r"><code>table(countOverlaps(my_random$gr, my_tss$gr))</code></pre>
<pre><code>
    0     1 
62670     4 </code></pre>
<p>Remove overlapping random sites.</p>
<pre class="r"><code>my_random$loci_no &lt;- my_random$loci[countOverlaps(my_random$gr, my_tss$gr) == 0, ]
dim(my_random$loci_no)</code></pre>
<pre><code>[1] 62670     4</code></pre>
<p>We will also check whether TSS regions overlap each other.</p>
<pre class="r"><code>table(countOverlaps(my_tss$gr, my_tss$gr, minoverlap = 1))</code></pre>
<pre><code>
    1     2     3     4     5     6     7     8     9    10    11    12    13 
25962 11653  6071  4042  2700  1935  1499  1104  1006   834   652   577   514 
   14    15    16    17    18    19    20    21    22    23    24    25    26 
  447   296   414   371   255   268   238   188   155    97   120    50    70 
   27    28    29    30    31    32    33    34    35    36    38    43    44 
   62   140    58    60    31    64    99    34    70   108    76    43    44 
   45    46    50    54    72 
   45    46    50    54    72 </code></pre>
<p>We will also create a set of non-overlapping TSS regions (the
conditional is <code>== 1</code> because we are comparing regions
against itself and 1 is the self overlap), where a TSS does not overlap
by a single bp.</p>
<pre class="r"><code>my_tss$loci_no &lt;- my_tss$loci[countOverlaps(my_tss$gr, my_tss$gr, minoverlap = 1) == 1, ]
dim(my_tss$loci_no)</code></pre>
<pre><code>[1] 25962     7</code></pre>
</div>
<div id="sequence" class="section level2">
<h2>Sequence</h2>
<p>At this point, we have our TSS and random coordinates but we are
missing the nucleotide sequences that correspond to the defined
coordinates. The <code>getSeq</code> function can be used to fetch the
sequence.</p>
<pre class="r"><code>my_random$seq &lt;- getSeq(
  BSgenome.Hsapiens.UCSC.hg38,
  names = my_random$loci$chr,
  start = my_random$loci$start,
  end = my_random$loci$end,
  strand = my_random$loci$strand
)

head(my_random$seq)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 AAGGA                                             chr8
[2]     5 ATGTG                                             chr5
[3]     5 TAGTG                                             chr21
[4]     5 AGATG                                             chrX
[5]     5 CAGAA                                             chr20
[6]     5 NNNNN                                             chr21</code></pre>
<p>Remove entries with N’s.</p>
<pre class="r"><code>my_random$seq &lt;- my_random$seq[! grepl(&quot;N&quot;, my_random$seq)]
head(my_random$seq)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 AAGGA                                             chr8
[2]     5 ATGTG                                             chr5
[3]     5 TAGTG                                             chr21
[4]     5 AGATG                                             chrX
[5]     5 CAGAA                                             chr20
[6]     5 TCCCT                                             chr13</code></pre>
<p>Obtain TSS sequences.</p>
<pre class="r"><code>my_tss$seq &lt;- getSeq(
  BSgenome.Hsapiens.UCSC.hg38,
  names = my_tss$loci$chromosome_name,
  start = my_tss$loci$tss_start,
  end = my_tss$loci$tss_end,
  strand = my_tss$loci$strand
)

my_tss$seq &lt;- my_tss$seq[! grepl(&quot;N&quot;, my_tss$seq)]
head(my_tss$seq)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 GCACT                                             chr8
[2]     5 TCATT                                             chr5
[3]     5 GGGTC                                             chr21
[4]     5 CAGGG                                             chrX
[5]     5 TCACG                                             chr20
[6]     5 TCGCC                                             chr21</code></pre>
<p>Obtain non-overlapping TSS sequences.</p>
<pre class="r"><code>my_tss$seq_no &lt;- getSeq(
  BSgenome.Hsapiens.UCSC.hg38,
  names = my_tss$loci_no$chromosome_name,
  start = my_tss$loci_no$tss_start,
  end = my_tss$loci_no$tss_end,
  strand = my_tss$loci_no$strand
)

my_tss$seq_no &lt;- my_tss$seq_no[! grepl(&quot;N&quot;, my_tss$seq_no)]
head(my_tss$seq_no)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 GCACT                                             chr8
[2]     5 TCATT                                             chr5
[3]     5 CAGGG                                             chrX
[4]     5 TCACG                                             chr20
[5]     5 TCGCC                                             chr21
[6]     5 TTCTT                                             chr13</code></pre>
<p>The <code>dinucleotideFrequency</code> function can be used to
calculate the di-nucleotide frequency given a sequence string.</p>
<pre class="r"><code>my_random$di &lt;- dinucleotideFrequency(my_random$seq)
colSums(my_random$di)</code></pre>
<pre><code>   AA    AC    AG    AT    CA    CC    CG    CT    GA    GC    GG    GT    TA 
23283 12211 16924 18043 17744 12813  2661 16981 14501 10691 12848 12093 15212 
   TC    TG    TT 
14454 17551 23314 </code></pre>
<pre class="r"><code>my_tss$di &lt;- dinucleotideFrequency(my_tss$seq)
colSums(my_tss$di)</code></pre>
<pre><code>   AA    AC    AG    AT    CA    CC    CG    CT    GA    GC    GG    GT    TA 
 8572 11739 20484  8622 24740 21756 16399 15887 16045 24409 19844 14664  6980 
   TC    TG    TT 
15055 14683 10817 </code></pre>
<pre class="r"><code>my_tss$di_no &lt;- dinucleotideFrequency(my_tss$seq_no)
colSums(my_tss$di_no)</code></pre>
<pre><code>   AA    AC    AG    AT    CA    CC    CG    CT    GA    GC    GG    GT    TA 
 3703  4981  8592  3758 10473  8977  6313  6640  6704 10003  8061  5764  2884 
   TC    TG    TT 
 6443  6151  4401 </code></pre>
<p>Plot di-nucleotide frequencies between TSS and random regions.</p>
<pre class="r"><code>my_df &lt;- data.frame(
  x = colMeans(my_tss$di),
  y = colMeans(my_random$di),
  di = colnames(my_tss$di)
)

ggplot(my_df, aes(x, y, label = di)) +
  geom_point() +
  geom_text(nudge_x = 0.005) +
  geom_abline(slope = 1, lty = 3) +
  labs(x = &quot;TSS&quot;, y = &quot;Random&quot;) +
  theme_bw()</code></pre>
<p><img src="figure/tss_predict.Rmd/plot_di-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_di-1">
Past versions of plot_di-1.png
</button>
</p>
<div id="fig-plot_di-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/9ec0dec626f4d02266a3b53c5ae208519acea033/docs/figure/tss_predict.Rmd/plot_di-1.png" target="_blank">9ec0dec</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/figure/tss_predict.Rmd/plot_di-1.png" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot di-nucleotide frequencies between overlapping and
non-overlapping TSS regions.</p>
<pre class="r"><code>my_df &lt;- data.frame(
  x = colMeans(my_tss$di),
  y = colMeans(my_tss$di_no),
  di = colnames(my_tss$di)
)

ggplot(my_df, aes(x, y, label = di)) +
  geom_point() +
  geom_text(nudge_x = 0.005) +
  geom_abline(slope = 1, lty = 3) +
  labs(x = &quot;TSS&quot;, y = &quot;TSS no overlap&quot;) +
  theme_bw()</code></pre>
<p><img src="figure/tss_predict.Rmd/plot_di_tss-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_di_tss-1">
Past versions of plot_di_tss-1.png
</button>
</p>
<div id="fig-plot_di_tss-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/9ec0dec626f4d02266a3b53c5ae208519acea033/docs/figure/tss_predict.Rmd/plot_di_tss-1.png" target="_blank">9ec0dec</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="random-forests" class="section level2">
<h2>Random Forests</h2>
<p>A Random Forest classifier generally performs well and we will use it
to train a model to predict whether a sequence is a TSS or not based on
di-nucleotide frequencies.</p>
<pre class="r"><code>load_package(&#39;randomForest&#39;, source = &quot;cran&quot;)</code></pre>
<p>Combine TSS and random data. (We saw that overlapping and
non-overlapping TSSs had similar di-nucleotide frequencies, so we will
use the overlapping set.)</p>
<pre class="r"><code>as.data.frame(my_tss$di) %&gt;%
  mutate(class = &#39;tss&#39;) -&gt; my_data

as.data.frame(my_random$di) %&gt;%
  mutate(class = &#39;random&#39;) %&gt;%
  rbind(my_data) %&gt;%
  mutate(class = factor(class)) -&gt; my_data

dim(my_data)</code></pre>
<pre><code>[1] 123005     17</code></pre>
<p>Check out <code>my_data</code>.</p>
<pre class="r"><code>my_data[c(1:3, (nrow(my_data)-2):(nrow(my_data))), ]</code></pre>
<pre><code>       AA AC AG AT CA CC CG CT GA GC GG GT TA TC TG TT  class
1       1  0  1  0  0  0  0  0  1  0  1  0  0  0  0  0 random
2       0  0  0  1  0  0  0  0  0  0  0  1  0  0  2  0 random
3       0  0  1  0  0  0  0  0  0  0  0  1  1  0  1  0 random
123003  0  0  1  1  0  0  0  0  1  0  0  0  0  0  1  0    tss
123004  0  0  0  0  0  1  0  1  0  0  0  0  0  1  0  1    tss
123005  0  1  0  0  0  1  0  1  0  0  0  0  0  0  0  1    tss</code></pre>
<p>Split 80/20.</p>
<pre class="r"><code>set.seed(1984)
idx &lt;- sample(nrow(my_data), nrow(my_data)*0.8)

train &lt;- my_data[idx, ]
test &lt;- my_data[-idx, ]

prop.table(table(train$class))</code></pre>
<pre><code>
   random       tss 
0.4904577 0.5095423 </code></pre>
<pre class="r"><code>prop.table(table(test$class))</code></pre>
<pre><code>
   random       tss 
0.4905492 0.5094508 </code></pre>
<p>Train Random Forest.</p>
<pre class="r"><code>my_rf &lt;- randomForest(
  class ~ .,
  data = train,
  importance = TRUE,
  do.trace = 100,
  ntree = 500
)</code></pre>
<pre><code>ntree      OOB      1      2
  100:  27.13% 26.86% 27.38%
  200:  27.04% 26.68% 27.38%
  300:  27.03% 26.65% 27.39%
  400:  26.95% 26.69% 27.21%
  500:  26.95% 26.68% 27.20%</code></pre>
<pre class="r"><code>my_rf</code></pre>
<pre><code>
Call:
 randomForest(formula = class ~ ., data = train, importance = TRUE,      do.trace = 100, ntree = 500) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 4

        OOB estimate of  error rate: 26.95%
Confusion matrix:
       random   tss class.error
random  35385 12878   0.2668297
tss     13640 36501   0.2720329</code></pre>
<p>Train using more features.</p>
<pre class="r"><code>my_rf_6 &lt;- randomForest(
  class ~ .,
  data = train,
  importance = TRUE,
  do.trace = 100,
  ntree = 500,
  mtry = 6
)</code></pre>
<pre><code>ntree      OOB      1      2
  100:  27.03% 27.31% 26.75%
  200:  27.09% 27.74% 26.46%
  300:  27.06% 27.79% 26.35%
  400:  27.02% 27.81% 26.25%
  500:  27.09% 27.91% 26.30%</code></pre>
<pre class="r"><code>my_rf_6</code></pre>
<pre><code>
Call:
 randomForest(formula = class ~ ., data = train, importance = TRUE,      do.trace = 100, ntree = 500, mtry = 6) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 6

        OOB estimate of  error rate: 27.09%
Confusion matrix:
       random   tss class.error
random  34794 13469   0.2790751
tss     13185 36956   0.2629585</code></pre>
<p>Feature importance.</p>
<pre class="r"><code>rn &lt;- round(importance(my_rf), 2)
rn[order(rn[,3], decreasing=TRUE),]</code></pre>
<pre><code>   random   tss MeanDecreaseAccuracy MeanDecreaseGini
AA  35.80 64.89                83.54          1499.41
CG  58.97 57.37                77.89          2750.53
CA -42.96 72.08                76.85           832.48
AT  -7.48 68.13                74.31           886.97
AC  -2.52 52.71                73.42           447.21
GG  -7.25 58.49                72.85           494.45
TG   8.80 50.53                70.33           535.25
CT  21.70 42.34                68.66           504.98
AG -20.31 52.35                62.32           554.21
GT   9.17 45.51                61.79           421.48
GA -22.80 53.08                61.59           400.75
TT  13.61 51.23                58.04          1061.71
CC  -0.12 53.14                57.82           676.20
TA -18.31 54.24                53.68           599.27
GC  27.85 43.24                53.18           873.14
TC  15.60 34.67                52.01           416.18</code></pre>
<p>Plot feature importance.</p>
<pre class="r"><code>varImpPlot(my_rf)</code></pre>
<p><img src="figure/tss_predict.Rmd/importance_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-importance_plot-1">
Past versions of importance_plot-1.png
</button>
</p>
<div id="fig-importance_plot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/9ec0dec626f4d02266a3b53c5ae208519acea033/docs/figure/tss_predict.Rmd/importance_plot-1.png" target="_blank">9ec0dec</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/figure/tss_predict.Rmd/importance_plot-1.png" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Use our Random Forest classifier to make predictions on the test
data.</p>
<pre class="r"><code>data.predict &lt;- predict(my_rf, test)
prop.table(table(observed = test$class, predict = data.predict), 1)</code></pre>
<pre><code>        predict
observed    random       tss
  random 0.7365761 0.2634239
  tss    0.2675337 0.7324663</code></pre>
<p>Use <code>pROC</code> package to calculate the area under the ROC
curve.</p>
<pre class="r"><code>load_package(&quot;pROC&quot;, source = &quot;cran&quot;)</code></pre>
<p>Area under the ROC curve.</p>
<pre class="r"><code>roc(my_rf$y, my_rf$votes[, &#39;random&#39;])</code></pre>
<pre><code>Setting levels: control = random, case = tss</code></pre>
<pre><code>Setting direction: controls &gt; cases</code></pre>
<pre><code>
Call:
roc.default(response = my_rf$y, predictor = my_rf$votes[, &quot;random&quot;])

Data: my_rf$votes[, &quot;random&quot;] in 48263 controls (my_rf$y random) &gt; 50141 cases (my_rf$y tss).
Area under the curve: 0.7908</code></pre>
<p>Plot ROC with the <code>verification</code> package.</p>
<pre class="r"><code>load_package(&#39;verification&#39;, source = &quot;cran&quot;)</code></pre>
<p>Plot ROC.</p>
<pre class="r"><code>aucc &lt;- roc.area(as.integer(train$class==&#39;tss&#39;), my_rf$votes[,2])$A
roc.plot(as.integer(train$class==&#39;tss&#39;), my_rf$votes[,2], main=&quot;&quot;)</code></pre>
<pre><code>Warning in roc.plot.default(as.integer(train$class == &quot;tss&quot;), my_rf$votes[, :
Large amount of unique predictions used as thresholds. Consider specifying
thresholds.</code></pre>
<pre class="r"><code>legend(&quot;bottomright&quot;, bty=&quot;n&quot;, sprintf(&quot;Area Under the Curve (AUC) = %1.3f&quot;, aucc))
title(main=&quot;OOB ROC Curve Random Forest for predicting TSS&quot;)</code></pre>
<p><img src="figure/tss_predict.Rmd/plot_roc-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_roc-1">
Past versions of plot_roc-1.png
</button>
</p>
<div id="fig-plot_roc-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/9ec0dec626f4d02266a3b53c5ae208519acea033/docs/figure/tss_predict.Rmd/plot_roc-1.png" target="_blank">9ec0dec</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/figure/tss_predict.Rmd/plot_roc-1.png" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="nucleotide-frequency-across-tss" class="section level3">
<h3>Nucleotide frequency across TSS</h3>
<p>The di-nucleotide frequencies were useful in predicting TSSs but we
lost the positional information of each di-nucleotide and the relative
position of a di-nucleotide may be important.</p>
<p>To confirm this, we can use the <code>nucleotideFrequencyAt</code>
function to return the single nucleotide frequency at a specific
location.</p>
<pre class="r"><code>nucleotideFrequencyAt(x = my_tss$seq, at = 3)</code></pre>
<pre><code>    A     C     G     T 
26030  9426 24056  3162 </code></pre>
<p>We will use <code>sapply</code> to calculate the nucleotide
frequencies at all sites.</p>
<pre class="r"><code>nuc_freq &lt;- sapply(1:((params$span*2)+1), function(x) nucleotideFrequencyAt(my_tss$seq, x))

nuc_freq_norm &lt;- apply(nuc_freq, 2, function(x) x/sum(x))
colnames(nuc_freq_norm) &lt;- c(&#39;m2&#39;, &#39;m1&#39;, &#39;t&#39;, &#39;p1&#39;, &#39;p2&#39;)

nuc_freq_norm</code></pre>
<pre><code>         m2         m1          t        p1        p2
A 0.1252194 0.09062769 0.41532374 0.1573061 0.2356320
C 0.3273925 0.51676931 0.15039729 0.2624533 0.2344832
G 0.3261161 0.15159396 0.38382742 0.3345247 0.2694419
T 0.2212720 0.24100903 0.05045154 0.2457159 0.2604429</code></pre>
<p>There is a higher CT (pyrimidine) ratio one position before the TSS
(<code>m1</code> [minus 1] column) and a higher AG (purine) ratio at the
TSS (<code>t</code> column) forming the classic pyrimidine–purine (PyPu)
di-nucleotide at a TSS.</p>
<p>Given this positional bias, we will calculate di-nucleotide
frequencies at all positions using the
<code>dinucleotideFrequency</code> function.</p>
<pre class="r"><code>dinuc_freq_tss &lt;- lapply(1:(params$span*2), function(x){
  dinucleotideFrequency(subseq(my_tss$seq, x, x+1))
})

dinuc_freq_random &lt;- lapply(1:(params$span*2), function(x){
  dinucleotideFrequency(subseq(my_random$seq, x, x+1))
})

dinuc_freq_tss_df &lt;- do.call(cbind.data.frame, dinuc_freq_tss)
dinuc_freq_random_df &lt;- do.call(cbind.data.frame, dinuc_freq_random)

my_colname &lt;- paste0(
  &#39;X&#39;,
  rep(1:(params$span*2), each = 16),
  &quot;_&quot;,
  colnames(dinuc_freq_tss_df)[1:16]
)

colnames(dinuc_freq_tss_df) &lt;- my_colname
colnames(dinuc_freq_random_df) &lt;- my_colname

head(dinuc_freq_tss_df)</code></pre>
<pre><code>  X1_AA X1_AC X1_AG X1_AT X1_CA X1_CC X1_CG X1_CT X1_GA X1_GC X1_GG X1_GT X1_TA
1     0     0     0     0     0     0     0     0     0     1     0     0     0
2     0     0     0     0     0     0     0     0     0     0     0     0     0
3     0     0     0     0     0     0     0     0     0     0     1     0     0
4     0     0     0     0     1     0     0     0     0     0     0     0     0
5     0     0     0     0     0     0     0     0     0     0     0     0     0
6     0     0     0     0     0     0     0     0     0     0     0     0     0
  X1_TC X1_TG X1_TT X2_AA X2_AC X2_AG X2_AT X2_CA X2_CC X2_CG X2_CT X2_GA X2_GC
1     0     0     0     0     0     0     0     1     0     0     0     0     0
2     1     0     0     0     0     0     0     1     0     0     0     0     0
3     0     0     0     0     0     0     0     0     0     0     0     0     0
4     0     0     0     0     0     1     0     0     0     0     0     0     0
5     1     0     0     0     0     0     0     1     0     0     0     0     0
6     1     0     0     0     0     0     0     0     0     1     0     0     0
  X2_GG X2_GT X2_TA X2_TC X2_TG X2_TT X3_AA X3_AC X3_AG X3_AT X3_CA X3_CC X3_CG
1     0     0     0     0     0     0     0     1     0     0     0     0     0
2     0     0     0     0     0     0     0     0     0     1     0     0     0
3     1     0     0     0     0     0     0     0     0     0     0     0     0
4     0     0     0     0     0     0     0     0     0     0     0     0     0
5     0     0     0     0     0     0     0     1     0     0     0     0     0
6     0     0     0     0     0     0     0     0     0     0     0     0     0
  X3_CT X3_GA X3_GC X3_GG X3_GT X3_TA X3_TC X3_TG X3_TT X4_AA X4_AC X4_AG X4_AT
1     0     0     0     0     0     0     0     0     0     0     0     0     0
2     0     0     0     0     0     0     0     0     0     0     0     0     0
3     0     0     0     0     1     0     0     0     0     0     0     0     0
4     0     0     0     1     0     0     0     0     0     0     0     0     0
5     0     0     0     0     0     0     0     0     0     0     0     0     0
6     0     0     1     0     0     0     0     0     0     0     0     0     0
  X4_CA X4_CC X4_CG X4_CT X4_GA X4_GC X4_GG X4_GT X4_TA X4_TC X4_TG X4_TT
1     0     0     0     1     0     0     0     0     0     0     0     0
2     0     0     0     0     0     0     0     0     0     0     0     1
3     0     0     0     0     0     0     0     0     0     1     0     0
4     0     0     0     0     0     0     1     0     0     0     0     0
5     0     0     1     0     0     0     0     0     0     0     0     0
6     0     1     0     0     0     0     0     0     0     0     0     0</code></pre>
<p>Combine TSS and random data.</p>
<pre class="r"><code>dinuc_freq_tss_df %&gt;%
  mutate(class = &#39;tss&#39;) -&gt; my_data2

dinuc_freq_random_df %&gt;%
  mutate(class = &#39;random&#39;) %&gt;%
  rbind(my_data2) %&gt;%
  mutate(class = factor(class)) -&gt; my_data2

saveRDS(object = my_data2, file = &quot;data/tss.rds&quot;)

dim(my_data2)</code></pre>
<pre><code>[1] 123005     65</code></pre>
<p>Split 80/20.</p>
<pre class="r"><code>set.seed(1984)
idx &lt;- sample(nrow(my_data2), nrow(my_data2)*0.8)

train2 &lt;- my_data2[idx, ]
test2 &lt;- my_data2[-idx, ]

dim(train2)</code></pre>
<pre><code>[1] 98404    65</code></pre>
<pre class="r"><code>dim(test2)</code></pre>
<pre><code>[1] 24601    65</code></pre>
<p>Train Random Forest.</p>
<pre class="r"><code>my_rf2 &lt;- randomForest(
  class ~ .,
  data = train2,
  importance = TRUE,
  do.trace = 100,
  ntree = 500
)</code></pre>
<pre><code>ntree      OOB      1      2
  100:  23.61% 25.44% 21.85%
  200:  23.57% 25.23% 21.97%
  300:  23.57% 25.17% 22.03%
  400:  23.59% 25.20% 22.05%
  500:  23.59% 25.11% 22.12%</code></pre>
<pre class="r"><code>my_rf2</code></pre>
<pre><code>
Call:
 randomForest(formula = class ~ ., data = train2, importance = TRUE,      do.trace = 100, ntree = 500) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 8

        OOB estimate of  error rate: 23.59%
Confusion matrix:
       random   tss class.error
random  36143 12120   0.2511240
tss     11091 39050   0.2211962</code></pre>
<p>Feature importance.</p>
<pre class="r"><code>rn2 &lt;- round(importance(my_rf2), 2)
head(rn2[order(rn2[,3], decreasing=TRUE),])</code></pre>
<pre><code>      random   tss MeanDecreaseAccuracy MeanDecreaseGini
X4_CG  27.77 41.79                47.49           308.70
X3_CA  35.17 16.16                38.48           424.26
X1_AT  41.21 10.50                36.20           345.93
X2_CG  38.74 19.75                33.97          1943.38
X2_CA  29.40 20.47                33.35          1830.46
X4_TG   0.24 27.13                31.65           112.98</code></pre>
<p>Feature importance plot.</p>
<pre class="r"><code>varImpPlot(my_rf2)</code></pre>
<p><img src="figure/tss_predict.Rmd/importance_plot2-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-importance_plot2-1">
Past versions of importance_plot2-1.png
</button>
</p>
<div id="fig-importance_plot2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/9ec0dec626f4d02266a3b53c5ae208519acea033/docs/figure/tss_predict.Rmd/importance_plot2-1.png" target="_blank">9ec0dec</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Compare prediction results of our first model and our second
model.</p>
<pre class="r"><code># results from first model
data.predict &lt;- predict(my_rf, test, type = &quot;vote&quot;)
auroc &lt;- roc.area(as.integer(test$class==&#39;tss&#39;), data.predict[,2])$A
roc.plot(as.integer(test$class==&#39;tss&#39;), data.predict[,2], main=&quot;ROC&quot;)
legend(&quot;bottomright&quot;, bty=&quot;n&quot;, sprintf(&quot;Area Under the Curve (AUC) = %1.3f&quot;, auroc))</code></pre>
<p><img src="figure/tss_predict.Rmd/predict2-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># results from second model
data.predict2 &lt;- predict(my_rf2, test2, type = &quot;vote&quot;)
auroc2 &lt;- roc.area(as.integer(test2$class==&#39;tss&#39;), data.predict2[,2])$A
roc.plot(as.integer(test2$class==&#39;tss&#39;), data.predict2[,2], main=&quot;ROC&quot;)
legend(&quot;bottomright&quot;, bty=&quot;n&quot;, sprintf(&quot;Area Under the Curve (AUC) = %1.3f&quot;, auroc2))</code></pre>
<p><img src="figure/tss_predict.Rmd/predict2-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>By including the positional information of the di-nucleotides, we
have a better TSS classifier, which makes sense since not all the bases
upstream and downstream of the TSS have a nucleotide preference and thus
were diluting the signal in our first model where we calculated the
overall di-nucleotide frequency.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] verification_1.42                 dtw_1.23-1                       
 [3] proxy_0.4-27                      CircStats_0.2-6                  
 [5] MASS_7.3-58.1                     boot_1.3-28                      
 [7] fields_14.1                       viridis_0.6.2                    
 [9] viridisLite_0.4.1                 spam_2.9-1                       
[11] pROC_1.18.0                       randomForest_4.7-1.1             
[13] BSgenome.Hsapiens.UCSC.hg38_1.4.4 BSgenome_1.66.1                  
[15] rtracklayer_1.58.0                Biostrings_2.66.0                
[17] XVector_0.38.0                    GenomicRanges_1.50.1             
[19] GenomeInfoDb_1.34.2               IRanges_2.32.0                   
[21] S4Vectors_0.36.0                  BiocGenerics_0.44.0              
[23] biomaRt_2.54.0                    forcats_0.5.2                    
[25] stringr_1.4.1                     dplyr_1.0.10                     
[27] purrr_0.3.5                       readr_2.1.3                      
[29] tidyr_1.2.1                       tibble_3.1.8                     
[31] ggplot2_3.4.0                     tidyverse_1.3.2                  
[33] workflowr_1.7.0                  

loaded via a namespace (and not attached):
  [1] readxl_1.4.1                backports_1.4.1            
  [3] BiocFileCache_2.6.0         plyr_1.8.7                 
  [5] BiocParallel_1.32.0         digest_0.6.30              
  [7] htmltools_0.5.3             fansi_1.0.3                
  [9] magrittr_2.0.3              memoise_2.0.1              
 [11] googlesheets4_1.0.1         tzdb_0.3.0                 
 [13] modelr_0.1.9                matrixStats_0.62.0         
 [15] timechange_0.1.1            prettyunits_1.1.1          
 [17] colorspace_2.0-3            blob_1.2.3                 
 [19] rvest_1.0.3                 rappdirs_0.3.3             
 [21] haven_2.5.1                 xfun_0.34                  
 [23] callr_3.7.3                 crayon_1.5.2               
 [25] RCurl_1.98-1.9              jsonlite_1.8.3             
 [27] glue_1.6.2                  gtable_0.3.1               
 [29] gargle_1.2.1                zlibbioc_1.44.0            
 [31] DelayedArray_0.24.0         maps_3.4.1                 
 [33] scales_1.2.1                DBI_1.1.3                  
 [35] Rcpp_1.0.9                  progress_1.2.2             
 [37] bit_4.0.4                   dotCall64_1.0-2            
 [39] httr_1.4.4                  ellipsis_0.3.2             
 [41] pkgconfig_2.0.3             XML_3.99-0.12              
 [43] farver_2.1.1                sass_0.4.2                 
 [45] dbplyr_2.2.1                utf8_1.2.2                 
 [47] tidyselect_1.2.0            labeling_0.4.2             
 [49] rlang_1.0.6                 later_1.3.0                
 [51] AnnotationDbi_1.60.0        munsell_0.5.0              
 [53] cellranger_1.1.0            tools_4.2.0                
 [55] cachem_1.0.6                cli_3.4.1                  
 [57] generics_0.1.3              RSQLite_2.2.18             
 [59] broom_1.0.1                 evaluate_0.17              
 [61] fastmap_1.1.0               yaml_2.3.6                 
 [63] processx_3.8.0              knitr_1.40                 
 [65] bit64_4.0.5                 fs_1.5.2                   
 [67] KEGGREST_1.38.0             whisker_0.4                
 [69] xml2_1.3.3                  compiler_4.2.0             
 [71] rstudioapi_0.14             filelock_1.0.2             
 [73] curl_4.3.3                  png_0.1-7                  
 [75] reprex_2.0.2                bslib_0.4.1                
 [77] stringi_1.7.8               highr_0.9                  
 [79] ps_1.7.2                    lattice_0.20-45            
 [81] Matrix_1.5-1                vctrs_0.5.0                
 [83] pillar_1.8.1                lifecycle_1.0.3            
 [85] jquerylib_0.1.4             bitops_1.0-7               
 [87] httpuv_1.6.6                R6_2.5.1                   
 [89] BiocIO_1.8.0                promises_1.2.0.1           
 [91] gridExtra_2.3               codetools_0.2-18           
 [93] assertthat_0.2.1            SummarizedExperiment_1.28.0
 [95] rprojroot_2.0.3             rjson_0.2.21               
 [97] withr_2.5.0                 GenomicAlignments_1.34.0   
 [99] Rsamtools_2.14.0            GenomeInfoDbData_1.2.9     
[101] parallel_4.2.0              hms_1.1.2                  
[103] grid_4.2.0                  rmarkdown_2.17             
[105] MatrixGenerics_1.10.0       googledrive_2.0.0          
[107] git2r_0.30.1                getPass_0.2-2              
[109] Biobase_2.58.0              lubridate_1.9.0            
[111] restfulr_0.0.15            </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
