<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2022-04-26" />

<title>Sequence composition and Random Forests</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Sequence composition and Random Forests</h1>
<h4 class="date">2022-04-26</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-04-26
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree56b1a796a52ee226a819290f1a36cf591ed9fd60targetblank56b1a79a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/davetang/muse/tree/56b1a796a52ee226a819290f1a36cf591ed9fd60" target="_blank">56b1a79</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree56b1a796a52ee226a819290f1a36cf591ed9fd60targetblank56b1a79a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/davetang/muse/tree/56b1a796a52ee226a819290f1a36cf591ed9fd60" target="_blank">56b1a79</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.1.2/

Untracked files:
    Untracked:  analysis/cell_ranger.Rmd

Unstaged changes:
    Modified:   .Rprofile

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/tss_predict.Rmd</code>) and HTML (<code>docs/tss_predict.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/56b1a796a52ee226a819290f1a36cf591ed9fd60/analysis/tss_predict.Rmd" target="_blank">56b1a79</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-26
</td>
<td>
Positional di-nuc freq
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/tss_predict.html" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/06164641bd8f7307763af70a208dad096f524607/analysis/tss_predict.Rmd" target="_blank">0616464</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
<td>
Random forest model
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/fd41c759545ae9429df95eaf79a7ddca9447d5c9/docs/tss_predict.html" target="_blank">fd41c75</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/e372c815ee309a4b014e441268a2ec5416f8ced8/analysis/tss_predict.Rmd" target="_blank">e372c81</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-22
</td>
<td>
TSS
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Transcriptional starting sites (TSSs) demarcate the first position in the DNA sequence that gets transcribed into RNA. In this work, we will try to build a classifier to try to predict TSSs. We will train our classifier using the NCBI Reference Sequence Database also known as RefSeq.</p>
</div>
<div id="packages" class="section level2">
<h2>Packages</h2>
<p>Function for loading required Bioconductor packages (and install them first, if missing).</p>
<pre class="r"><code>load_package &lt;- function(x, source = &quot;bioc&quot;){
  if(!require(x, character.only = TRUE, quietly = TRUE)){
    if (source == &quot;bioc&quot;){
      BiocManager::install(x, character.only = TRUE)
    } else if (source == &quot;cran&quot;){
      install.packages(x, character.only = TRUE)
    } else {
      stop(&quot;Unrecognised source&quot;)
    }
    library(x, character.only = TRUE)
  }
}</code></pre>
</div>
<div id="refseq" class="section level2">
<h2>RefSeq</h2>
<p>We will use the Bioconductor package <code>biomaRt</code> to download the entire collection of RefSeq sequences.</p>
<pre class="r"><code>load_package(&#39;biomaRt&#39;)</code></pre>
<p>Find the human gene set.</p>
<pre class="r"><code>ensembl &lt;- useMart(&#39;ensembl&#39;, host = params$host)
biomaRt::listDatasets(ensembl) %&gt;%
  filter(grepl(&#39;human&#39;, description, TRUE))</code></pre>
<pre><code>                dataset              description    version
1 hsapiens_gene_ensembl Human genes (GRCh38.p13) GRCh38.p13</code></pre>
<p>Find RefSeq attributes.</p>
<pre class="r"><code>ensembl &lt;- useMart(&#39;ensembl&#39;, dataset = &#39;hsapiens_gene_ensembl&#39;, host = params$host)
listAttributes(ensembl) %&gt;%
  filter(grepl(&#39;refseq&#39;, description, TRUE))</code></pre>
<pre><code>                           name                                  description
1        transcript_mane_select        RefSeq match transcript (MANE Select)
2 transcript_mane_plus_clinical RefSeq match transcript (MANE Plus Clinical)
3                   refseq_mrna                               RefSeq mRNA ID
4         refseq_mrna_predicted                     RefSeq mRNA predicted ID
5                  refseq_ncrna                              RefSeq ncRNA ID
6        refseq_ncrna_predicted                    RefSeq ncRNA predicted ID
7                refseq_peptide                            RefSeq peptide ID
8      refseq_peptide_predicted                  RefSeq peptide predicted ID
          page
1 feature_page
2 feature_page
3 feature_page
4 feature_page
5 feature_page
6 feature_page
7 feature_page
8 feature_page</code></pre>
<p>Fetch all RefSeq mRNAs on assembled chromosomes.</p>
<pre class="r"><code>my_chr &lt;- c(1:22, &#39;X&#39;, &#39;Y&#39;)

my_refseq &lt;- getBM(
  attributes=&#39;refseq_mrna&#39;,
  filters = &#39;chromosome_name&#39;,
  values = my_chr,
  mart = ensembl
)</code></pre>
<p>Number of RefSeq IDs.</p>
<pre class="r"><code>dim(my_refseq)</code></pre>
<pre><code>[1] 61953     1</code></pre>
<p>Build table. Note that when working with transcripts, use the attributes <code>transcript_start</code> and <code>transcript_end</code>, and not the attributes <code>start_position</code> and <code>end_position</code>; using those will give you the Ensembl gene coordinates and not the RefSeq coordinates, which is what we want! (I have retrieved the Ensembl coordinates for illustrative purposes.)</p>
<pre class="r"><code>my_att &lt;- c(
  &#39;refseq_mrna&#39;,
  &#39;chromosome_name&#39;,
  &#39;transcript_start&#39;,
  &#39;transcript_end&#39;,
  &#39;start_position&#39;,
  &#39;end_position&#39;,
  &#39;strand&#39;
)

my_refseq_loci &lt;- getBM(
  attributes = my_att,
  filters = c(&#39;refseq_mrna&#39;, &#39;chromosome_name&#39;),
  values = list(refseq_mrna = my_refseq$refseq_mrna, chromosome_name = my_chr),
  mart = ensembl
)</code></pre>
<p>Check out the table.</p>
<pre class="r"><code>head(my_refseq_loci)</code></pre>
<pre><code>  refseq_mrna chromosome_name transcript_start transcript_end start_position
1   NM_000015               8         18391282       18401218       18391282
2   NM_000021              14         73136507       73223691       73136418
3   NM_000053              13         51932669       52011450       51930436
4   NM_000059              13         32315508       32400268       32315086
5   NM_000071              21         43053191       43075835       43053191
6   NM_000084               X         50067576       50092406       49922596
  end_position strand
1     18401218      1
2     73223691      1
3     52012125     -1
4     32400268      1
5     43076943     -1
6     50099235      1</code></pre>
<p>Check for duplicated entries.</p>
<pre class="r"><code>table(duplicated(my_refseq_loci$refseq_mrna))</code></pre>
<pre><code>
FALSE  TRUE 
61953    25 </code></pre>
<p>Removed duplicated entries.</p>
<pre class="r"><code>my_refseq_loci_uniq &lt;- my_refseq_loci[!duplicated(my_refseq_loci$refseq_mrna),]
dim(my_refseq_loci_uniq)</code></pre>
<pre><code>[1] 61953     7</code></pre>
<p>We will modify <code>chromosome_name</code> and <code>strand</code> to make it compatible with <code>BSgenome.Hsapiens.UCSC.hg38</code>.</p>
<pre class="r"><code>my_refseq_loci_uniq %&gt;%
  mutate(strand = ifelse(strand == 1, yes = &#39;+&#39;, no = &#39;-&#39;)) %&gt;%
  mutate(chromosome_name = sub(&quot;^&quot;, &quot;chr&quot;, chromosome_name)) -&gt; my_refseq_loci_uniq

head(my_refseq_loci_uniq)</code></pre>
<pre><code>  refseq_mrna chromosome_name transcript_start transcript_end start_position
1   NM_000015            chr8         18391282       18401218       18391282
2   NM_000021           chr14         73136507       73223691       73136418
3   NM_000053           chr13         51932669       52011450       51930436
4   NM_000059           chr13         32315508       32400268       32315086
5   NM_000071           chr21         43053191       43075835       43053191
6   NM_000084            chrX         50067576       50092406       49922596
  end_position strand
1     18401218      +
2     73223691      +
3     52012125      -
4     32400268      +
5     43076943      -
6     50099235      +</code></pre>
</div>
<div id="transcription-start-site" class="section level2">
<h2>Transcription Start Site</h2>
<p>The TSS is <code>transcript_start</code> when the transcript is on the <code>+</code> strand and is <code>transcript_end</code> on the <code>-</code> strand. We want to retrieve sequence upstream and downstream of the TSS, so we will subtract and add coordinates accordingly.</p>
<pre class="r"><code>my_tss &lt;- list()

my_tss$loci &lt;- my_refseq_loci_uniq %&gt;%
  dplyr::select(-c(start_position, end_position)) %&gt;%
  mutate(tss_start = if_else(strand == &quot;+&quot;, transcript_start - params$span, transcript_end - params$span)) %&gt;%
  mutate(tss_end = if_else(strand == &quot;+&quot;, transcript_start + params$span, transcript_end + params$span))

head(my_tss$loci)</code></pre>
<pre><code>  refseq_mrna chromosome_name transcript_start transcript_end strand tss_start
1   NM_000015            chr8         18391282       18401218      +  18391280
2   NM_000021           chr14         73136507       73223691      +  73136505
3   NM_000053           chr13         51932669       52011450      -  52011448
4   NM_000059           chr13         32315508       32400268      +  32315506
5   NM_000071           chr21         43053191       43075835      -  43075833
6   NM_000084            chrX         50067576       50092406      +  50067574
   tss_end
1 18391284
2 73136509
3 52011452
4 32315510
5 43075837
6 50067578</code></pre>
</div>
<div id="random-loci" class="section level2">
<h2>Random loci</h2>
<p>Sample a bunch of random sequences in hg38.</p>
<pre class="r"><code>load_package(&#39;BSgenome.Hsapiens.UCSC.hg38&#39;)</code></pre>
<p>Save chromosomes and their sizes.</p>
<pre class="r"><code>hg38_ref &lt;- seqnames(BSgenome.Hsapiens.UCSC.hg38)
hg38_chr &lt;- hg38_ref[!grepl(&quot;_&quot;, hg38_ref)]
hg38_chr &lt;- hg38_chr[!grepl(&quot;chrM&quot;, hg38_chr)]

hg38_chr_size &lt;- sapply(hg38_chr, function(x){
  length(BSgenome.Hsapiens.UCSC.hg38[[x]])
})

head(hg38_chr_size)</code></pre>
<pre><code>     chr1      chr2      chr3      chr4      chr5      chr6 
248956422 242193529 198295559 190214555 181538259 170805979 </code></pre>
<p>Sample chromosome based on length of chromosome.</p>
<pre class="r"><code>set.seed(1984)
n &lt;- nrow(my_tss$loci)
schr &lt;- sample(
  hg38_chr,
  n,
  TRUE,
  prob = hg38_chr_size/sum(hg38_chr_size)
)

sstrand &lt;- sample(c(&quot;+&quot;, &quot;-&quot;), n, TRUE)
sstart &lt;- sapply(schr, function(x){
  sample(hg38_chr_size[x] - (params$span * 2), 1)
})
send &lt;- sstart + (params$span*2)

my_random &lt;- list()

my_random$loci &lt;- data.frame(
  chr = schr,
  start = sstart,
  end = send,
  strand = sstrand
)

head(my_random$loci)</code></pre>
<pre><code>    chr     start       end strand
1 chr10 132635027 132635031      -
2  chr7 115323451 115323455      -
3  chr6  27058086  27058090      -
4  chr5 164216413 164216417      +
5 chr13   1446812   1446816      +
6 chr17  52786687  52786691      -</code></pre>
</div>
<div id="removing-overlaps" class="section level2">
<h2>Removing overlaps</h2>
<p>Since some of the randomly sampled regions may overlap/intersect with TSSs, I used the GenomicRanges package to remove these overlapping regions:</p>
<pre class="r"><code>load_package(&quot;GenomicRanges&quot;)</code></pre>
<p>Create a GRanges object given an object, my_refseq_loci</p>
<pre class="r"><code>my_tss$gr &lt;- with(
  my_tss$loci,
  GRanges(chromosome_name,
          IRanges(tss_start, tss_end, names = refseq_mrna),
          strand
  )
)

head(my_tss$gr)</code></pre>
<pre><code>GRanges object with 6 ranges and 0 metadata columns:
            seqnames            ranges strand
               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
  NM_000015     chr8 18391280-18391284      +
  NM_000021    chr14 73136505-73136509      +
  NM_000053    chr13 52011448-52011452      -
  NM_000059    chr13 32315506-32315510      +
  NM_000071    chr21 43075833-43075837      -
  NM_000084     chrX 50067574-50067578      +
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Create random GRanges object.</p>
<pre class="r"><code>my_random$gr &lt;- with(
  my_random$loci,
  GRanges(chr,
          IRanges(start, end, names = paste(chr, start, end, sep = &quot;_&quot;)),
          strand
  )
)

head(my_random$gr)</code></pre>
<pre><code>GRanges object with 6 ranges and 0 metadata columns:
                            seqnames              ranges strand
                               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
  chr10_132635027_132635031    chr10 132635027-132635031      -
   chr7_115323451_115323455     chr7 115323451-115323455      -
     chr6_27058086_27058090     chr6   27058086-27058090      -
   chr5_164216413_164216417     chr5 164216413-164216417      +
      chr13_1446812_1446816    chr13     1446812-1446816      +
    chr17_52786687_52786691    chr17   52786687-52786691      -
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Overlap between random regions and TSS’s. Note that there are cases where a random region overlaps more than one TSS, which indicates that TSS regions are overlapping.</p>
<pre class="r"><code>table(countOverlaps(my_random$gr, my_tss$gr))</code></pre>
<pre><code>
    0     1     4     5 
61947     4     1     1 </code></pre>
<p>Remove overlapping random sites.</p>
<pre class="r"><code>my_random$loci_no &lt;- my_random$loci[countOverlaps(my_random$gr, my_tss$gr) == 0, ]
dim(my_random$loci_no)</code></pre>
<pre><code>[1] 61947     4</code></pre>
<p>Check overlapping TSS regions.</p>
<pre class="r"><code>table(countOverlaps(my_tss$gr, my_tss$gr, minoverlap = 1))</code></pre>
<pre><code>
    1     2     3     4     5     6     7     8     9    10    11    12    13 
25851 11455  5969  3918  2691  1911  1455  1051  1015   834   663   557   475 
   14    15    16    17    18    19    20    21    22    23    24    25    26 
  430   258   363   394   280   258   220   209   155    92    96    75    96 
   27    28    29    30    31    32    33    35    36    38    43    44    45 
   62   168    87    30    62    64    66   105   108    76    43    44    45 
   46    50    54    72 
   46    50    54    72 </code></pre>
<p>Create a set of non-overlapping TSS regions (the conditional is <code>== 1</code> because we are comparing regions against itself and 1 is the self overlap), where a TSS does not overlap by a single bp.</p>
<pre class="r"><code>my_tss$loci_no &lt;- my_tss$loci[countOverlaps(my_tss$gr, my_tss$gr, minoverlap = 1) == 1, ]
dim(my_tss$loci_no)</code></pre>
<pre><code>[1] 25851     7</code></pre>
</div>
<div id="sequence" class="section level2">
<h2>Sequence</h2>
<p>Now to fetch the sequences and calculate the dinucleotide frequencies:</p>
<pre class="r"><code>my_random$seq &lt;- getSeq(
  BSgenome.Hsapiens.UCSC.hg38,
  names = my_random$loci$chr,
  start = my_random$loci$start,
  end = my_random$loci$end,
  strand = my_random$loci$strand
)

head(my_random$seq)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 CATCA                                             chr10
[2]     5 CTAGT                                             chr7
[3]     5 AGAAT                                             chr6
[4]     5 AATGT                                             chr5
[5]     5 NNNNN                                             chr13
[6]     5 TTACT                                             chr17</code></pre>
<p>Remove entries with N’s.</p>
<pre class="r"><code>my_random$seq &lt;- my_random$seq[! grepl(&quot;N&quot;, my_random$seq)]
head(my_random$seq)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 CATCA                                             chr10
[2]     5 CTAGT                                             chr7
[3]     5 AGAAT                                             chr6
[4]     5 AATGT                                             chr5
[5]     5 TTACT                                             chr17
[6]     5 CCTCC                                             chr1</code></pre>
<p>Obtain TSS sequences.</p>
<pre class="r"><code>my_tss$seq &lt;- getSeq(
  BSgenome.Hsapiens.UCSC.hg38,
  names = my_tss$loci$chromosome_name,
  start = my_tss$loci$tss_start,
  end = my_tss$loci$tss_end,
  strand = my_tss$loci$strand
)

my_tss$seq &lt;- my_tss$seq[! grepl(&quot;N&quot;, my_tss$seq)]
head(my_tss$seq)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 GCACT                                             chr8
[2]     5 TCGGA                                             chr14
[3]     5 CCCTC                                             chr13
[4]     5 GCAGA                                             chr13
[5]     5 GGGTC                                             chr21
[6]     5 CAGGG                                             chrX</code></pre>
<p>Obtain TSS sequences non-overlapping.</p>
<pre class="r"><code>my_tss$seq_no &lt;- getSeq(
  BSgenome.Hsapiens.UCSC.hg38,
  names = my_tss$loci_no$chromosome_name,
  start = my_tss$loci_no$tss_start,
  end = my_tss$loci_no$tss_end,
  strand = my_tss$loci_no$strand
)

my_tss$seq_no &lt;- my_tss$seq_no[! grepl(&quot;N&quot;, my_tss$seq_no)]
head(my_tss$seq_no)</code></pre>
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]     5 GCACT                                             chr8
[2]     5 TCGGA                                             chr14
[3]     5 CCCTC                                             chr13
[4]     5 GCAGA                                             chr13
[5]     5 CAGGG                                             chrX
[6]     5 TCACG                                             chr20</code></pre>
<p>Calculate the di-nucleotide frequency.</p>
<pre class="r"><code>my_random$di &lt;- dinucleotideFrequency(my_random$seq)
colSums(my_random$di)</code></pre>
<pre><code>   AA    AC    AG    AT    CA    CC    CG    CT    GA    GC    GG    GT    TA 
23038 11832 16208 18247 16874 12151  2429 16470 13965 10067 12230 12207 15480 
   TC    TG    TT 
14141 17269 23132 </code></pre>
<pre class="r"><code>my_tss$di &lt;- dinucleotideFrequency(my_tss$seq)
colSums(my_tss$di)</code></pre>
<pre><code>   AA    AC    AG    AT    CA    CC    CG    CT    GA    GC    GG    GT    TA 
 8678 11634 20201  8509 24443 21320 16226 15711 15760 24256 19408 14537  6959 
   TC    TG    TT 
14903 14539 10728 </code></pre>
<pre class="r"><code>my_tss$di_no &lt;- dinucleotideFrequency(my_tss$seq_no)
colSums(my_tss$di_no)</code></pre>
<pre><code>   AA    AC    AG    AT    CA    CC    CG    CT    GA    GC    GG    GT    TA 
 3721  4956  8551  3729 10443  8911  6271  6625  6646  9977  8019  5738  2861 
   TC    TG    TT 
 6439  6118  4399 </code></pre>
<p>Plot di-nucleotide frequencies between TSS and random regions.</p>
<pre class="r"><code>my_df &lt;- data.frame(
  x = colMeans(my_tss$di),
  y = colMeans(my_random$di),
  di = colnames(my_tss$di)
)

ggplot(my_df, aes(x, y, label = di)) +
  geom_point() +
  geom_text(nudge_x = 0.005) +
  geom_abline(slope = 1, lty = 3) +
  labs(x = &quot;TSS&quot;, y = &quot;Random&quot;) +
  theme_bw()</code></pre>
<p><img src="figure/tss_predict.Rmd/plot_di-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_di-1">
Past versions of plot_di-1.png
</button>
</p>
<div id="fig-plot_di-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/figure/tss_predict.Rmd/plot_di-1.png" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot di-nucleotide frequencies between overlapping and non-overlapping TSS regions.</p>
<pre class="r"><code>my_df &lt;- data.frame(
  x = colMeans(my_tss$di),
  y = colMeans(my_tss$di_no),
  di = colnames(my_tss$di)
)

ggplot(my_df, aes(x, y, label = di)) +
  geom_point() +
  geom_text(nudge_x = 0.005) +
  geom_abline(slope = 1, lty = 3) +
  labs(x = &quot;TSS&quot;, y = &quot;TSS no overlap&quot;) +
  theme_bw()</code></pre>
<p><img src="figure/tss_predict.Rmd/plot_di_tss-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="random-forests" class="section level2">
<h2>Random Forests</h2>
<p>Use Random Forests to train a predictor using the TSSs and random di-nucleotide frequencies.</p>
<pre class="r"><code>load_package(&#39;randomForest&#39;, source = &quot;cran&quot;)</code></pre>
<p>Combine TSS and random data.</p>
<pre class="r"><code>as.data.frame(my_tss$di) %&gt;%
  mutate(class = &#39;tss&#39;) -&gt; my_data

as.data.frame(my_random$di) %&gt;%
  mutate(class = &#39;random&#39;) %&gt;%
  rbind(my_data) %&gt;%
  mutate(class = factor(class)) -&gt; my_data

dim(my_data)</code></pre>
<pre><code>[1] 120888     17</code></pre>
<p>Split 80/20.</p>
<pre class="r"><code>set.seed(1984)
idx &lt;- sample(nrow(my_data), nrow(my_data)*0.8)

train &lt;- my_data[idx, ]
test &lt;- my_data[-idx, ]

dim(train)</code></pre>
<pre><code>[1] 96710    17</code></pre>
<pre class="r"><code>dim(test)</code></pre>
<pre><code>[1] 24178    17</code></pre>
<p>Train Random Forest.</p>
<pre class="r"><code>my_rf &lt;- randomForest(
  class ~ .,
  data = train,
  importance = TRUE,
  do.trace = 100,
  ntree = 200
)</code></pre>
<pre><code>ntree      OOB      1      2
  100:  26.89% 26.77% 27.00%
  200:  26.85% 26.47% 27.22%</code></pre>
<pre class="r"><code>my_rf</code></pre>
<pre><code>
Call:
 randomForest(formula = class ~ ., data = train, importance = TRUE,      do.trace = 100, ntree = 200) 
               Type of random forest: classification
                     Number of trees: 200
No. of variables tried at each split: 4

        OOB estimate of  error rate: 26.85%
Confusion matrix:
       random   tss class.error
random  34622 12462   0.2646759
tss     13508 36118   0.2721960</code></pre>
<p>Feature importance.</p>
<pre class="r"><code>rn &lt;- round(importance(my_rf), 2)
rn[order(rn[,3], decreasing=TRUE),]</code></pre>
<pre><code>   random   tss MeanDecreaseAccuracy MeanDecreaseGini
AA  21.74 40.17                52.33          1505.47
AT  -1.87 44.72                48.11           939.64
GG  -3.66 35.79                47.02           478.09
CA -26.74 45.55                46.95           858.20
CG  34.59 34.79                46.47          2599.94
CT  15.54 27.60                45.84           488.97
AC  -0.84 34.24                45.43           442.64
GT   8.18 29.41                43.86           418.54
TG   5.15 32.84                39.78           545.91
AG -13.89 34.59                39.47           539.19
TT  10.58 31.98                39.28          1103.73
GA -15.40 36.52                38.76           394.84
TA -10.21 32.41                35.69           623.41
CC   2.66 33.37                34.07           699.46
TC  12.13 21.00                32.92           428.78
GC  18.90 25.14                31.12          1014.36</code></pre>
<p>As a plot.</p>
<pre class="r"><code>varImpPlot(my_rf)</code></pre>
<p><img src="figure/tss_predict.Rmd/importance_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-importance_plot-1">
Past versions of importance_plot-1.png
</button>
</p>
<div id="fig-importance_plot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/figure/tss_predict.Rmd/importance_plot-1.png" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Predict.</p>
<pre class="r"><code>data.predict &lt;- predict(my_rf, test)
prop.table(table(observed = test$class, predict = data.predict), 1)</code></pre>
<pre><code>        predict
observed    random       tss
  random 0.7326808 0.2673192
  tss    0.2746005 0.7253995</code></pre>
<p>Use <code>pROC</code> package to get area under curve.</p>
<pre class="r"><code>load_package(&quot;pROC&quot;, source = &quot;cran&quot;)</code></pre>
<p>Check ROC.</p>
<pre class="r"><code>roc(my_rf$y, my_rf$votes[, &#39;random&#39;])</code></pre>
<pre><code>Setting levels: control = random, case = tss</code></pre>
<pre><code>Setting direction: controls &gt; cases</code></pre>
<pre><code>
Call:
roc.default(response = my_rf$y, predictor = my_rf$votes[, &quot;random&quot;])

Data: my_rf$votes[, &quot;random&quot;] in 47084 controls (my_rf$y random) &gt; 49626 cases (my_rf$y tss).
Area under the curve: 0.7945</code></pre>
<p>Plot ROC with the <code>verification</code> package.</p>
<pre class="r"><code>load_package(&#39;verification&#39;, source = &quot;cran&quot;)</code></pre>
<p>Plot ROC.</p>
<pre class="r"><code>aucc &lt;- roc.area(as.integer(train$class==&#39;tss&#39;), my_rf$votes[,2])$A
roc.plot(as.integer(train$class==&#39;tss&#39;), my_rf$votes[,2], main=&quot;&quot;)</code></pre>
<pre><code>Warning in roc.plot.default(as.integer(train$class == &quot;tss&quot;), my_rf$votes[, :
Large amount of unique predictions used as thresholds. Consider specifying
thresholds.</code></pre>
<pre class="r"><code>legend(&quot;bottomright&quot;, bty=&quot;n&quot;, sprintf(&quot;Area Under the Curve (AUC) = %1.3f&quot;, aucc))
title(main=&quot;OOB ROC Curve Random Forest for predicting TSS&quot;)</code></pre>
<p><img src="figure/tss_predict.Rmd/plot_roc-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_roc-1">
Past versions of plot_roc-1.png
</button>
</p>
<div id="fig-plot_roc-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/fd116a1c7914d28d8b119ee53618de541ff44d1c/docs/figure/tss_predict.Rmd/plot_roc-1.png" target="_blank">fd116a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-04-24
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="nucleotide-frequency-across-tss" class="section level3">
<h3>Nucleotide frequency across TSS</h3>
<p>The <code>nucleotideFrequencyAt</code> function will return the single nucleotide frequency at a specific location.</p>
<pre class="r"><code>nucleotideFrequencyAt(my_tss$seq, 3)</code></pre>
<pre><code>    A     C     G     T 
25775  9399 23652  3127 </code></pre>
<p>Use <code>sapply</code> to calculate at all sites.</p>
<p>There is a higher CT (pyrimidine) ratio one position before the TSS (second column) and a higher AG (purine) ratio at the TSS (third column) forming the classic pyrimidine–purine (PyPu) di-nucleotide at a TSS.</p>
<pre class="r"><code>nuc_freq &lt;- sapply(1:((params$span*2)+1), function(x) nucleotideFrequencyAt(my_tss$seq, x))
apply(nuc_freq, 2, function(x) x/sum(x))</code></pre>
<pre><code>       [,1]       [,2]       [,3]      [,4]      [,5]
A 0.1251755 0.09239262 0.41604119 0.1576679 0.2352267
C 0.3242781 0.51500331 0.15171178 0.2631834 0.2340968
G 0.3270221 0.15193776 0.38177328 0.3330912 0.2691234
T 0.2235243 0.24066631 0.05047375 0.2460575 0.2615531</code></pre>
<p>Calculate di-nucleotide frequencies at all sites.</p>
<pre class="r"><code>dinuc_freq_tss &lt;- lapply(1:(params$span*2), function(x){
  dinucleotideFrequency(subseq(my_tss$seq, x, x+1))
})

dinuc_freq_random &lt;- lapply(1:(params$span*2), function(x){
  dinucleotideFrequency(subseq(my_random$seq, x, x+1))
})

dinuc_freq_tss_df &lt;- do.call(cbind.data.frame, dinuc_freq_tss)
dinuc_freq_random_df &lt;- do.call(cbind.data.frame, dinuc_freq_random)

my_colname &lt;- paste0(
  &#39;X&#39;,
  rep(1:(params$span*2), each = 16),
  &quot;_&quot;,
  colnames(dinuc_freq_tss_df)[1:16]
)

colnames(dinuc_freq_tss_df) &lt;- my_colname
colnames(dinuc_freq_random_df) &lt;- my_colname

head(dinuc_freq_tss_df)</code></pre>
<pre><code>  X1_AA X1_AC X1_AG X1_AT X1_CA X1_CC X1_CG X1_CT X1_GA X1_GC X1_GG X1_GT X1_TA
1     0     0     0     0     0     0     0     0     0     1     0     0     0
2     0     0     0     0     0     0     0     0     0     0     0     0     0
3     0     0     0     0     0     1     0     0     0     0     0     0     0
4     0     0     0     0     0     0     0     0     0     1     0     0     0
5     0     0     0     0     0     0     0     0     0     0     1     0     0
6     0     0     0     0     1     0     0     0     0     0     0     0     0
  X1_TC X1_TG X1_TT X2_AA X2_AC X2_AG X2_AT X2_CA X2_CC X2_CG X2_CT X2_GA X2_GC
1     0     0     0     0     0     0     0     1     0     0     0     0     0
2     1     0     0     0     0     0     0     0     0     1     0     0     0
3     0     0     0     0     0     0     0     0     1     0     0     0     0
4     0     0     0     0     0     0     0     1     0     0     0     0     0
5     0     0     0     0     0     0     0     0     0     0     0     0     0
6     0     0     0     0     0     1     0     0     0     0     0     0     0
  X2_GG X2_GT X2_TA X2_TC X2_TG X2_TT X3_AA X3_AC X3_AG X3_AT X3_CA X3_CC X3_CG
1     0     0     0     0     0     0     0     1     0     0     0     0     0
2     0     0     0     0     0     0     0     0     0     0     0     0     0
3     0     0     0     0     0     0     0     0     0     0     0     0     0
4     0     0     0     0     0     0     0     0     1     0     0     0     0
5     1     0     0     0     0     0     0     0     0     0     0     0     0
6     0     0     0     0     0     0     0     0     0     0     0     0     0
  X3_CT X3_GA X3_GC X3_GG X3_GT X3_TA X3_TC X3_TG X3_TT X4_AA X4_AC X4_AG X4_AT
1     0     0     0     0     0     0     0     0     0     0     0     0     0
2     0     0     0     1     0     0     0     0     0     0     0     0     0
3     1     0     0     0     0     0     0     0     0     0     0     0     0
4     0     0     0     0     0     0     0     0     0     0     0     0     0
5     0     0     0     0     1     0     0     0     0     0     0     0     0
6     0     0     0     1     0     0     0     0     0     0     0     0     0
  X4_CA X4_CC X4_CG X4_CT X4_GA X4_GC X4_GG X4_GT X4_TA X4_TC X4_TG X4_TT
1     0     0     0     1     0     0     0     0     0     0     0     0
2     0     0     0     0     1     0     0     0     0     0     0     0
3     0     0     0     0     0     0     0     0     0     1     0     0
4     0     0     0     0     1     0     0     0     0     0     0     0
5     0     0     0     0     0     0     0     0     0     1     0     0
6     0     0     0     0     0     0     1     0     0     0     0     0</code></pre>
<p>Combine TSS and random data.</p>
<pre class="r"><code>dinuc_freq_tss_df %&gt;%
  mutate(class = &#39;tss&#39;) -&gt; my_data2

dinuc_freq_random_df %&gt;%
  mutate(class = &#39;random&#39;) %&gt;%
  rbind(my_data2) %&gt;%
  mutate(class = factor(class)) -&gt; my_data2

dim(my_data2)</code></pre>
<pre><code>[1] 120888     65</code></pre>
<p>Split 80/20.</p>
<pre class="r"><code>set.seed(1984)
idx &lt;- sample(nrow(my_data2), nrow(my_data2)*0.8)

train2 &lt;- my_data2[idx, ]
test2 &lt;- my_data2[-idx, ]

dim(train2)</code></pre>
<pre><code>[1] 96710    65</code></pre>
<pre class="r"><code>dim(test2)</code></pre>
<pre><code>[1] 24178    65</code></pre>
<p>Train Random Forest.</p>
<pre class="r"><code>my_rf2 &lt;- randomForest(
  class ~ .,
  data = train2,
  importance = TRUE,
  do.trace = 100,
  ntree = 200
)</code></pre>
<pre><code>ntree      OOB      1      2
  100:  23.53% 25.82% 21.36%
  200:  23.56% 26.09% 21.16%</code></pre>
<pre class="r"><code>my_rf2</code></pre>
<pre><code>
Call:
 randomForest(formula = class ~ ., data = train2, importance = TRUE,      do.trace = 100, ntree = 200) 
               Type of random forest: classification
                     Number of trees: 200
No. of variables tried at each split: 8

        OOB estimate of  error rate: 23.56%
Confusion matrix:
       random   tss class.error
random  34802 12282   0.2608529
tss     10501 39125   0.2116028</code></pre>
<p>Feature importance.</p>
<pre class="r"><code>rn2 &lt;- round(importance(my_rf2), 2)
head(rn2[order(rn2[,3], decreasing=TRUE),])</code></pre>
<pre><code>      random   tss MeanDecreaseAccuracy MeanDecreaseGini
X4_CG  16.35 23.54                26.19           299.22
X1_AT  28.87  7.19                24.64           373.22
X2_CG  29.80 12.67                23.00          2091.22
X3_CA  23.48  7.72                22.45           323.41
X4_TA  18.52  9.46                20.97           142.35
X1_TG  19.69  5.41                20.80           328.67</code></pre>
<p>As a plot.</p>
<pre class="r"><code>varImpPlot(my_rf2)</code></pre>
<p><img src="figure/tss_predict.Rmd/importance_plot2-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Compare prediction results of our first model and our second model.</p>
<pre class="r"><code># results from first model
data.predict &lt;- predict(my_rf, test)
prop.table(table(observed = test$class, predict = data.predict), 1)</code></pre>
<pre><code>        predict
observed    random       tss
  random 0.7326808 0.2673192
  tss    0.2747627 0.7252373</code></pre>
<pre class="r"><code># results from second model
data.predict2 &lt;- predict(my_rf2, test2)
prop.table(table(observed = test2$class, predict = data.predict2), 1)</code></pre>
<pre><code>        predict
observed    random       tss
  random 0.7383343 0.2616657
  tss    0.2103513 0.7896487</code></pre>
<p>By including the positional information of the di-nucleotides, we have a better TSS classifier, which makes sense since not all the bases upstream and downstream of the TSS have a nucleotide preference and thus were diluting the signal in our first model where we calculated the overall di-nucleotide frequency.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.3 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] verification_1.42                 dtw_1.22-3                       
 [3] proxy_0.4-26                      CircStats_0.2-6                  
 [5] MASS_7.3-54                       boot_1.3-28                      
 [7] fields_13.3                       viridis_0.6.2                    
 [9] viridisLite_0.4.0                 spam_2.8-0                       
[11] pROC_1.18.0                       randomForest_4.7-1               
[13] BSgenome.Hsapiens.UCSC.hg38_1.4.4 BSgenome_1.62.0                  
[15] rtracklayer_1.54.0                Biostrings_2.62.0                
[17] XVector_0.34.0                    GenomicRanges_1.46.1             
[19] GenomeInfoDb_1.30.1               IRanges_2.28.0                   
[21] S4Vectors_0.32.4                  BiocGenerics_0.40.0              
[23] biomaRt_2.50.3                    forcats_0.5.1                    
[25] stringr_1.4.0                     dplyr_1.0.7                      
[27] purrr_0.3.4                       readr_2.1.1                      
[29] tidyr_1.1.4                       tibble_3.1.6                     
[31] ggplot2_3.3.5                     tidyverse_1.3.1                  
[33] workflowr_1.7.0                  

loaded via a namespace (and not attached):
  [1] colorspace_2.0-2            rjson_0.2.21               
  [3] ellipsis_0.3.2              rprojroot_2.0.2            
  [5] fs_1.5.2                    rstudioapi_0.13            
  [7] farver_2.1.0                bit64_4.0.5                
  [9] AnnotationDbi_1.56.2        fansi_1.0.2                
 [11] lubridate_1.8.0             xml2_1.3.3                 
 [13] cachem_1.0.6                knitr_1.37                 
 [15] jsonlite_1.7.3              Rsamtools_2.10.0           
 [17] broom_0.7.11                dbplyr_2.1.1               
 [19] png_0.1-7                   compiler_4.1.2             
 [21] httr_1.4.2                  backports_1.4.1            
 [23] assertthat_0.2.1            Matrix_1.3-4               
 [25] fastmap_1.1.0               cli_3.1.1                  
 [27] later_1.3.0                 htmltools_0.5.2            
 [29] prettyunits_1.1.1           tools_4.1.2                
 [31] dotCall64_1.0-1             gtable_0.3.0               
 [33] glue_1.6.1                  GenomeInfoDbData_1.2.7     
 [35] maps_3.4.0                  rappdirs_0.3.3             
 [37] Rcpp_1.0.8                  Biobase_2.54.0             
 [39] cellranger_1.1.0            jquerylib_0.1.4            
 [41] vctrs_0.3.8                 xfun_0.29                  
 [43] ps_1.6.0                    rvest_1.0.2                
 [45] lifecycle_1.0.1             restfulr_0.0.13            
 [47] XML_3.99-0.9                getPass_0.2-2              
 [49] zlibbioc_1.40.0             scales_1.1.1               
 [51] hms_1.1.1                   promises_1.2.0.1           
 [53] MatrixGenerics_1.6.0        parallel_4.1.2             
 [55] SummarizedExperiment_1.24.0 yaml_2.2.2                 
 [57] curl_4.3.2                  gridExtra_2.3              
 [59] memoise_2.0.1               stringi_1.7.6              
 [61] RSQLite_2.2.12              highr_0.9                  
 [63] BiocIO_1.4.0                filelock_1.0.2             
 [65] BiocParallel_1.28.3         rlang_1.0.0                
 [67] pkgconfig_2.0.3             bitops_1.0-7               
 [69] matrixStats_0.62.0          evaluate_0.14              
 [71] lattice_0.20-45             labeling_0.4.2             
 [73] GenomicAlignments_1.30.0    bit_4.0.4                  
 [75] processx_3.5.2              tidyselect_1.1.1           
 [77] plyr_1.8.7                  magrittr_2.0.2             
 [79] R6_2.5.1                    generics_0.1.1             
 [81] DelayedArray_0.20.0         DBI_1.1.2                  
 [83] pillar_1.6.5                haven_2.4.3                
 [85] whisker_0.4                 withr_2.4.3                
 [87] KEGGREST_1.34.0             RCurl_1.98-1.6             
 [89] modelr_0.1.8                crayon_1.4.2               
 [91] utf8_1.2.2                  BiocFileCache_2.2.1        
 [93] tzdb_0.2.0                  rmarkdown_2.11             
 [95] progress_1.2.2              grid_4.1.2                 
 [97] readxl_1.3.1                blob_1.2.2                 
 [99] callr_3.7.0                 git2r_0.29.0               
[101] reprex_2.0.1                digest_0.6.29              
[103] httpuv_1.6.5                munsell_0.5.0              </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
