---
title: "Using the cBioPortalData R package"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Following the [vignette](https://bioconductor.org/packages/release/bioc/vignettes/cBioPortalData/inst/doc/cBioPortalData.html).

>The cBioPortalData package was developed to allow programmatical access to the data resources available on the cBioPortal using R.

>The MultiAssayExperiment package is used to integrate, represent, and coordinate multiple experiments for the studies availble in the cBioPortal. This package in conjunction with curatedTCGAData give access to a large trove of publicly available bioinformatic data.

>The AnVIL is a cloud computing resource developed in part by the National Human Genome Research Institute. The AnVIL package provides end-user and developer functionality. For the end-user, AnVIL provides fast binary package installation, utitlities for working with Terra / AnVIL table and data resources, and convenient functions for file movement to and from Google cloud storage. For developers, AnVIL provides programatic access to the Terra, Leonardo, Rawls, and Dockstore RESTful programming interface, including helper functions to transform JSON responses to formats more amenable to manipulation in R.

## Packages

```{r install_packages, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("AnVIL")
BiocManager::install("cBioPortalData")
```

Load packages.

```{r load_packages, message=FALSE, warning=FALSE}
library("AnVIL")
library("cBioPortalData")

packageVersion("AnVIL")
packageVersion("cBioPortalData")
```

## Available studies

Identifying available studies.

```{r get_studies}
cbio <- cBioPortal()
studies <- getStudies(cbio, buildReport = TRUE)
dim(studies)
```

The `api_build` and `pack_build` columns will show the availability of each `studyId` for either download method: `cBioPortalData` and `cBioDataPack`, respectively.

```{r check_out_studies}
studies |>
  dplyr::select(studyId, allSampleCount, api_build, pack_build) |>
  head()
```

## Downloading

There are two main user-facing functions for downloading data from the cBioPortal API.

* `cBioDataPack` makes use of the tarball distribution of study data. This is useful when the user wants to download and analyze the entirety of the data as available from the cBioPortal.org website.

* `cBioPortalData` allows a more flexible approach to obtaining study data based on the available parameters such as molecular profile identifiers. This option is useful for users who have a set of gene symbols or identifiers and would like to get a smaller subset of the data that correspond to a particular molecular profile.

```{r ovarian_cancer}
idx <- grep("ovarian", studies$name, ignore.case = TRUE)
studies[idx, ] |>
  dplyr::select(studyId, allSampleCount, api_build, pack_build) |>
  head()
```

Download `ov_tcga` study.

```{r ov_tcga, message=FALSE, warning=FALSE}
ov_tcga <- cBioDataPack("ov_tcga", ask = FALSE)
ov_tcga
```

Info on ov_tcga_pan_can_atlas_2018.

```{r ov_tcga_pan_can_atlas_2018}
studies |>
  dplyr::filter(studyId == "ov_tcga_pan_can_atlas_2018") |>
  t()
```

## Metadata

Using `colData`.

```{r coldata}
colnames(colData(ov_tcga))
```

Check out the data.

```{r check_out_ov_tcga}
colData(ov_tcga)[1:6, 1:6]
```

Check what metadata is available.

```{r check_metadata_available}
apply(colData(ov_tcga), 2, function(x){
  y = factor(x != "[Not Available]" & x != "[Not Applicable]", levels = c(TRUE, FALSE))
  table(y)
}) |>
  rbind.data.frame() |>
  t()
```

Other metadata.

```{r metadata}
metadata(ov_tcga)
```

## Survival analysis

### Packages

Install packages for performing and visualising the survival analysis.

```{r install_sur_packages, eval=FALSE}
install.packages("survival")
install.packages("survminer")
```

Load packages.

```{r load_sur_packages, message=FALSE, warning=FALSE}
library("survival")
library("survminer")
```

### Prepare data

Disease free status.

```{r dfs_status}
table(colData(ov_tcga)$DFS_STATUS)
```

Disease free months.

```{r dfs_months}
head(colData(ov_tcga)$DFS_MONTHS)
```

Survival status.

```{r os_status}
table(colData(ov_tcga)$OS_STATUS)
```

Survival months.

```{r class_os_months}
class(colData(ov_tcga)$OS_MONTHS)
```

View data.

```{r col_data_ov_tcga}
colData(ov_tcga) |>
  as.data.frame() |>
  dplyr::select(DFS_STATUS, DFS_MONTHS, OS_STATUS, OS_MONTHS) |>
  head()
```

Prepare data.

```{r ov_surv, warning=FALSE, message=FALSE}
colData(ov_tcga) |>
  as.data.frame() |>
  dplyr::select(DFS_STATUS, DFS_MONTHS, OS_STATUS, OS_MONTHS, GRADE) |>
  dplyr::mutate(DFS_STATUS = ifelse(DFS_STATUS == "[Not Available]", NA, as.numeric(substr(DFS_STATUS, 1, 1)))) |>
  dplyr::mutate(DFS_MONTHS = ifelse(DFS_MONTHS == "[Not Available]", NA, as.numeric(DFS_MONTHS))) |>
  dplyr::mutate(OS_STATUS = ifelse(OS_STATUS == "[Not Available]", NA, as.numeric(substr(OS_STATUS, 1, 1)))) |>
  dplyr::mutate(OS_MONTHS = ifelse(OS_MONTHS == "[Not Available]", NA, as.numeric(OS_MONTHS))) |>
  dplyr::mutate(GRADE = ifelse(GRADE == "[Not Available]", NA, GRADE)) |>
  dplyr::mutate(GRADE = ifelse(GRADE == "G2" | GRADE == "G3", GRADE, NA)) -> ov_surv
```

We specify a simple survival model using tumour grade as a covariate and we draw the K-M plot. (The `substr` function creates a vector where 1 = deceased and 0 = living.)

```{r os_grade, fig.width=8, fig.height=8}
fit <- survfit(
    Surv(OS_MONTHS, as.numeric(substr(OS_STATUS, 1, 1))) ~ GRADE,
    data = ov_surv
)

ggsurvplot(
  fit,
  data = ov_surv,
  pval = TRUE,
  pval.method = TRUE,
  conf.int = TRUE,
  risk.table = TRUE,
  risk.table.col = "strata",
  linetype = "strata",
  surv.median.line = "hv",
  ggtheme = theme_bw(),
  ncensor.plot = TRUE
)
```

Disease free status.

```{r dfs_grade, fig.width=8, fig.height=8}
fit2 <- survfit(
    Surv(DFS_MONTHS, as.numeric(substr(DFS_STATUS, 1, 1))) ~ GRADE,
    data = ov_surv
)

ggsurvplot(
  fit2,
  data = ov_surv,
  pval = TRUE,
  pval.method = TRUE,
  conf.int = TRUE,
  risk.table = TRUE,
  risk.table.col = "strata",
  linetype = "strata",
  surv.median.line = "hv",
  ggtheme = theme_bw(),
  ncensor.plot = TRUE
)
```

Cox Proportional-Hazards Model.

```{r cox_os_survival}
grade_os_cox <- coxph(Surv(OS_MONTHS, OS_STATUS) ~ GRADE, data = ov_surv)
summary(grade_os_cox)
```

Having grade G3 has a hazard ratio of 1.17 but is not significant.

Repeat analysis using disease free survival.

```{r cox_dfs_survival}
grade_dfs_cox <- coxph(Surv(DFS_MONTHS, DFS_STATUS) ~ GRADE, data = ov_surv)
summary(grade_dfs_cox)
```

## Useful links

* [Multi-omic Integration of cBioPortal and TCGA data with MultiAssayExperiment](https://waldronlab.io/MultiAssayWorkshop/)
The last two columns will show the availability of each studyId for either download method (pack_build for cBioDataPack and api_build for cBioPortalData).

