---
title: "Downloading molecular signatures from MSigDB in R"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Following the [vignette](https://bioconductor.org/packages/release/data/experiment/vignettes/msigdb/inst/doc/msigdb.html).

>The molecular signatures database (MSigDB) is one of the largest collections of molecular signatures or gene expression signatures. A variety of gene expression signatures are hosted on this database including experimentally derived signatures and signatures representing pathways and ontologies from other curated databases. This rich collection of gene expression signatures (>25,000) can facilitate a wide variety of signature-based analyses, the most popular being gene set enrichment analyses. These signatures can be used to perform enrichment analysis in a DE experiment using tools such as {GSEA}, {fry} (from {limma}) and {camera} (from {limma}). Alternatively, they can be used to perform single-sample gene-set analysis of individual transcriptomic profiles using approaches such as {singscore}, {ssGSEA} and {GSVA}.
>
>This package provides the gene sets in the MSigDB in the form of `GeneSet` objects. This data structure is specifically designed to store information about gene sets, including their member genes and metadata. Other packages, such as {msigdbr} and {EGSEAdata} provide these gene sets too, however, they do so by storing them as lists or tibbles. These structures are not specific to gene sets therefore do not allow storage of important metadata associated with each gene set, for example, their short and long descriptions. Additionally, the lack of structure allows creation of invalid gene sets. Accessory functions implemented in the {GSEABase} package provide a neat interface to interact with `GeneSet` objects.

## Package

Install GSVA. (Dependencies are listed in the Imports section in the [DESCRIPTION](https://github.com/rcastelo/GSVA/blob/devel/DESCRIPTION) file.)

```{r install_fgsea, message=FALSE, warning=FALSE}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!require("msigdb", quietly = TRUE))
  BiocManager::install("msigdb")
```

Load package.

```{r load_msigdb, message=FALSE, warning=FALSE}
library(msigdb)
packageVersion("msigdb")
```

## Downloading the MSigDB database

In order to download the MSigDB database, we need to load {ExperimentHub} and {GSEABase}.

```{r load_additional, message=FALSE, warning=FALSE}
library(ExperimentHub)
library(GSEABase)
```

Query an `ExperimentHub` object.

```{r query_msigdb, warning=FALSE, message=FALSE}
eh <- ExperimentHub(ask = FALSE)
AnnotationHub::query(x = eh, pattern = 'msigdb')
```

Specify a more specific pattern to look for only human collections.

```{r query_msigdb_hs}
AnnotationHub::query(x = eh, pattern = 'msigdb.*hs.SYM')
```

The experiment hubs seem to be ordered from earliest to latest.

```{r query_msigdb_hs_latest}
AnnotationHub::query(x = eh, pattern = 'msigdb.*hs.SYM') |>
  tail(1) -> msigdb_hs_latest

names(msigdb_hs_latest)
msigdb_hs_latest
```

Data can be downloaded using the unique ID.

```{r download_hs_latest, warning=FALSE, message=FALSE}
eh[[names(msigdb_hs_latest)]]
```

Data can also be downloaded using `msigdb::getMsigdb()`.

```{r get_msigdb}
msigdb_ver <- sub(pattern = "msigdb.v(.*).hs.SYM", replacement = "\\1", msigdb_hs_latest$title)

msigdb.hs <- msigdb::getMsigdb(org = "hs", id = "SYM", version = msigdb_ver)
msigdb.hs
```

## Accessing the GeneSet and GeneSetCollection objects

A `GeneSetCollection` object is effectively a list and therefore all list processing functions work.

```{r msigdb_length}
length(msigdb.hs)
```

Each signature is stored in a `GeneSet` object and can be processed using functions from the {GSEABase} package.

```{r gene_ids}
gs <- msigdb.hs[[1000]]
geneIds(gs)
```

Details of a gene set.

```{r gs_details}
details(gs)
```

[Human collection types](https://www.gsea-msigdb.org/gsea/msigdb):

* H - hallmark gene sets are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.
* C1 - positional gene sets corresponding to human chromosome cytogenetic bands.
* C2 - curated gene sets from online pathway databases, publications in PubMed, and knowledge of domain experts.
* C3 - regulatory target gene sets based on gene target predictions for microRNA seed sequences and predicted transcription factor binding sites.
* C4 - computational gene sets defined by mining large collections of cancer-oriented expression data.
* C5 - ontology gene sets consist of genes annotated by the same ontology term.
* C6 - oncogenic signature gene sets defined directly from microarray gene expression data from cancer gene perturbations.
* C7 - immunologic signature gene sets represent cell states and perturbations within the immune system.
* C8 - cell type signature gene sets curated from cluster markers identified in single-cell sequencing studies of human tissue.

```{r collection_types}
table(sapply(lapply(msigdb.hs, collectionType), bcCategory))
```
